{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0749210f-5ca9-4a66-b740-d2dc650c082c",
   "metadata": {},
   "source": [
    "# Every Person Is Different\n",
    "\n",
    "1. Each person is different. We showed this previously (Lim YW, Schmieder R, Haynes M, Willner D, Furlan M, Youle M, Abbott K, Edwards R, Evangelista J, Conrad D, Rohwer F. 2013. Metagenomics and metatranscriptomics: windows on CF-associated viral and microbial communities. J Cyst Fibros 12:154–164) and I think Katrine and others have followed this up a lot, so it’s not that interesting a story arc, but needs commenting on.\n",
    "\n",
    "  a. There may be individual differences but also some microbiome generalizations based on: \n",
    "\n",
    "        i. pancreatic sufficiency vs insufficiency - see Jupyter/pancreatic sufficiency.ipynb\n",
    "        \n",
    "        ii. CF related diabetes/pre-diabetes/no diabetes\n",
    "        \n",
    "        iii. Age (we know that diversity tends to drop with age/declining lung function/increasing treatment)\n",
    "        \n",
    "        iv. Gender\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "464c107c-1329-4ed3-8d68-ca09025a04fa",
   "metadata": {},
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import re\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.colors import ListedColormap\n",
    "import matplotlib.dates as mdates\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import json\n",
    "\n",
    "from itertools import cycle\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "from adjustText import adjust_text\n",
    "\n",
    "# there is a FutureWarning in sklearn StandardScalar which is really annoying. This ignores it.\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)"
   ],
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "id": "fa9410c4-51dc-4689-921e-3c411ec60d34",
   "metadata": {},
   "source": [
    "# Read the data frames \n",
    "\n",
    "We also correct the names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f4179952-80e3-4e15-badb-934fbe717bd7",
   "metadata": {},
   "source": [
    "corrections = {\n",
    "    \"MGI\" : { \n",
    "        '1112926_20171212_S' : '1447437_20171212_S',\n",
    "        '1128691_20170206_S' : '1128691_20171206_S',\n",
    "        '1255498_20171212_S' : '1590009_20171212_S',\n",
    "        '1316979_20171215_S' : '1651490_20171215_S',\n",
    "        '1598281_20180508_S' : '1588281_20180508_S',\n",
    "        '1723809_20180227_S' : '1085876_20180227_S',\n",
    "        '649354_20170206_S' : '639354_20171206_S',\n",
    "        '652927_20180226_S' : '715927_20180226_S',\n",
    "        '658355_20180301_S' : '658355_20180327_S',\n",
    "        '777851_20170918_S' : '778851_20170918_S',\n",
    "        '788707_20181126_S' : '788707_20181129_S'\n",
    "    },\n",
    "    \"minion\" : {\n",
    "        '1112926_20171212_S' : '1447437_20171212_S',\n",
    "        '1255498_20171212_S' : '1590009_20171212_S',\n",
    "        '1316979_20171215_S' : '1651490_20171215_S',\n",
    "        '1598281_20180508_S' : '1588281_20180508_S',\n",
    "        '698917_20190119_S' : '698917_20180119_S'\n",
    "        }\n",
    "}\n",
    "\n",
    "def read_taxonomy(tax_file, firstchar, sequence_type):\n",
    "    \"\"\"\n",
    "    Read the taxonomy file and return a data frame\n",
    "    \"\"\"\n",
    "    \n",
    "    if sequence_type.lower() == 'mgi':\n",
    "        sequence_type = 'MGI'\n",
    "    elif sequence_type.lower() == 'minion':\n",
    "        sequence_type = 'minion'\n",
    "    else:\n",
    "        print(f\"Sorry. Don't know what {sequence_type} is supposed to be\", sys.stderr)\n",
    "        return None\n",
    "       \n",
    "    df = pd.read_csv(tax_file, sep='\\t', compression='gzip')\n",
    "    df = df[df['taxonomy'].str.contains('k__Bacteria')]\n",
    "    df = df[~df['taxonomy'].str.endswith(f'{firstchar}__')]\n",
    "    df = df.set_index('taxonomy')\n",
    "    df = df.rename(columns=corrections[sequence_type])\n",
    "    df.index = df.index.str.split(';').str[-1]\n",
    "    df = df.sort_index(axis=1)\n",
    "    return df\n",
    "\n",
    "def read_metadata(sequence_type):\n",
    "    if sequence_type.lower() == 'mgi':\n",
    "        sequence_type = 'MGI'\n",
    "    elif sequence_type.lower() == 'minion':\n",
    "        sequence_type = 'minion'\n",
    "    else:\n",
    "        print(f\"Sorry. Don't know what {sequence_type} is supposed to be\", sys.stderr)\n",
    "        return None\n",
    "    \n",
    "    metadata = pd.read_csv(\"../Metadata/Metadata.txt\", encoding='windows-1252', sep=\"\\t\", index_col=0)\n",
    "    metadata = metadata[~metadata[sequence_type].isna()]\n",
    "\n",
    "    for ix in metadata.index:\n",
    "        s = metadata.loc[ix, sequence_type]\n",
    "        if s in corrections[sequence_type]:\n",
    "            metadata.loc[ix, sequence_type] = corrections[sequence_type][s]\n",
    "    return metadata\n",
    "\n",
    "def sorted_presence_absence(df1, df2, minrowsum=0, asc_sort=False):\n",
    "    \"\"\"\n",
    "    Join the two tables and return the sorted version\n",
    "    \"\"\"\n",
    "    # filter so we only include samples sequenced on both MGI and MinION\n",
    "    common_columns = df1.columns.intersection(df2.columns)\n",
    "    df1_both = df1[common_columns]\n",
    "    df2_both = df2[common_columns]\n",
    "    \n",
    "    # create a presence/absence matrix\n",
    "    df1_presence = (df1_both > 0).astype(int)\n",
    "    df2_presence = (df2_both > 0).astype(int)*2\n",
    "\n",
    "    # here we filter on the minimum number of columns each taxa is in if requested\n",
    "    if minrowsum > 0:\n",
    "        df1_presence = df1_presence.loc[df1_presence[df1_presence.sum(axis=1) > minrowsum].index]\n",
    "        df2_presence = df2_presence.loc[df2_presence[df2_presence.sum(axis=1) > (2 * minrowsum)].index]\n",
    "    \n",
    "    # combine the two matrices and sort them\n",
    "    both = df1_presence.add(df2_presence, fill_value=0)\n",
    "    sboth = both.loc[both.sum(axis=1).sort_values(ascending=asc_sort).index]\n",
    "    sboth = sboth.sort_index(axis=1) # sort by column names\n",
    "\n",
    "    return sboth"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3f413595-9a6e-4d55-aaaf-d625aa1576de",
   "metadata": {},
   "source": [
    "tax='genus'\n",
    "# read the data \n",
    "mgi_df = read_taxonomy(f\"../MGI/Taxonomy/MGI_reads_{tax}.normalised.tsv.gz\", tax[0], \"mgi\")\n",
    "min_df = read_taxonomy(f\"../MinION/Taxonomy/Minion_read_based_annotations_{tax}.normalised.tsv.gz\", tax[0], \"minion\")\n",
    "mgi_metadata = read_metadata(\"MGI\")\n",
    "minion_metadata = read_metadata(\"minion\")"
   ],
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "id": "7c248d2c-eba0-4f27-a450-23da09f191ab",
   "metadata": {},
   "source": [
    "### Selecting MGI data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "43ff1062-2b1c-4b0a-9325-9c392e6b6451",
   "metadata": {},
   "source": [
    "sequence_type = \"MGI\"\n",
    "# sequence_type = \"minion\"\n",
    "\n",
    "if sequence_type == \"MGI\":\n",
    "    df = mgi_df.T\n",
    "    metadata = mgi_metadata\n",
    "else:\n",
    "    df = min_df.T\n",
    "    metadata = minion_metadata\n",
    "\n",
    "metadata = metadata[~metadata[sequence_type].isna()]"
   ],
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "id": "2cc87c3b-2b77-4e14-9bb0-c013d5894285",
   "metadata": {},
   "source": [
    "## Choosing some colours and markers\n",
    "\n",
    "There are 62 people! It will never work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "50d6959e-ae37-4052-9d5f-65681b6d9cc4",
   "metadata": {},
   "source": [
    "# Create a color palette with distinct colors for each person\n",
    "colours = sns.color_palette(\"Set3\", 12) * 6  # Repeating the Set3 palette to cover 62 unique people\n",
    "markers = ['o', 's', 'D', '^', 'v', 'p', '*', 'H', 'X', 'P', '+', 'x']  # A list of markers\n",
    "marker_cycle = cycle(markers)  # Cycle through the marker types\n"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1a1c1353-4873-4693-986a-b683d0586756",
   "metadata": {},
   "source": [
    "# Normalization/Transformation\n",
    "# Square root transform\n",
    "sqrt_df = np.sqrt(df)\n",
    "\n",
    "# perform the PCA\n",
    "pca = PCA(n_components=2)\n",
    "\n",
    "pca_result = pca.fit_transform(sqrt_df)\n",
    "# Create a DataFrame with PCA results\n",
    "pca_df = pd.DataFrame(data=pca_result, index=df.index, columns=['PC1', 'PC2'])\n",
    "\n",
    "# Get loadings\n",
    "loadings = pca.components_.T * np.sqrt(pca.explained_variance_)\n",
    "# Create a DataFrame for loadings with feature names\n",
    "# Select the top 5 most important loadings\n",
    "loadings_df = pd.DataFrame(loadings, index=df.columns, columns=['PC1', 'PC2'])\n",
    "\n",
    "# Create a DataFrame for top loadings\n",
    "top_loadings_df = loadings_df.loc[loadings_df['PC1'].abs().sort_values(ascending=False).index]\n",
    "top_loadings_df.head()\n",
    "\n",
    "\n",
    "explained_variance = pca.explained_variance_ratio_ * 100\n",
    "pc1_variance = explained_variance[0]\n",
    "pc2_variance = explained_variance[1]"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ae252d6-e30a-486d-b16a-d98a4d65cb8a",
   "metadata": {},
   "source": [
    "# Plot the PCA with individual markers and colors for each person\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# create a LUT for the people\n",
    "person_colour = {}\n",
    "for idx, person in enumerate(set(metadata.pwCF_ID.values)):\n",
    "    person_colour[person] = colours[idx]\n",
    "\n",
    "for idx, MGI in enumerate(pca_df.index):\n",
    "    person = metadata[metadata[\"MGI\"] == MGI][\"pwCF_ID\"].values[0]\n",
    "    \n",
    "    plt.scatter(pca_df.loc[MGI, 'PC1'], pca_df.loc[MGI, 'PC2'], \n",
    "                color=person_colour[person], label=person, \n",
    "                marker=next(marker_cycle), s=20, alpha=0.7)\n",
    "\n",
    "\n",
    "# Add labels and legend\n",
    "plt.title(f'{tax.capitalize()}-level taxonomic PCA of the MGI Sequencing by person')\n",
    "plt.xlabel(f'Principal Component 1 ({pc1_variance:.2f} % of the variation)')\n",
    "plt.ylabel(f'Principal Component 2 ({pc2_variance:.2f} % of the variation)')\n",
    "plt.legend([])\n",
    "# Display legend with the person labels\n",
    "#plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', markerscale=2)\n",
    "\n",
    "plt.savefig('img/every_person_pca.png')\n",
    "# Show plot\n",
    "plt.show()"
   ],
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "id": "241f7452-5bff-45bd-a0f6-2fd2cd363a8f",
   "metadata": {},
   "source": [
    "## Which people do we have more than once?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e7323edc-b2b1-47f5-96d5-30a5db7a8668",
   "metadata": {},
   "source": [
    "metadata.groupby(by='pwCF_ID').count().sort_values(by='MGI', ascending=False)"
   ],
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "id": "ef259413-073f-40d5-9b12-24d7f0baa39f",
   "metadata": {},
   "source": [
    "# Filter the DF for some of the people"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0e87e764-22b6-4e68-a7df-8b64b2bbd9e3",
   "metadata": {},
   "source": [
    "interesting_people = list(map(str, metadata.groupby(by='pwCF_ID').count().sort_values(by='MGI', ascending=False).index[0:6]))\n",
    "interesting_people"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "403dada6-1d3d-4b86-acdb-f42138cdb15a",
   "metadata": {},
   "source": [
    "interesing_pwCF = [x for x in mgi_df.columns if x.startswith(tuple(interesting_people))]\n",
    "interesing_pwCF_df = mgi_df[interesing_pwCF]"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3ac6133f-0459-44da-9043-f64f0902ca56",
   "metadata": {},
   "source": [
    "sns.displot(\n",
    "    data=interesing_pwCF_df.sort_values(by=interesing_pwCF[0]),\n",
    "    kind=\"kde\", height=6,\n",
    "    multiple=\"fill\", clip=(0, None),\n",
    "    palette=\"ch:rot=-.25,hue=1,light=.75\",\n",
    ")"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "13a2549f-85cf-4e0f-b5cd-de86aa7066b7",
   "metadata": {},
   "source": [
    "sns.heatmap(\n",
    "    data=interesing_pwCF_df.sort_values(by=interesing_pwCF[0]),\n",
    "    cmap='coolwarm_r'\n",
    ")"
   ],
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "id": "501e2fdf-dc1f-485e-a592-e0777f5ffd3d",
   "metadata": {},
   "source": [
    "# Plot a figure for just one person!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "18743bfe-8eb9-4e42-abb0-6cabac54d229",
   "metadata": {},
   "source": [
    "pta = [x for x in mgi_df.columns if x.startswith(interesting_people[0])]\n",
    "pta_df = mgi_df[pta]\n",
    "minrowsum = 1000\n",
    "pta_df = pta_df.loc[pta_df[pta_df.sum(axis=1) > minrowsum].index]"
   ],
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "id": "e0361196-cc21-4766-acc5-5aa0e67652a8",
   "metadata": {},
   "source": [
    "## Maybe a ridge plot?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a1672fd6-ad8b-4134-9ede-900149d89c8a",
   "metadata": {},
   "source": [
    "\n",
    "pwCF_ID = interesting_people[0]\n",
    "pta = [x for x in mgi_df.columns if x.startswith(pwCF_ID)]\n",
    "pta_df = mgi_df[pta]\n",
    "minrowsum = 1000\n",
    "pta_df = pta_df.loc[pta_df[pta_df.sum(axis=1) > minrowsum].index]\n",
    "\n",
    "\n",
    "renames = {x: int(x.replace(f'{pwCF_ID}_', '').replace('_S', '')) for x in pta_df.columns}\n",
    "pta_df = pta_df.rename(columns=renames)\n",
    "\n",
    "\n",
    "df_long = pd.melt(pta_df.reset_index(), id_vars='taxonomy', var_name='Date', value_name='Value')\n",
    "df_long['Date'] = pd.to_datetime(df_long['Date'], format='%Y%m%d')\n",
    "df_long['Date_Num'] = mdates.date2num(df_long['Date'])\n",
    "df_long"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cc258bd9-84dc-473f-acda-3e6511535085",
   "metadata": {},
   "source": [
    "# Get unique bacteria names\n",
    "bacteria_list = df_long['taxonomy'].unique()\n",
    "\n",
    "# Set up the plot\n",
    "fig, ax = plt.subplots(figsize=(12, 8))\n",
    "\n",
    "# Define vertical spacing between ridges\n",
    "spacing = 2.0\n",
    "\n",
    "# Plot each bacteria's data\n",
    "for i, bacteria in enumerate(bacteria_list):\n",
    "    # Subset the data\n",
    "    subset = df_long[df_long['taxonomy'] == bacteria]\n",
    "    # Normalize values for consistent plotting\n",
    "    values = subset['Value'] / subset['Value'].max()\n",
    "    # Offset for the ridge\n",
    "    offset = i * spacing\n",
    "    # Plot the ridge\n",
    "    ax.fill_between(subset['Date'], offset, values + offset, alpha=0.7)\n",
    "    # Add bacteria name\n",
    "    ax.text(subset['Date'].min(), offset + spacing / 2, bacteria, va='center', ha='right')\n",
    "\n",
    "# Adjust y-axis limits\n",
    "ax.set_ylim(-spacing, len(bacteria_list) * spacing)\n",
    "# Hide y-axis labels\n",
    "ax.set_yticks([])\n",
    "\n",
    "date_list = df_long['Date'].unique()\n",
    "# Set x-axis label\n",
    "plt.xlabel('Date')\n",
    "ax.set_xticks(date_list)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['left'].set_visible(False)\n",
    "#ax.spines['bottom'].set_visible(False)\n",
    "plt.title(f\"pwCF {pwCF_ID}\")\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8131c991-1fe4-4544-bdc1-523626c58bb0",
   "metadata": {},
   "source": [
    "fig, axes = plt.subplots(2, 3, figsize=(16, 22), sharey=True)\n",
    "\n",
    "# first, we filter out the bacteria to get the top X (currently 50) entries\n",
    "all_pts = [x for x in mgi_df.columns if any(x.startswith(y) for y in interesting_people)]\n",
    "pta_df = mgi_df[all_pts]\n",
    "# find the top 50 row sums\n",
    "minrowsum = sorted(pta_df.sum(axis=1), reverse=True)[50]\n",
    "pta_df = pta_df.loc[pta_df[pta_df.sum(axis=1) > minrowsum].index]\n",
    "\n",
    "bacteria_name = {x:x.replace(f\"{tax[0]}__\", \"\").replace(\" Family \", \"_\").replace(\"Candidatus \", \"\").split(\" \")[0] for x in pta_df.index.unique()}\n",
    "bacteria_name = dict(sorted(bacteria_name.items(), key=lambda item: item[1], reverse=True))\n",
    "bacteria_list = list(bacteria_name.keys())\n",
    "\n",
    "row = 0\n",
    "col = 0\n",
    "# Get unique bacteria names\n",
    "for i, pwCF_ID in enumerate(interesting_people):\n",
    "    \n",
    "    pta = [x for x in mgi_df.columns if x.startswith(pwCF_ID)]\n",
    "    pta_df = mgi_df[pta]\n",
    "    minrowsum = 1000\n",
    "    #pta_df = pta_df.loc[pta_df[pta_df.sum(axis=1) > minrowsum].index]\n",
    "    renames = {x: int(x.replace(f'{pwCF_ID}_', '').replace('_S', '')) for x in pta_df.columns}\n",
    "    pta_df = pta_df.rename(columns=renames)\n",
    "\n",
    "    df_long = pd.melt(pta_df.reset_index(), id_vars='taxonomy', var_name='Date', value_name='Value')\n",
    "    df_long['Date'] = pd.to_datetime(df_long['Date'], format='%Y%m%d')\n",
    "    df_long['Date_Num'] = mdates.date2num(df_long['Date'])\n",
    "    df_long\n",
    "\n",
    "    \n",
    "\n",
    "    ax = axes[row][col]\n",
    "\n",
    "    # Define vertical spacing between ridges\n",
    "    spacing = 2.0\n",
    "\n",
    "    # Plot each bacteria's data\n",
    "    for ix, bacteria in enumerate(bacteria_list):\n",
    "        # Subset the data\n",
    "        subset = df_long[df_long['taxonomy'] == bacteria]\n",
    "        # Normalize values for consistent plotting\n",
    "        values = subset['Value'] / subset['Value'].max()\n",
    "        # Offset for the ridge\n",
    "        offset = ix * spacing\n",
    "        # Plot the ridge\n",
    "        ax.fill_between(subset['Date'], offset, values + offset, alpha=0.7)\n",
    "        # Add bacteria name\n",
    "        if col == 0:\n",
    "            ax.text(subset['Date'].min(), offset + spacing / 2, bacteria_name[bacteria], va='center', ha='right')\n",
    "\n",
    "    # Adjust y-axis limits\n",
    "    ax.set_ylim(-spacing, len(bacteria_list) * spacing)\n",
    "    # Hide y-axis labels\n",
    "    ax.set_yticks([])\n",
    "\n",
    "    date_list = df_long['Date'].unique()\n",
    "    # Set x-axis label\n",
    "    plt.xlabel('Date')\n",
    "    ax.set_xticks(date_list)\n",
    "    plt.setp(ax.get_yticklabels(), fontsize=6)\n",
    "    plt.setp(ax.get_xticklabels(), rotation=45, ha=\"right\", rotation_mode=\"anchor\", fontsize=6)\n",
    "    \"\"\"\n",
    "    for j, label in enumerate(ax.get_xticklabels()):\n",
    "        if j % 2 == 0:\n",
    "            label.set_y(0)  # Position below x-axis\n",
    "        else:\n",
    "            label.set_visible(False)\n",
    "    \"\"\"\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    #ax.spines['bottom'].set_visible(False)\n",
    "    ax.title.set_text(f\"pwCF {pwCF_ID}\")  \n",
    "\n",
    "    if i == 2:\n",
    "        row = 1\n",
    "        col = 0\n",
    "    else:\n",
    "        col += 1\n",
    "\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "fig.align_xlabels()\n",
    "# Display the plot\n",
    "plt.savefig('img/every_person_is_different_mgi.svg')\n",
    "plt.savefig('img/every_person_is_different_mgi.png')\n",
    "plt.show()\n",
    "\n",
    "    "
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dbc75bc7-35bf-4da3-a010-cb03502a0495",
   "metadata": {},
   "source": [
    "fig, axes = plt.subplots(2, 3, figsize=(16, 22), sharey=True)\n",
    "\n",
    "# first, we filter out the bacteria to get the top X (currently 50) entries\n",
    "all_pts = [x for x in min_df.columns if any(x.startswith(y) for y in interesting_people)]\n",
    "pta_df = min_df[all_pts]\n",
    "# find the top 50 row sums\n",
    "minrowsum = sorted(pta_df.sum(axis=1), reverse=True)[50]\n",
    "pta_df = pta_df.loc[pta_df[pta_df.sum(axis=1) > minrowsum].index]\n",
    "\n",
    "bacteria_name = {x:x.replace(f\"{tax[0]}__\", \"\").replace(\" Family \", \"_\").replace(\"Candidatus \", \"\").split(\" \")[0] for x in pta_df.index.unique()}\n",
    "bacteria_name = dict(sorted(bacteria_name.items(), key=lambda item: item[1], reverse=True))\n",
    "bacteria_list = list(bacteria_name.keys())\n",
    "\n",
    "\n",
    "row = 0\n",
    "col = 0\n",
    "# Get unique bacteria names\n",
    "for i, pwCF_ID in enumerate(interesting_people):\n",
    "    \n",
    "    pta = [x for x in min_df.columns if x.startswith(pwCF_ID)]\n",
    "    pta_df = min_df[pta]\n",
    "    minrowsum = 1000\n",
    "    #pta_df = pta_df.loc[pta_df[pta_df.sum(axis=1) > minrowsum].index]\n",
    "    renames = {x: int(x.replace(f'{pwCF_ID}_', '').replace('_S', '')) for x in pta_df.columns}\n",
    "    pta_df = pta_df.rename(columns=renames)\n",
    "\n",
    "    df_long = pd.melt(pta_df.reset_index(), id_vars='taxonomy', var_name='Date', value_name='Value')\n",
    "    df_long['Date'] = pd.to_datetime(df_long['Date'], format='%Y%m%d')\n",
    "    df_long['Date_Num'] = mdates.date2num(df_long['Date'])\n",
    "    df_long\n",
    "\n",
    "    \n",
    "\n",
    "    ax = axes[row][col]\n",
    "\n",
    "    # Define vertical spacing between ridges\n",
    "    spacing = 2.0\n",
    "\n",
    "    # Plot each bacteria's data\n",
    "    for ix, bacteria in enumerate(bacteria_list):\n",
    "        # Subset the data\n",
    "        subset = df_long[df_long['taxonomy'] == bacteria]\n",
    "        # Normalize values for consistent plotting\n",
    "        values = subset['Value'] / subset['Value'].max()\n",
    "        # Offset for the ridge\n",
    "        offset = ix * spacing\n",
    "        # Plot the ridge\n",
    "        ax.fill_between(subset['Date'], offset, values + offset, alpha=0.7)\n",
    "        # Add bacteria name\n",
    "        if col == 0:\n",
    "            ax.text(subset['Date'].min(), offset + spacing / 2, bacteria_name[bacteria], va='center', ha='right')\n",
    "\n",
    "    # Adjust y-axis limits\n",
    "    ax.set_ylim(-spacing, len(bacteria_list) * spacing)\n",
    "    # Hide y-axis labels\n",
    "    ax.set_yticks([])\n",
    "\n",
    "    date_list = df_long['Date'].unique()\n",
    "    # Set x-axis label\n",
    "    plt.xlabel('Date')\n",
    "    ax.set_xticks(date_list)\n",
    "    plt.setp(ax.get_yticklabels(), fontsize=6)\n",
    "    plt.setp(ax.get_xticklabels(), rotation=45, ha=\"right\", rotation_mode=\"anchor\", fontsize=6)\n",
    "    \"\"\"\n",
    "    for j, label in enumerate(ax.get_xticklabels()):\n",
    "        if j % 2 == 0:\n",
    "            label.set_y(0)  # Position below x-axis\n",
    "        else:\n",
    "            label.set_visible(False)\n",
    "    \"\"\"\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    #ax.spines['bottom'].set_visible(False)\n",
    "    ax.title.set_text(f\"pwCF {pwCF_ID}\")  \n",
    "\n",
    "    if i == 2:\n",
    "        row = 1\n",
    "        col = 0\n",
    "    else:\n",
    "        col += 1\n",
    "\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "fig.align_xlabels()\n",
    "# Display the plot\n",
    "plt.savefig('img/every_person_is_different_minion.png')\n",
    "plt.show()\n",
    "\n",
    "    "
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e88eed6-3add-42e4-8432-58c1e01cbe00",
   "metadata": {},
   "source": [],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e759e15d-2eba-43b3-a1ac-46464fa54148",
   "metadata": {},
   "source": [],
   "outputs": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BioinfieConda",
   "language": "python",
   "name": "bioinfieconda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
