{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a7a213c0-97e1-4338-b943-3166b9e1f91a",
   "metadata": {},
   "source": [
    "# Linear Mixed Models\n",
    "\n",
    "Build some linear mixed models about our data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b54c76e8-0aaf-4da7-be76-3b56a655b8d2",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "780ba312-1f7d-4751-b01f-0718dc08222a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from socket import gethostname\n",
    "\n",
    "hostname = gethostname()\n",
    "\n",
    "import re\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.collections import PatchCollection\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.dates as mdates\n",
    "from matplotlib.colors import ListedColormap\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import json\n",
    "import random\n",
    "\n",
    "from itertools import cycle\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder, OrdinalEncoder\n",
    "from sklearn.inspection import permutation_importance\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier\n",
    "from sklearn.metrics import mean_squared_error\n",
    "\n",
    "from scipy.stats import linregress\n",
    "\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "from statsmodels.stats.outliers_influence import variance_inflation_factor\n",
    "\n",
    "\n",
    "# there is a FutureWarning in sklearn StandardScalar which is really annoying. This ignores it.\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "try:\n",
    "  import google.colab\n",
    "  IN_COLAB = True\n",
    "  !pip install adjustText\n",
    "  from google.colab import drive\n",
    "  drive.mount('/content/drive')\n",
    "  datadir = '/content/drive/MyDrive/Projects/CF/Adelaide/CF_Data_Analysis'\n",
    "except ImportError:\n",
    "  IN_COLAB = False\n",
    "  datadir = '..'\n",
    "\n",
    "\n",
    "if hostname.startswith('hpc-node'):\n",
    "    IN_DEEPTHOUGHT = True\n",
    "    sys.path.append('..')\n",
    "else:\n",
    "    IN_DEEPTHOUGHT = False\n",
    "\n",
    "import cf_analysis_lib\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d320c78c-0cb0-4bcf-af63-3bb8668872dc",
   "metadata": {},
   "source": [
    "## Read the data frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ecd087d9-65e7-4859-801f-019893047960",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The subsystems df has shape: (127, 769)\n",
      "The taxonomy df has shape: (127, 3581)\n",
      "The metadata df has shape: (127, 166)\n"
     ]
    }
   ],
   "source": [
    "sequence_type = \"MGI\"\n",
    "datadir = '..'\n",
    "#sslevel = 'level2_norm_ss.tsv.gz'\n",
    "sslevel = 'subsystems_norm_ss.tsv.gz'\n",
    "ss_df = cf_analysis_lib.read_subsystems(os.path.join(datadir, sequence_type, \"FunctionalAnalysis\", \"subsystems\", sslevel), sequence_type)\n",
    "ss_df = ss_df.T\n",
    "print(f\"The subsystems df has shape: {ss_df.shape}\")\n",
    "\n",
    "taxa = \"genus\"\n",
    "genus_otu = cf_analysis_lib.read_taxonomy(datadir, sequence_type, taxa)\n",
    "genus_otu = genus_otu.T\n",
    "print(f\"The taxonomy df has shape: {genus_otu.shape}\")\n",
    "metadata = cf_analysis_lib.read_metadata(datadir, sequence_type, categorise=True)\n",
    "print(f\"The metadata df has shape: {metadata.shape}\")\n",
    "\n",
    "df = ss_df.merge(genus_otu, left_index=True, right_index=True, how='inner')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11120eb9-bc3a-41b9-baaf-e2eafc646602",
   "metadata": {},
   "source": [
    "## Rename the columns\n",
    "\n",
    "Statsmodels can't handle column names with spaces or special characters, and so we rename them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9a58a052-34b9-4257-b44b-fe0654c6b535",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = [re.sub(r'\\W+', '_', col) for col in df.columns]\n",
    "metadata.columns = [re.sub(r'\\W+', '_', col) for col in metadata.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4c2d8609-5ccc-421b-bed5-032b9bf66c9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid</th>\n",
       "      <th>2_O_alpha_mannosyl_D_glycerate_utilization</th>\n",
       "      <th>2_aminophenol_Metabolism</th>\n",
       "      <th>2_ketoacid_oxidoreductases_disambiguation</th>\n",
       "      <th>2_oxoglutarate_dehydrogenase_</th>\n",
       "      <th>2_phosphoglycolate_salvage</th>\n",
       "      <th>3_amino_5_hydroxybenzoic_Acid_Synthesis</th>\n",
       "      <th>4_hydroxybenzoyl_CoA_reductase</th>\n",
       "      <th>5_methylaminomethyl_2_thiouridine</th>\n",
       "      <th>A_Hypothetical_Protein_Related_to_Proline_Metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Oceaniferula</th>\n",
       "      <th>Persicirhabdus</th>\n",
       "      <th>Phragmitibacter</th>\n",
       "      <th>Prosthecobacter</th>\n",
       "      <th>Roseibacillus</th>\n",
       "      <th>Roseimicrobium</th>\n",
       "      <th>Sulfuriroseicoccus</th>\n",
       "      <th>Verrucomicrobium</th>\n",
       "      <th>Eremiobacter</th>\n",
       "      <th>Methylomirabilis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1068841_20180306_S</th>\n",
       "      <td>10.085904</td>\n",
       "      <td>2784.895948</td>\n",
       "      <td>516.160945</td>\n",
       "      <td>231.778018</td>\n",
       "      <td>122.573392</td>\n",
       "      <td>342.327431</td>\n",
       "      <td>783.041862</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4136.703664</td>\n",
       "      <td>827.439614</td>\n",
       "      <td>...</td>\n",
       "      <td>1.347606</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.539042</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.539042</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1447437_20171212_S</th>\n",
       "      <td>59.260325</td>\n",
       "      <td>1065.438272</td>\n",
       "      <td>543.947408</td>\n",
       "      <td>554.136026</td>\n",
       "      <td>428.171446</td>\n",
       "      <td>301.915763</td>\n",
       "      <td>679.310468</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3913.676651</td>\n",
       "      <td>732.956657</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1128691_20171206_S</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>426.619709</td>\n",
       "      <td>912.767750</td>\n",
       "      <td>49.606943</td>\n",
       "      <td>213.640568</td>\n",
       "      <td>277.798880</td>\n",
       "      <td>423.312580</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2093.412992</td>\n",
       "      <td>236.459761</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 4350 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid  \\\n",
       "1068841_20180306_S                                      10.085904   \n",
       "1447437_20171212_S                                      59.260325   \n",
       "1128691_20171206_S                                       0.000000   \n",
       "\n",
       "                    2_O_alpha_mannosyl_D_glycerate_utilization  \\\n",
       "1068841_20180306_S                                 2784.895948   \n",
       "1447437_20171212_S                                 1065.438272   \n",
       "1128691_20171206_S                                  426.619709   \n",
       "\n",
       "                    2_aminophenol_Metabolism  \\\n",
       "1068841_20180306_S                516.160945   \n",
       "1447437_20171212_S                543.947408   \n",
       "1128691_20171206_S                912.767750   \n",
       "\n",
       "                    2_ketoacid_oxidoreductases_disambiguation  \\\n",
       "1068841_20180306_S                                 231.778018   \n",
       "1447437_20171212_S                                 554.136026   \n",
       "1128691_20171206_S                                  49.606943   \n",
       "\n",
       "                    2_oxoglutarate_dehydrogenase_  2_phosphoglycolate_salvage  \\\n",
       "1068841_20180306_S                     122.573392                  342.327431   \n",
       "1447437_20171212_S                     428.171446                  301.915763   \n",
       "1128691_20171206_S                     213.640568                  277.798880   \n",
       "\n",
       "                    3_amino_5_hydroxybenzoic_Acid_Synthesis  \\\n",
       "1068841_20180306_S                               783.041862   \n",
       "1447437_20171212_S                               679.310468   \n",
       "1128691_20171206_S                               423.312580   \n",
       "\n",
       "                    4_hydroxybenzoyl_CoA_reductase  \\\n",
       "1068841_20180306_S                             0.0   \n",
       "1447437_20171212_S                             0.0   \n",
       "1128691_20171206_S                             0.0   \n",
       "\n",
       "                    5_methylaminomethyl_2_thiouridine  \\\n",
       "1068841_20180306_S                        4136.703664   \n",
       "1447437_20171212_S                        3913.676651   \n",
       "1128691_20171206_S                        2093.412992   \n",
       "\n",
       "                    A_Hypothetical_Protein_Related_to_Proline_Metabolism  ...  \\\n",
       "1068841_20180306_S                                         827.439614     ...   \n",
       "1447437_20171212_S                                         732.956657     ...   \n",
       "1128691_20171206_S                                         236.459761     ...   \n",
       "\n",
       "                    Oceaniferula  Persicirhabdus  Phragmitibacter  \\\n",
       "1068841_20180306_S      1.347606             0.0              0.0   \n",
       "1447437_20171212_S      0.000000             0.0              0.0   \n",
       "1128691_20171206_S      0.000000             0.0              0.0   \n",
       "\n",
       "                    Prosthecobacter  Roseibacillus  Roseimicrobium  \\\n",
       "1068841_20180306_S              0.0       0.539042             0.0   \n",
       "1447437_20171212_S              0.0       0.000000             0.0   \n",
       "1128691_20171206_S              0.0       0.000000             0.0   \n",
       "\n",
       "                    Sulfuriroseicoccus  Verrucomicrobium  Eremiobacter  \\\n",
       "1068841_20180306_S                 0.0          0.539042           0.0   \n",
       "1447437_20171212_S                 0.0          0.000000           0.0   \n",
       "1128691_20171206_S                 0.0          0.000000           0.0   \n",
       "\n",
       "                    Methylomirabilis  \n",
       "1068841_20180306_S               0.0  \n",
       "1447437_20171212_S               0.0  \n",
       "1128691_20171206_S               0.0  \n",
       "\n",
       "[3 rows x 4350 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60c44b67-6ef6-4019-a371-66dd8371a421",
   "metadata": {},
   "source": [
    "#### Remove the columns that start with numbers\n",
    "\n",
    "The lmm breaks on these!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "73e569b9-47ff-4226-893a-b57d76a19952",
   "metadata": {},
   "outputs": [],
   "source": [
    "nos = {\"1\": \"one\", \"2\": \"two\", \"3\": \"three\", \"4\": \"four\", \"5\": \"five\", \"6\": \"six\", \"7\" : \"seven\", \"8\": \"eight\", \"9\" : \"nine\"}\n",
    "new_df_cols = {}\n",
    "for c in df.columns:\n",
    "    if c[0] in nos:\n",
    "        new_df_cols[c] = c.replace(c[0], nos[c[0]], 1)\n",
    "df = df.rename(columns=new_df_cols)\n",
    "\n",
    "newmetacols = {}\n",
    "for c in metadata.columns:\n",
    "    if c[0] in nos:\n",
    "        newmetacols[c] = c.replace(c[0], nos[c[0]], 1)\n",
    "metadata = metadata.rename(columns=newmetacols)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "997b5c90-a628-4d2a-be56-ad847b1b92dd",
   "metadata": {},
   "source": [
    "## Convert categories and objects to numeric forms\n",
    "\n",
    "We use numbers if we have them, we keep the MGI/MinION ID so we can use it to join tables, but we don't want to use it as a predictor.\n",
    "\n",
    "We convert categories to their codes so we get numbers. \n",
    "\n",
    "We drop the other columns so we don't try and predict on free text columns.\n",
    "\n",
    "I'm not sure what to do about dates, yet, so I ignore them for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2b41c3cf-14a7-4acb-a8db-9a020f2330d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping minion: object\n",
      "Dropping Sample_date: datetime64[ns]\n",
      "Dropping Room: object\n",
      "Dropping Culture_Result_Matched_with_sequenced_sample_: object\n",
      "Dropping Cutured_in_previous_12_months: object\n",
      "Dropping Others_cultured: object\n",
      "Dropping Spec_IgG: object\n",
      "Dropping CF_gene_2: object\n",
      "Dropping Notes: object\n",
      "Dropping Next_PA_Positive_Date: datetime64[ns]\n",
      "Dropping Culture_for_next_sample: object\n",
      "Dropping Antibiotics_duration_: object\n",
      "Dropping SAGC_ULN: object\n",
      "Dropping Index_I7: object\n",
      "Dropping Index_I5: object\n"
     ]
    }
   ],
   "source": [
    "# we make a copy of each data frame and then overwrite the columns. Internally this is more efficient than making a df one\n",
    "# column at a time.\n",
    "encoded_metadata = metadata.copy()\n",
    "to_delete = ['Pseudomonas_Culture'] # this is a duplicate columns with 'CS_Pseudomonas_aeruginosa'\n",
    "for col in metadata:\n",
    "    if pd.api.types.is_numeric_dtype(metadata[col]):\n",
    "        continue\n",
    "    elif col == sequence_type:\n",
    "        continue\n",
    "    elif metadata[col].dtypes == 'category':\n",
    "        encoded_metadata[col] = metadata[col].cat.codes\n",
    "    else:\n",
    "        print(f'Dropping {col}: {metadata[col].dtypes}')\n",
    "        to_delete.append(col)\n",
    "\n",
    "encoded_metadata = encoded_metadata.drop(columns=to_delete)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e592190-26ca-4bf2-892d-b76f750cbd86",
   "metadata": {},
   "source": [
    "## Remove or collapse highly correlated metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5bf7ffa3-4acd-4c22-9494-0e0a25e67ccd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        From                               To  Correlation\n",
      "303                 Hospital              Paediatric_vs_Adult     0.979930\n",
      "451                      Age                       Age_groups     0.964916\n",
      "1203          H2_Uncorrected                     H2_Corrected     0.974218\n",
      "1353         CH4_Uncorrected                    CH4_Corrected     0.968859\n",
      "2851                     FVC                             FEV1     0.939640\n",
      "3001                    FEV1                        Best_FEV1     0.932290\n",
      "7201    CS_Aspergillus_niger           CS_Aspergillus_terreus     1.000000\n",
      "11552  N12M_M_intracellulare  N12M_Achromobacter_xylosoxidans     0.901607\n",
      "21151     Sum_of_antibiotics                      Sum_of_meds     0.937157\n"
     ]
    }
   ],
   "source": [
    "correlation_matrix = encoded_metadata.drop(columns=sequence_type).corr(method='pearson')\n",
    "\n",
    "# Find highly correlated pairs (absolute correlation > 0.8)\n",
    "high_corr_metadata = correlation_matrix.unstack().reset_index()\n",
    "high_corr_metadata.columns = ['From', 'To', 'Correlation']\n",
    "high_corr_metadata = high_corr_metadata[\n",
    "    (high_corr_metadata['Correlation'].abs() > 0.9) & (high_corr_metadata['From'] != high_corr_metadata['To'])\n",
    "]\n",
    "\n",
    "# Drop duplicate pairs (e.g., (A, B) and (B, A))\n",
    "high_corr_metadata = high_corr_metadata.drop_duplicates(subset=['Correlation'])\n",
    "\n",
    "print(high_corr_metadata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "247d04ca-92c3-4e10-acad-5d10012d3e66",
   "metadata": {},
   "outputs": [],
   "source": [
    "encoded_metadata = encoded_metadata.drop(columns=high_corr_metadata['To'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34002072-251a-4fac-ad2f-013fcf0b5e35",
   "metadata": {},
   "source": [
    "## Remove or collapse highly correlated abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7ba71a6c-72c6-429a-b2ce-92b953b89083",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping 68 highly correlated variables\n"
     ]
    }
   ],
   "source": [
    "correlation_matrix = df.corr(method='pearson')\n",
    "\n",
    "# Find highly correlated pairs (absolute correlation > 0.8)\n",
    "high_corr_abundance = correlation_matrix.unstack().reset_index()\n",
    "high_corr_abundance.columns = ['From', 'To', 'Correlation']\n",
    "high_corr_abundance = high_corr_abundance[\n",
    "    (high_corr_abundance['Correlation'].abs() > 0.9999) & (high_corr_abundance['From'] != high_corr_abundance['To'])\n",
    "]\n",
    "\n",
    "# Drop duplicate pairs (e.g., (A, B) and (B, A))\n",
    "high_corr_abundance = high_corr_abundance.drop_duplicates(subset=['Correlation'])\n",
    "\n",
    "print(f\"Dropping {len(high_corr_abundance['To'])} highly correlated variables\")\n",
    "df = df.drop(columns=high_corr_abundance['To'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "220248b4-7cea-4117-b614-42b49cba9320",
   "metadata": {},
   "source": [
    "# Predict something \"simple\"\n",
    "\n",
    "Here, we are just trying to predict Pseudomonas positive in the next 12 months!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2bbff15f-b949-43ac-a018-b84d95d544e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df: (127, 4317)\n",
      "metadata: (127, 166)\n",
      "df_combined: (127, 4459)\n"
     ]
    }
   ],
   "source": [
    "df_combined = df.merge(encoded_metadata, left_index=True, right_index=True, how='inner')\n",
    "df_combined = df_combined.dropna(subset=['N12M_Pseudomonas_aeruginosa', \"pwCF_ID\"])\n",
    "df_combined[\"N12M_Pseudomonas_aeruginosa\"] = df_combined[\"N12M_Pseudomonas_aeruginosa\"].astype(int)\n",
    "df_combined = df_combined.reset_index()\n",
    "\n",
    "print(f\"df: {df.shape}\")\n",
    "print(f\"metadata: {metadata.shape}\")\n",
    "print(f\"df_combined: {df_combined.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2a532c44-0036-4bb4-b67c-d029b3608139",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_column_number = 100\n",
    "random_columns = np.random.choice(df.columns, size=random_column_number, replace=False)\n",
    "#random_columns = ['Pedobacter']\n",
    "predictors = \" + \".join(random_columns)  # Concatenate all bacterial abundance column names\n",
    "formula = f\"N12M_Pseudomonas_aeruginosa ~ {predictors} + CS_Pseudomonas_aeruginosa\"\n",
    "all_columns = random_columns.tolist() + [\"N12M_Pseudomonas_aeruginosa\", \"CS_Pseudomonas_aeruginosa\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b82d9dc-83d0-4c1f-8f42-083acfacaabe",
   "metadata": {},
   "source": [
    "### Check for any NA values in our data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "86e4f33c-5048-43c8-9aed-ed0909170c09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined[all_columns].columns[df_combined[all_columns].isna().any() | df_combined[all_columns].isin([np.inf, -np.inf]).any()].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c864b5c1-c03f-448e-a591-702c5791a760",
   "metadata": {},
   "source": [
    "## Use PCA to reduce the dimensionality of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f850070e-19d2-4b98-8527-7f96782c995c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>two_3_diacetamido_2_3_dideoxy_d_mannuronic_acid</th>\n",
       "      <th>two_O_alpha_mannosyl_D_glycerate_utilization</th>\n",
       "      <th>two_aminophenol_Metabolism</th>\n",
       "      <th>two_ketoacid_oxidoreductases_disambiguation</th>\n",
       "      <th>two_oxoglutarate_dehydrogenase_</th>\n",
       "      <th>two_phosphoglycolate_salvage</th>\n",
       "      <th>three_amino_5_hydroxybenzoic_Acid_Synthesis</th>\n",
       "      <th>four_hydroxybenzoyl_CoA_reductase</th>\n",
       "      <th>five_methylaminomethyl_2_thiouridine</th>\n",
       "      <th>...</th>\n",
       "      <th>Sum_of_steroid_mabs</th>\n",
       "      <th>DNA_extraction__conc</th>\n",
       "      <th>DNA_Conc_ng_ul_</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total_Clusters_Passing_Filter_Million_</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>10.085904</td>\n",
       "      <td>2784.895948</td>\n",
       "      <td>516.160945</td>\n",
       "      <td>231.778018</td>\n",
       "      <td>122.573392</td>\n",
       "      <td>342.327431</td>\n",
       "      <td>783.041862</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4136.703664</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1.070</td>\n",
       "      <td>42.800</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "      <td>99372.537072</td>\n",
       "      <td>21836.600136</td>\n",
       "      <td>759.873855</td>\n",
       "      <td>-3885.516024</td>\n",
       "      <td>2646.814699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1447437_20171212_S</td>\n",
       "      <td>59.260325</td>\n",
       "      <td>1065.438272</td>\n",
       "      <td>543.947408</td>\n",
       "      <td>554.136026</td>\n",
       "      <td>428.171446</td>\n",
       "      <td>301.915763</td>\n",
       "      <td>679.310468</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3913.676651</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1.510</td>\n",
       "      <td>32.600</td>\n",
       "      <td>498</td>\n",
       "      <td>4.7</td>\n",
       "      <td>-73297.506592</td>\n",
       "      <td>-171978.570617</td>\n",
       "      <td>159669.904865</td>\n",
       "      <td>58545.595994</td>\n",
       "      <td>17062.589621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1128691_20171206_S</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>426.619709</td>\n",
       "      <td>912.767750</td>\n",
       "      <td>49.606943</td>\n",
       "      <td>213.640568</td>\n",
       "      <td>277.798880</td>\n",
       "      <td>423.312580</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2093.412992</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1.750</td>\n",
       "      <td>33.400</td>\n",
       "      <td>535</td>\n",
       "      <td>4.5</td>\n",
       "      <td>-215950.928538</td>\n",
       "      <td>-161130.110407</td>\n",
       "      <td>-304256.714750</td>\n",
       "      <td>278876.087314</td>\n",
       "      <td>85151.898742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1128691_20171218_S</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>659.087578</td>\n",
       "      <td>864.137047</td>\n",
       "      <td>139.140711</td>\n",
       "      <td>235.318676</td>\n",
       "      <td>355.174973</td>\n",
       "      <td>538.254855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2050.494687</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3.400</td>\n",
       "      <td>30.800</td>\n",
       "      <td>352</td>\n",
       "      <td>4.8</td>\n",
       "      <td>-193031.382090</td>\n",
       "      <td>-150311.164287</td>\n",
       "      <td>-251948.106696</td>\n",
       "      <td>240655.956824</td>\n",
       "      <td>78552.608901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1128691_20180116_S</td>\n",
       "      <td>14.478968</td>\n",
       "      <td>159.268644</td>\n",
       "      <td>593.637673</td>\n",
       "      <td>48.263225</td>\n",
       "      <td>358.354449</td>\n",
       "      <td>202.705547</td>\n",
       "      <td>348.701804</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1570.967988</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.708</td>\n",
       "      <td>27.400</td>\n",
       "      <td>500</td>\n",
       "      <td>3.8</td>\n",
       "      <td>-236806.519898</td>\n",
       "      <td>-165494.526901</td>\n",
       "      <td>-311704.727626</td>\n",
       "      <td>272863.998846</td>\n",
       "      <td>89995.184983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>895293_20180502_S</td>\n",
       "      <td>11.343610</td>\n",
       "      <td>250.504720</td>\n",
       "      <td>1278.046721</td>\n",
       "      <td>264.054032</td>\n",
       "      <td>86.589556</td>\n",
       "      <td>229.708102</td>\n",
       "      <td>753.089661</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3424.824904</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.442</td>\n",
       "      <td>16.900</td>\n",
       "      <td>584</td>\n",
       "      <td>4.6</td>\n",
       "      <td>89868.551921</td>\n",
       "      <td>14859.996373</td>\n",
       "      <td>-7895.488631</td>\n",
       "      <td>-5847.483561</td>\n",
       "      <td>15289.446458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>896213_20180427_S</td>\n",
       "      <td>63.502022</td>\n",
       "      <td>2074.399372</td>\n",
       "      <td>465.681492</td>\n",
       "      <td>134.059823</td>\n",
       "      <td>136.882135</td>\n",
       "      <td>328.093778</td>\n",
       "      <td>1035.435741</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3407.941825</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.436</td>\n",
       "      <td>477</td>\n",
       "      <td>0.2</td>\n",
       "      <td>-31332.700478</td>\n",
       "      <td>24224.658597</td>\n",
       "      <td>-10936.213508</td>\n",
       "      <td>3919.906686</td>\n",
       "      <td>-48850.089953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>913873_20180417_S</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2031.296007</td>\n",
       "      <td>532.006097</td>\n",
       "      <td>145.092572</td>\n",
       "      <td>120.910477</td>\n",
       "      <td>447.368763</td>\n",
       "      <td>890.707178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3192.036583</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.162</td>\n",
       "      <td>863</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1846.767313</td>\n",
       "      <td>70152.638463</td>\n",
       "      <td>11545.295524</td>\n",
       "      <td>24828.061940</td>\n",
       "      <td>-20604.331376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>980574_20180403_S</td>\n",
       "      <td>15.281074</td>\n",
       "      <td>1205.624053</td>\n",
       "      <td>444.731949</td>\n",
       "      <td>383.080719</td>\n",
       "      <td>119.209942</td>\n",
       "      <td>228.425711</td>\n",
       "      <td>929.423029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3243.276243</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.216</td>\n",
       "      <td>34.600</td>\n",
       "      <td>364</td>\n",
       "      <td>2.2</td>\n",
       "      <td>-8912.360273</td>\n",
       "      <td>89645.220424</td>\n",
       "      <td>14458.018251</td>\n",
       "      <td>-1053.182273</td>\n",
       "      <td>26759.977336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>983493_20180123_S</td>\n",
       "      <td>490.780540</td>\n",
       "      <td>556.217945</td>\n",
       "      <td>1145.154592</td>\n",
       "      <td>812.514449</td>\n",
       "      <td>257.278065</td>\n",
       "      <td>498.960215</td>\n",
       "      <td>452.608720</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2574.961898</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.698</td>\n",
       "      <td>21.600</td>\n",
       "      <td>469</td>\n",
       "      <td>4.4</td>\n",
       "      <td>-53904.030899</td>\n",
       "      <td>-88562.758637</td>\n",
       "      <td>-117509.952258</td>\n",
       "      <td>57285.790006</td>\n",
       "      <td>37141.200905</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows Ã— 4464 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  index  two_3_diacetamido_2_3_dideoxy_d_mannuronic_acid  \\\n",
       "0    1068841_20180306_S                                        10.085904   \n",
       "1    1447437_20171212_S                                        59.260325   \n",
       "2    1128691_20171206_S                                         0.000000   \n",
       "3    1128691_20171218_S                                         0.000000   \n",
       "4    1128691_20180116_S                                        14.478968   \n",
       "..                  ...                                              ...   \n",
       "122   895293_20180502_S                                        11.343610   \n",
       "123   896213_20180427_S                                        63.502022   \n",
       "124   913873_20180417_S                                         0.000000   \n",
       "125   980574_20180403_S                                        15.281074   \n",
       "126   983493_20180123_S                                       490.780540   \n",
       "\n",
       "     two_O_alpha_mannosyl_D_glycerate_utilization  two_aminophenol_Metabolism  \\\n",
       "0                                     2784.895948                  516.160945   \n",
       "1                                     1065.438272                  543.947408   \n",
       "2                                      426.619709                  912.767750   \n",
       "3                                      659.087578                  864.137047   \n",
       "4                                      159.268644                  593.637673   \n",
       "..                                            ...                         ...   \n",
       "122                                    250.504720                 1278.046721   \n",
       "123                                   2074.399372                  465.681492   \n",
       "124                                   2031.296007                  532.006097   \n",
       "125                                   1205.624053                  444.731949   \n",
       "126                                    556.217945                 1145.154592   \n",
       "\n",
       "     two_ketoacid_oxidoreductases_disambiguation  \\\n",
       "0                                     231.778018   \n",
       "1                                     554.136026   \n",
       "2                                      49.606943   \n",
       "3                                     139.140711   \n",
       "4                                      48.263225   \n",
       "..                                           ...   \n",
       "122                                   264.054032   \n",
       "123                                   134.059823   \n",
       "124                                   145.092572   \n",
       "125                                   383.080719   \n",
       "126                                   812.514449   \n",
       "\n",
       "     two_oxoglutarate_dehydrogenase_  two_phosphoglycolate_salvage  \\\n",
       "0                         122.573392                    342.327431   \n",
       "1                         428.171446                    301.915763   \n",
       "2                         213.640568                    277.798880   \n",
       "3                         235.318676                    355.174973   \n",
       "4                         358.354449                    202.705547   \n",
       "..                               ...                           ...   \n",
       "122                        86.589556                    229.708102   \n",
       "123                       136.882135                    328.093778   \n",
       "124                       120.910477                    447.368763   \n",
       "125                       119.209942                    228.425711   \n",
       "126                       257.278065                    498.960215   \n",
       "\n",
       "     three_amino_5_hydroxybenzoic_Acid_Synthesis  \\\n",
       "0                                     783.041862   \n",
       "1                                     679.310468   \n",
       "2                                     423.312580   \n",
       "3                                     538.254855   \n",
       "4                                     348.701804   \n",
       "..                                           ...   \n",
       "122                                   753.089661   \n",
       "123                                  1035.435741   \n",
       "124                                   890.707178   \n",
       "125                                   929.423029   \n",
       "126                                   452.608720   \n",
       "\n",
       "     four_hydroxybenzoyl_CoA_reductase  five_methylaminomethyl_2_thiouridine  \\\n",
       "0                                  0.0                           4136.703664   \n",
       "1                                  0.0                           3913.676651   \n",
       "2                                  0.0                           2093.412992   \n",
       "3                                  0.0                           2050.494687   \n",
       "4                                  0.0                           1570.967988   \n",
       "..                                 ...                                   ...   \n",
       "122                                0.0                           3424.824904   \n",
       "123                                0.0                           3407.941825   \n",
       "124                                0.0                           3192.036583   \n",
       "125                                0.0                           3243.276243   \n",
       "126                                0.0                           2574.961898   \n",
       "\n",
       "     ...  Sum_of_steroid_mabs  DNA_extraction__conc  DNA_Conc_ng_ul_  \\\n",
       "0    ...                    0                 1.070           42.800   \n",
       "1    ...                    0                 1.510           32.600   \n",
       "2    ...                    0                 1.750           33.400   \n",
       "3    ...                    0                 3.400           30.800   \n",
       "4    ...                    0                 0.708           27.400   \n",
       "..   ...                  ...                   ...              ...   \n",
       "122  ...                    0                 0.442           16.900   \n",
       "123  ...                    0                 0.000            0.436   \n",
       "124  ...                    0                 0.000            0.162   \n",
       "125  ...                    0                 0.216           34.600   \n",
       "126  ...                    0                 0.698           21.600   \n",
       "\n",
       "     Mean_Size_BP  Total_Clusters_Passing_Filter_Million_            PC1  \\\n",
       "0             417                                     2.8   99372.537072   \n",
       "1             498                                     4.7  -73297.506592   \n",
       "2             535                                     4.5 -215950.928538   \n",
       "3             352                                     4.8 -193031.382090   \n",
       "4             500                                     3.8 -236806.519898   \n",
       "..            ...                                     ...            ...   \n",
       "122           584                                     4.6   89868.551921   \n",
       "123           477                                     0.2  -31332.700478   \n",
       "124           863                                     0.0   -1846.767313   \n",
       "125           364                                     2.2   -8912.360273   \n",
       "126           469                                     4.4  -53904.030899   \n",
       "\n",
       "               PC2            PC3            PC4           PC5  \n",
       "0     21836.600136     759.873855   -3885.516024   2646.814699  \n",
       "1   -171978.570617  159669.904865   58545.595994  17062.589621  \n",
       "2   -161130.110407 -304256.714750  278876.087314  85151.898742  \n",
       "3   -150311.164287 -251948.106696  240655.956824  78552.608901  \n",
       "4   -165494.526901 -311704.727626  272863.998846  89995.184983  \n",
       "..             ...            ...            ...           ...  \n",
       "122   14859.996373   -7895.488631   -5847.483561  15289.446458  \n",
       "123   24224.658597  -10936.213508    3919.906686 -48850.089953  \n",
       "124   70152.638463   11545.295524   24828.061940 -20604.331376  \n",
       "125   89645.220424   14458.018251   -1053.182273  26759.977336  \n",
       "126  -88562.758637 -117509.952258   57285.790006  37141.200905  \n",
       "\n",
       "[127 rows x 4464 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pca = PCA(n_components=5)\n",
    "pca_components = pca.fit_transform(df_combined[df.columns])  # Add bacterial abundance columns\n",
    "pca_df = pd.DataFrame(pca_components, columns=[\"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\"])\n",
    "\n",
    "# Add PCA components to the combined DataFrame\n",
    "df_combined2 = pd.concat([df_combined.reset_index(drop=True), pca_df], axis=1)\n",
    "df_combined2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ad01662a-ccea-4d6b-951e-5fa715c8ffed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  Mixed Linear Model Regression Results\n",
      "=========================================================================\n",
      "Model:            MixedLM Dependent Variable: N12M_Pseudomonas_aeruginosa\n",
      "No. Observations: 127     Method:             REML                       \n",
      "No. Groups:       64      Scale:              0.0212                     \n",
      "Min. group size:  1       Log-Likelihood:     -120.4513                  \n",
      "Max. group size:  7       Converged:          Yes                        \n",
      "Mean group size:  2.0                                                    \n",
      "-----------------------------------------------------------------------------\n",
      "               Coef.      Std.Err.       z        P>|z|     [0.025     0.975]\n",
      "-----------------------------------------------------------------------------\n",
      "Intercept       0.441        0.089      4.984     0.000      0.268      0.615\n",
      "PC1            -0.000        0.000     -2.318     0.020     -0.000     -0.000\n",
      "PC2            -0.000        0.000     -3.338     0.001     -0.000     -0.000\n",
      "PC3             0.000        0.000      0.045     0.964     -0.000      0.000\n",
      "PC4             0.000        0.000      2.394     0.017      0.000      0.000\n",
      "PC5             0.000        0.000      3.210     0.001      0.000      0.000\n",
      "Group Var       0.486        0.896                                           \n",
      "=========================================================================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "formula = \"N12M_Pseudomonas_aeruginosa ~ PC1 + PC2 + PC3 + PC4 + PC5\"\n",
    "# Fit the GLMM\n",
    "model = smf.mixedlm(\n",
    "    formula=formula,\n",
    "    data=df_combined2,\n",
    "    groups=df_combined2[\"pwCF_ID\"],  # Grouping by pwCF_ID for random effects\n",
    ")\n",
    "#family=sm.families.Binomial()  # Logistic link function\n",
    "result = model.fit()\n",
    "\n",
    "# Print the model summary\n",
    "print(result.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7753552b-08a2-4850-a69d-37488fef0338",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N12M_Pseudomonas_aeruginosa ~ MukBEF_Chromosome_Condensation + Gryllotalpicola + Soehngenia + Dactylopiibacterium + Thermodesulfobium + Heme_O_and_Heme_A_biosynthesis + Mogibacterium + Inositol_catabolism + Resistance_to_the_fluoroquinolones_norfloxacin_and_ciprofloxacin + Hongsoonwoonella + Puniceispirillum + Efflux_ABC_transporters_BmrCD_and_LmrCD_involved_in_multidrug_resistance + Thermocrinis + tRNA_aminoacylation_Gln + Roseovarius + Escherichia + Portiera + Symbiobacterium + Arcanobacterium + Sedimenticola + Brasilonema + Spelaeicoccus + Phycobacter + Orientia + Anaerobic_dimethyl_sulfoxide_reductase_EC_1_8_5_3_ + Urea_carboxylase_and_Allophanate_hydrolase_cluster + Acidicaldus + Fluviispira + Botrimarina + Mycosynbacter + Oleiphilus + Brotolimicola + Anaerofilum + Roseofilum + Beijerinckia + Thiofilum + Draconibacterium + Marinisporobacter + Phenazine_biosynthesis + Isoptericola + Hepatoplasma + Caldimicrobium + Rhodobaculum + Anaerotignum + Persicitalea + Cytobacillus + Boudabousia + Criibacterium + Ferriphaselus + Bacilliculturomica + Dolosigranulum + Atopococcus + Heliophilum + Kaustia + Microaerobacter + Translation_elongation_factors_bacterial + Ethanoligenens + Gabonia + Spirillospora + Neglectibacter + Hapalosiphon + Martelella + Gorillibacterium + Anoxybacter + Mammaliicoccus + Pontivivens + Hespellia + Marasmitruncus + Head_to_head_olefinic_hydrocarbon_biosynthesis + Gaetbulibacter + Croceicoccus + Nitrosospira + Histidine_Degradation + Chloroflexus + Fangia + Limnoglobus + Alkalilimnicola + Translation_elongation_factor_G_family + Biotin_synthesis_cluster + Pedobacter + Rhizobium + Xylophilus + Mupirocin_resistance + Kinetoplastibacterium + Thaumasiovibrio + Terriglobus + Desulfogranum + Cardiolipin_biosynthesis + Thermosulfidibacter + Williamsia + Biotin_synthesis_utilization + Desulfohalovibrio + Ornithinibacillus + Mesobaculum + Actinocrispum + Thermobacillus + Mycofactocin_system_cluster + Kangiella + TAM_transport_system + Longitalea + CS_Pseudomonas_aeruginosa\n"
     ]
    }
   ],
   "source": [
    "#formula = f\"N12M_Pseudomonas_aeruginosa ~ {predictors.replace('2_oxoglutarate_dehydrogenase_ +', '')} + CS_Pseudomonas_aeruginosa\"\n",
    "formula = f\"N12M_Pseudomonas_aeruginosa ~ {predictors} + CS_Pseudomonas_aeruginosa\"\n",
    "print(formula)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "04378ee0-b921-42dc-b5ec-315d6d775f7a",
   "metadata": {},
   "outputs": [
    {
     "ename": "LinAlgError",
     "evalue": "Singular matrix",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLinAlgError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 9\u001b[0m\n\u001b[1;32m      3\u001b[0m model \u001b[38;5;241m=\u001b[39m smf\u001b[38;5;241m.\u001b[39mmixedlm(\n\u001b[1;32m      4\u001b[0m     formula\u001b[38;5;241m=\u001b[39mformula,\n\u001b[1;32m      5\u001b[0m     data\u001b[38;5;241m=\u001b[39mdf_combined,\n\u001b[1;32m      6\u001b[0m     groups\u001b[38;5;241m=\u001b[39mdf_combined[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpwCF_ID\u001b[39m\u001b[38;5;124m\"\u001b[39m],  \u001b[38;5;66;03m# Grouping by pwCF_ID for random effects\u001b[39;00m\n\u001b[1;32m      7\u001b[0m )\n\u001b[1;32m      8\u001b[0m \u001b[38;5;66;03m#family=sm.families.Binomial()  # Logistic link function\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;66;03m# Print the model summary\u001b[39;00m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28mprint\u001b[39m(result\u001b[38;5;241m.\u001b[39msummary())\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/regression/mixed_linear_model.py:2191\u001b[0m, in \u001b[0;36mMixedLM.fit\u001b[0;34m(self, start_params, reml, niter_sa, do_cg, fe_pen, cov_pen, free, full_output, method, **fit_kwargs)\u001b[0m\n\u001b[1;32m   2189\u001b[0m \u001b[38;5;66;03m# Try optimizing one or more times\u001b[39;00m\n\u001b[1;32m   2190\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m j \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(method)):\n\u001b[0;32m-> 2191\u001b[0m     rslt \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mMixedLM\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstart_params\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpacked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2192\u001b[0m \u001b[43m                                    \u001b[49m\u001b[43mskip_hessian\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   2193\u001b[0m \u001b[43m                                    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m[\u001b[49m\u001b[43mj\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2194\u001b[0m \u001b[43m                                    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mfit_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2195\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m rslt\u001b[38;5;241m.\u001b[39mmle_retvals[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mconverged\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n\u001b[1;32m   2196\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/model.py:563\u001b[0m, in \u001b[0;36mLikelihoodModel.fit\u001b[0;34m(self, start_params, method, maxiter, full_output, disp, fargs, callback, retall, skip_hessian, **kwargs)\u001b[0m\n\u001b[1;32m    560\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muse_t\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    562\u001b[0m optimizer \u001b[38;5;241m=\u001b[39m Optimizer()\n\u001b[0;32m--> 563\u001b[0m xopt, retvals, optim_settings \u001b[38;5;241m=\u001b[39m \u001b[43moptimizer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscore\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart_params\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    564\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mfargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    565\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mhessian\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhess\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    566\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    567\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mdisp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdisp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    568\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mmaxiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaxiter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    569\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mcallback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    570\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mretall\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretall\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    571\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mfull_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfull_output\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    572\u001b[0m \u001b[38;5;66;03m# Restore cov_type, cov_kwds and use_t\u001b[39;00m\n\u001b[1;32m    573\u001b[0m optim_settings\u001b[38;5;241m.\u001b[39mupdate(kwds)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/optimizer.py:241\u001b[0m, in \u001b[0;36mOptimizer._fit\u001b[0;34m(self, objective, gradient, start_params, fargs, kwargs, hessian, method, maxiter, full_output, disp, callback, retall)\u001b[0m\n\u001b[1;32m    238\u001b[0m     fit_funcs\u001b[38;5;241m.\u001b[39mupdate(extra_fit_funcs)\n\u001b[1;32m    240\u001b[0m func \u001b[38;5;241m=\u001b[39m fit_funcs[method]\n\u001b[0;32m--> 241\u001b[0m xopt, retvals \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobjective\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgradient\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart_params\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    242\u001b[0m \u001b[43m                     \u001b[49m\u001b[43mdisp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdisp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmaxiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaxiter\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    243\u001b[0m \u001b[43m                     \u001b[49m\u001b[43mretall\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretall\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfull_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfull_output\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m                     \u001b[49m\u001b[43mhess\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhessian\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    246\u001b[0m optim_settings \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124moptimizer\u001b[39m\u001b[38;5;124m'\u001b[39m: method, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstart_params\u001b[39m\u001b[38;5;124m'\u001b[39m: start_params,\n\u001b[1;32m    247\u001b[0m                   \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmaxiter\u001b[39m\u001b[38;5;124m'\u001b[39m: maxiter, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfull_output\u001b[39m\u001b[38;5;124m'\u001b[39m: full_output,\n\u001b[1;32m    248\u001b[0m                   \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdisp\u001b[39m\u001b[38;5;124m'\u001b[39m: disp, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfargs\u001b[39m\u001b[38;5;124m'\u001b[39m: fargs, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcallback\u001b[39m\u001b[38;5;124m'\u001b[39m: callback,\n\u001b[1;32m    249\u001b[0m                   \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mretall\u001b[39m\u001b[38;5;124m'\u001b[39m: retall, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mextra_fit_funcs\u001b[39m\u001b[38;5;124m\"\u001b[39m: extra_fit_funcs}\n\u001b[1;32m    250\u001b[0m optim_settings\u001b[38;5;241m.\u001b[39mupdate(kwargs)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/optimizer.py:529\u001b[0m, in \u001b[0;36m_fit_bfgs\u001b[0;34m(f, score, start_params, fargs, kwargs, disp, maxiter, callback, retall, full_output, hess)\u001b[0m\n\u001b[1;32m    527\u001b[0m norm \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnorm\u001b[39m\u001b[38;5;124m'\u001b[39m, np\u001b[38;5;241m.\u001b[39mInf)\n\u001b[1;32m    528\u001b[0m epsilon \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mepsilon\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;241m1.4901161193847656e-08\u001b[39m)\n\u001b[0;32m--> 529\u001b[0m retvals \u001b[38;5;241m=\u001b[39m \u001b[43moptimize\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfmin_bfgs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart_params\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscore\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    530\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mgtol\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgtol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnorm\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnorm\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mepsilon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mepsilon\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    531\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mmaxiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaxiter\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfull_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfull_output\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    532\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mdisp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdisp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mretall\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretall\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    533\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m full_output:\n\u001b[1;32m    534\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m retall:\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/optimize.py:1143\u001b[0m, in \u001b[0;36mfmin_bfgs\u001b[0;34m(f, x0, fprime, args, gtol, norm, epsilon, maxiter, full_output, disp, retall, callback)\u001b[0m\n\u001b[1;32m   1037\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1038\u001b[0m \u001b[38;5;124;03mMinimize a function using the BFGS algorithm.\u001b[39;00m\n\u001b[1;32m   1039\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1134\u001b[0m \n\u001b[1;32m   1135\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1136\u001b[0m opts \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgtol\u001b[39m\u001b[38;5;124m'\u001b[39m: gtol,\n\u001b[1;32m   1137\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnorm\u001b[39m\u001b[38;5;124m'\u001b[39m: norm,\n\u001b[1;32m   1138\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124meps\u001b[39m\u001b[38;5;124m'\u001b[39m: epsilon,\n\u001b[1;32m   1139\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdisp\u001b[39m\u001b[38;5;124m'\u001b[39m: disp,\n\u001b[1;32m   1140\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmaxiter\u001b[39m\u001b[38;5;124m'\u001b[39m: maxiter,\n\u001b[1;32m   1141\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreturn_all\u001b[39m\u001b[38;5;124m'\u001b[39m: retall}\n\u001b[0;32m-> 1143\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43m_minimize_bfgs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mx0\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfprime\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mopts\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1145\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m full_output:\n\u001b[1;32m   1146\u001b[0m     retlist \u001b[38;5;241m=\u001b[39m (res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mx\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfun\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mjac\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhess_inv\u001b[39m\u001b[38;5;124m'\u001b[39m],\n\u001b[1;32m   1147\u001b[0m                res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnfev\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnjev\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstatus\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/optimize.py:1235\u001b[0m, in \u001b[0;36m_minimize_bfgs\u001b[0;34m(fun, x0, args, jac, callback, gtol, norm, eps, maxiter, disp, return_all, finite_diff_rel_step, **unknown_options)\u001b[0m\n\u001b[1;32m   1232\u001b[0m pk \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m-\u001b[39mnp\u001b[38;5;241m.\u001b[39mdot(Hk, gfk)\n\u001b[1;32m   1233\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1234\u001b[0m     alpha_k, fc, gc, old_fval, old_old_fval, gfkp1 \u001b[38;5;241m=\u001b[39m \\\n\u001b[0;32m-> 1235\u001b[0m              \u001b[43m_line_search_wolfe12\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmyfprime\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mxk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgfk\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1236\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43mold_fval\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mold_old_fval\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mamin\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1e-100\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mamax\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1e100\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1237\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m _LineSearchError:\n\u001b[1;32m   1238\u001b[0m     \u001b[38;5;66;03m# Line search failed to find a better solution.\u001b[39;00m\n\u001b[1;32m   1239\u001b[0m     warnflag \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m2\u001b[39m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/optimize.py:1005\u001b[0m, in \u001b[0;36m_line_search_wolfe12\u001b[0;34m(f, fprime, xk, pk, gfk, old_fval, old_old_fval, **kwargs)\u001b[0m\n\u001b[1;32m    991\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    992\u001b[0m \u001b[38;5;124;03mSame as line_search_wolfe1, but fall back to line_search_wolfe2 if\u001b[39;00m\n\u001b[1;32m    993\u001b[0m \u001b[38;5;124;03msuitable step length is not found, and raise an exception if a\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1000\u001b[0m \n\u001b[1;32m   1001\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1003\u001b[0m extra_condition \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mpop(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mextra_condition\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m-> 1005\u001b[0m ret \u001b[38;5;241m=\u001b[39m \u001b[43mline_search_wolfe1\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfprime\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mxk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgfk\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1006\u001b[0m \u001b[43m                         \u001b[49m\u001b[43mold_fval\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mold_old_fval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1007\u001b[0m \u001b[43m                         \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1009\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ret[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m extra_condition \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1010\u001b[0m     xp1 \u001b[38;5;241m=\u001b[39m xk \u001b[38;5;241m+\u001b[39m ret[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m*\u001b[39m pk\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/linesearch.py:96\u001b[0m, in \u001b[0;36mline_search_wolfe1\u001b[0;34m(f, fprime, xk, pk, gfk, old_fval, old_old_fval, args, c1, c2, amax, amin, xtol)\u001b[0m\n\u001b[1;32m     92\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m np\u001b[38;5;241m.\u001b[39mdot(gval[\u001b[38;5;241m0\u001b[39m], pk)\n\u001b[1;32m     94\u001b[0m derphi0 \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mdot(gfk, pk)\n\u001b[0;32m---> 96\u001b[0m stp, fval, old_fval \u001b[38;5;241m=\u001b[39m \u001b[43mscalar_search_wolfe1\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     97\u001b[0m \u001b[43m        \u001b[49m\u001b[43mphi\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mderphi\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mold_fval\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mold_old_fval\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mderphi0\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     98\u001b[0m \u001b[43m        \u001b[49m\u001b[43mc1\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mc1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mc2\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mc2\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mamax\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mamax\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mamin\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mamin\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mxtol\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mxtol\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    100\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m stp, fc[\u001b[38;5;241m0\u001b[39m], gc[\u001b[38;5;241m0\u001b[39m], fval, old_fval, gval[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/linesearch.py:172\u001b[0m, in \u001b[0;36mscalar_search_wolfe1\u001b[0;34m(phi, derphi, phi0, old_phi0, derphi0, c1, c2, amax, amin, xtol)\u001b[0m\n\u001b[1;32m    170\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m task[:\u001b[38;5;241m2\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mFG\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m    171\u001b[0m     alpha1 \u001b[38;5;241m=\u001b[39m stp\n\u001b[0;32m--> 172\u001b[0m     phi1 \u001b[38;5;241m=\u001b[39m \u001b[43mphi\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    173\u001b[0m     derphi1 \u001b[38;5;241m=\u001b[39m derphi(stp)\n\u001b[1;32m    174\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/linesearch.py:84\u001b[0m, in \u001b[0;36mline_search_wolfe1.<locals>.phi\u001b[0;34m(s)\u001b[0m\n\u001b[1;32m     82\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mphi\u001b[39m(s):\n\u001b[1;32m     83\u001b[0m     fc[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m---> 84\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[43mxk\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43ms\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mpk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/_differentiable_functions.py:249\u001b[0m, in \u001b[0;36mScalarFunction.fun\u001b[0;34m(self, x)\u001b[0m\n\u001b[1;32m    247\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m np\u001b[38;5;241m.\u001b[39marray_equal(x, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mx):\n\u001b[1;32m    248\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_update_x_impl(x)\n\u001b[0;32m--> 249\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_update_fun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    250\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/_differentiable_functions.py:233\u001b[0m, in \u001b[0;36mScalarFunction._update_fun\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    231\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_update_fun\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    232\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf_updated:\n\u001b[0;32m--> 233\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_update_fun_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    234\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf_updated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/_differentiable_functions.py:137\u001b[0m, in \u001b[0;36mScalarFunction.__init__.<locals>.update_fun\u001b[0;34m()\u001b[0m\n\u001b[1;32m    136\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mupdate_fun\u001b[39m():\n\u001b[0;32m--> 137\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf \u001b[38;5;241m=\u001b[39m \u001b[43mfun_wrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/_differentiable_functions.py:134\u001b[0m, in \u001b[0;36mScalarFunction.__init__.<locals>.fun_wrapped\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m    130\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnfev \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    131\u001b[0m \u001b[38;5;66;03m# Send a copy because the user may overwrite it.\u001b[39;00m\n\u001b[1;32m    132\u001b[0m \u001b[38;5;66;03m# Overwriting results in undefined behaviour because\u001b[39;00m\n\u001b[1;32m    133\u001b[0m \u001b[38;5;66;03m# fun(self.x) will change self.x, with the two no longer linked.\u001b[39;00m\n\u001b[0;32m--> 134\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/model.py:531\u001b[0m, in \u001b[0;36mLikelihoodModel.fit.<locals>.f\u001b[0;34m(params, *args)\u001b[0m\n\u001b[1;32m    530\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m(params, \u001b[38;5;241m*\u001b[39margs):\n\u001b[0;32m--> 531\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloglike\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;241m/\u001b[39m nobs\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/regression/mixed_linear_model.py:1498\u001b[0m, in \u001b[0;36mMixedLM.loglike\u001b[0;34m(self, params, profile_fe)\u001b[0m\n\u001b[1;32m   1496\u001b[0m \u001b[38;5;66;03m# Move to the profile set\u001b[39;00m\n\u001b[1;32m   1497\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m profile_fe:\n\u001b[0;32m-> 1498\u001b[0m     fe_params, sing \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_fe_params\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcov_re\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvcomp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1499\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m sing:\n\u001b[1;32m   1500\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cov_sing \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/regression/mixed_linear_model.py:1345\u001b[0m, in \u001b[0;36mMixedLM.get_fe_params\u001b[0;34m(self, cov_re, vcomp, tol)\u001b[0m\n\u001b[1;32m   1343\u001b[0m     fe_params \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mdot(np\u001b[38;5;241m.\u001b[39mlinalg\u001b[38;5;241m.\u001b[39mpinv(xtxy[:, \u001b[38;5;241m0\u001b[39m:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]), xtxy[:, \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m])\n\u001b[1;32m   1344\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1345\u001b[0m     fe_params \u001b[38;5;241m=\u001b[39m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlinalg\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msolve\u001b[49m\u001b[43m(\u001b[49m\u001b[43mxtxy\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m:\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mxtxy\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1347\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m fe_params, sing\n",
      "File \u001b[0;32m<__array_function__ internals>:180\u001b[0m, in \u001b[0;36msolve\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/numpy/linalg/linalg.py:393\u001b[0m, in \u001b[0;36msolve\u001b[0;34m(a, b)\u001b[0m\n\u001b[1;32m    391\u001b[0m signature \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDD->D\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m isComplexType(t) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdd->d\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    392\u001b[0m extobj \u001b[38;5;241m=\u001b[39m get_linalg_error_extobj(_raise_linalgerror_singular)\n\u001b[0;32m--> 393\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43mgufunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mb\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msignature\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msignature\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mextobj\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextobj\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    395\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m wrap(r\u001b[38;5;241m.\u001b[39mastype(result_t, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m))\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/numpy/linalg/linalg.py:88\u001b[0m, in \u001b[0;36m_raise_linalgerror_singular\u001b[0;34m(err, flag)\u001b[0m\n\u001b[1;32m     87\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_raise_linalgerror_singular\u001b[39m(err, flag):\n\u001b[0;32m---> 88\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m LinAlgError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSingular matrix\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mLinAlgError\u001b[0m: Singular matrix"
     ]
    }
   ],
   "source": [
    "df_combined = df_combined.dropna(subset=all_columns)\n",
    "# Fit the GLMM\n",
    "model = smf.mixedlm(\n",
    "    formula=formula,\n",
    "    data=df_combined,\n",
    "    groups=df_combined[\"pwCF_ID\"],  # Grouping by pwCF_ID for random effects\n",
    ")\n",
    "#family=sm.families.Binomial()  # Logistic link function\n",
    "result = model.fit()\n",
    "\n",
    "# Print the model summary\n",
    "print(result.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5abbc39-d02a-473d-84a7-96a939b200d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "coefficients = pd.DataFrame({\n",
    "    \"Variable\": result.params.index,\n",
    "    \"Estimate\": result.params.values,\n",
    "    \"Std Error\": result.bse.values,\n",
    "    \"P-value\": result.pvalues.values\n",
    "})\n",
    "coefficients.to_csv(os.path.join('lmm', f'{taxa}_model_coefficients.tsv'), sep=\"\\t\", index=False)\n",
    "\n",
    "with open(os.path.join('lmm', f'{taxa}_model_results.tsv'), 'w') as f:\n",
    "    f.write(result.summary().as_text())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9ec84004-ba69-4f55-af46-32e707cce263",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MGI',\n",
       " 'pwCF_ID',\n",
       " 'IP_vs_OP',\n",
       " 'Hospital',\n",
       " 'Age',\n",
       " 'Gender',\n",
       " 'Sample_Type',\n",
       " 'H2_Uncorrected',\n",
       " 'CH4_Uncorrected',\n",
       " 'CO2',\n",
       " 'CH4_H2_ratio_corrected',\n",
       " 'Corr_',\n",
       " 'NTM',\n",
       " 'IgE',\n",
       " 'Spec_IgE',\n",
       " 'Precipitins',\n",
       " 'FVC',\n",
       " 'FEV1_best_FEV1',\n",
       " 'FEV1_RATIO_SCORE',\n",
       " 'FEV1_Obstruction_Rank',\n",
       " 'Cystic_Fibrosis_related_diabetes_CFRD_',\n",
       " 'Pancreatic_insufficiency_PI_',\n",
       " 'CF_gene_1',\n",
       " 'CFLD',\n",
       " 'CS_mucoid',\n",
       " 'CS_non_mucoid',\n",
       " 'CS_Pseudomonas_aeruginosa',\n",
       " 'CS_Oral_flora',\n",
       " 'CS_Stenophotomonas_maltophilia',\n",
       " 'CS_Aspergillus_fumigatus',\n",
       " 'CS_Aspergillus_flavus',\n",
       " 'CS_Candida_albicans',\n",
       " 'CS_Mycobacteroides_abscessus',\n",
       " 'CS_Mycobacterium_intracellulare',\n",
       " 'CS_Staphylococcus_aureus',\n",
       " 'CS_Inquilinus_limosus',\n",
       " 'CS_Achromobacter_xylosoxidans',\n",
       " 'CS_Burkholderia_cepacia',\n",
       " 'CS_NTM__Smear_negative_',\n",
       " 'CS_Mycolicibacter_terrae',\n",
       " 'CS_Aspergillus_nidulans',\n",
       " 'CS_MAC__Smear_negative_',\n",
       " 'CS_Penicillium',\n",
       " 'CS_Aspergillus_niger',\n",
       " 'CS_Scedosporium_apiospermum',\n",
       " 'CS_Lomentospora_prolificans',\n",
       " 'CS_Acremonium_species',\n",
       " 'CS_MDR_Pseudomonas_aeruginosa',\n",
       " 'CS_Haemophilus_influenzae',\n",
       " 'CS_Scedosporium_apiospermum_1',\n",
       " 'P12M_0_Nothing',\n",
       " 'P12M_1_Staphylococcus_aureus',\n",
       " 'P12M_2_Pseudomonas_aeruginosa',\n",
       " 'P12M_3_Stenotrophomonas_maltophilia',\n",
       " 'P12M_4_Rapid_NTM',\n",
       " 'P12M_5_Slow_NTM',\n",
       " 'P12M_6_Aspergillus',\n",
       " 'P12M_7_Burkholderia_cepacia',\n",
       " 'P12M_8_Haemophilus_influenza',\n",
       " 'P12M_9_Streptococcus_pneumoniae',\n",
       " 'P12M_10_Achromobacter_xylosoxidans',\n",
       " 'P12M_11_Serratia_marcesens',\n",
       " 'P12M_12_Inquilinus_limosus',\n",
       " 'P12M_13_Scedosporum_apiospermum',\n",
       " 'P12M_14_Moraxella_catarrhalis',\n",
       " 'Positive_for_PA_within_next_12_months',\n",
       " 'N12M_Pseudomonas_aeruginosa',\n",
       " 'N12M_Pseudomonas_aeruginosa_multi_resistant_',\n",
       " 'N12M_mucoid',\n",
       " 'N12M_non_mucoid',\n",
       " 'N12M_MAC',\n",
       " 'N12M_M_intracellulare',\n",
       " 'N12M_Candida_albicans',\n",
       " 'N12M_Aspergillus_fumigatus',\n",
       " 'N12M_Stenotrophomonas_maltophilia',\n",
       " 'N12M_Staphylococcus_aureus',\n",
       " 'N12M_M_gordonae',\n",
       " 'N12M_M_chelonae',\n",
       " 'O_Penicillium_species',\n",
       " 'O_Enterobacter_cloacae',\n",
       " 'O_Aspergillus_Flavus',\n",
       " 'O_Mycobacterium_abscessus',\n",
       " 'O_Cupriavidus_gilardii',\n",
       " 'O_Mycobacterium_intracellular',\n",
       " 'O_Scedosporium_apiospermum',\n",
       " 'O_Pseudomonas_aeruginosa_multi_resistant_',\n",
       " 'O_Streptococcus_pyogenes',\n",
       " 'O_Pseudomonas_aeruginosa_non_mucoid_',\n",
       " 'O_Chryseobacterium_indologenes',\n",
       " 'O_Bordetella_species',\n",
       " 'O_Klebsiella_oxytoca',\n",
       " 'O_Mycobacterium_mucogenicum',\n",
       " 'O_Mycobacterium_scotochromogen',\n",
       " 'O_Acinetobacter_species',\n",
       " 'O_Enterobacter_species',\n",
       " 'O_Aspergillus_niger',\n",
       " 'O_Aspergillus_terreus',\n",
       " 'O_Pseudomonas_aeruginosa_mucoid_',\n",
       " 'O_Proteus_mirabilis',\n",
       " 'O_talaromyces_purpurogenus',\n",
       " 'O_Trichosporon_mycotoxinivorans',\n",
       " 'Antibiotics_YN',\n",
       " 'IV',\n",
       " 'INH',\n",
       " 'PO',\n",
       " 'two_Amphotericin_B_Ambisome__INH',\n",
       " 'four_Amikacin_IV',\n",
       " 'two_Amikacin_INH',\n",
       " 'two_Amoxicillin_Potassium_clavulanate_Aug_Duo__PO',\n",
       " 'three_Azithromycin_IV',\n",
       " 'two_Azithromycin_PO',\n",
       " 'three_Aztreonam_IV',\n",
       " 'three_Cefopime_IV',\n",
       " 'four_Cefoxitin_IV',\n",
       " 'three_Ceftazidime_IV',\n",
       " 'two_Ceftazidime_INH',\n",
       " 'one_Cephalexin_PO',\n",
       " 'two_Ciprofloxacin_PO',\n",
       " 'two_Clarithromycin_PO',\n",
       " 'two_Clofazimine_PO',\n",
       " 'four_Colistin_IV',\n",
       " 'two_Colistin_IHN',\n",
       " 'one_Flucloaxcillin_PO',\n",
       " 'three_Imipenem',\n",
       " 'one_Itraconazole_Lozenoc__PO',\n",
       " 'three_Ivacaftor_Kalydeco_',\n",
       " 'three_Meropenem_IV',\n",
       " 'three_Methylpredinosolone_IV',\n",
       " 'three_Omalizumab_SC',\n",
       " 'three_piperacillin_sodium_tazobactam_sodium_Tazocin__IV',\n",
       " 'two_prednisolone_PO',\n",
       " 'one_Sulfamethoxazole_trimethoprim_Bactrim__PO',\n",
       " 'three_tobramycin_IV',\n",
       " 'two_tobramycin_INH',\n",
       " 'Sum_of_antibiotics',\n",
       " 'Sum_of_antifungals',\n",
       " 'Sum_of_steroid_mabs',\n",
       " 'DNA_extraction__conc',\n",
       " 'DNA_Conc_ng_ul_',\n",
       " 'Mean_Size_BP',\n",
       " 'Total_Clusters_Passing_Filter_Million_']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(encoded_metadata.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82853325-1d92-4615-8d4b-037bdd736a41",
   "metadata": {},
   "source": [
    "# We can't do all\n",
    "\n",
    "We get the 'maximum recursion depth exceeded' error.\n",
    "\n",
    "So lets subsample our predictors and then combine them!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "758680f6-0d39-4302-b106-d2a08fe53925",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Iteration 0\n",
      "Iteration 1000\n",
      "Iteration 2000\n",
      "Iteration 3000\n",
      "Iteration 4000\n",
      "Iteration 5000\n",
      "Iteration 6000\n",
      "Iteration 7000\n",
      "Iteration 8000\n",
      "Iteration 9000\n",
      "Syntax warnings\n",
      "{'N12M_Pseudomonas_aeruginosa', '~', '+', 'CS_Pseudomonas_aeruginosa'}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                           Predictor  Mean Estimate  \\\n",
      "0                            ABC_transporter_YeiABEF       0.000533   \n",
      "1                             ABC_transporter_YxeMNO       0.012208   \n",
      "2             ABC_transporter_of_unknown_substrate_X      -0.000249   \n",
      "3            ABC_transporter_tungstate_TC_3_A_1_6_2_      -0.008973   \n",
      "4                     ABC_type_iron_transport_system       0.000091   \n",
      "...                                              ...            ...   \n",
      "4306                         tRNA_aminoacylation_Val       0.000022   \n",
      "4307  tRNA_modification_5_oxyacetyl_uridine_cmo5U34_      -0.001655   \n",
      "4308                     tRNA_nucleotidyltransferase       0.000452   \n",
      "4309                                   tRNA_splicing       0.000088   \n",
      "4310                                 tRNA_thiolation      -0.000320   \n",
      "\n",
      "      Mean Std Err  Mean P-value  \n",
      "0     4.506153e+09      0.490405  \n",
      "1     1.705472e-02      0.463453  \n",
      "2     1.059203e-03      0.528974  \n",
      "3     1.372700e-01      0.524505  \n",
      "4     6.433104e-03      0.512489  \n",
      "...            ...           ...  \n",
      "4306  3.543736e-04      0.555696  \n",
      "4307  8.987979e-03      0.524013  \n",
      "4308  8.297940e-04      0.470000  \n",
      "4309  6.852919e-04      0.529497  \n",
      "4310  1.116621e-03      0.481735  \n",
      "\n",
      "[4311 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "all_predictors = df.columns \n",
    "num_predictors_per_model = 100\n",
    "\n",
    "results = []\n",
    "\n",
    "num_iterations = 10000\n",
    "\n",
    "syntax_warnings = set()\n",
    "\n",
    "for i in range(num_iterations):\n",
    "    if i % 1000 == 0:\n",
    "        print(f\"Iteration {i}\", file=sys.stderr)\n",
    "    subset_predictors = random.sample(list(all_predictors), num_predictors_per_model) # Randomly select predictors\n",
    "    formula = f\"N12M_Pseudomonas_aeruginosa ~ {' + '.join(subset_predictors)} + CS_Pseudomonas_aeruginosa\"\n",
    "    df_combined_na = df_combined.dropna(subset=subset_predictors)\n",
    "    \n",
    "    try:\n",
    "        with warnings.catch_warnings():\n",
    "            # this just suppresses some statsmodels warnings.\n",
    "            warnings.filterwarnings(\"ignore\", module=\"statsmodels\")\n",
    "            model = smf.mixedlm(\n",
    "                formula=formula,\n",
    "                data=df_combined_na,\n",
    "                groups=df_combined_na[\"pwCF_ID\"]\n",
    "            )\n",
    "            result = model.fit()\n",
    "        \n",
    "        df_result = pd.DataFrame({\n",
    "            \"Predictor\": result.params.index,\n",
    "            \"Estimate\": result.params.values,\n",
    "            \"Std Err\": result.bse.values,\n",
    "            \"P-value\": result.pvalues.values\n",
    "        })\n",
    "        df_result[\"Iteration\"] = i  # Add iteration number for tracking\n",
    "        \n",
    "        # Append to results\n",
    "        results.append(df_result)\n",
    "    except Exception as e:\n",
    "        if isinstance(e, SyntaxError):\n",
    "            warn_set = set(formula.split())\n",
    "            if not syntax_warnings:\n",
    "                syntax_warnings = warn_set\n",
    "            else:\n",
    "                syntax_warnings = syntax_warnings.intersection(warn_set)\n",
    "        \n",
    "        # print(f\"Iteration {i} failed: {e} > {type(e)}\", file=sys.stderr)\n",
    "        # if isinstance(e, SyntaxError):\n",
    "        #     print(f\"\\tformula: {formula}\", file=sys.stderr)\n",
    "\n",
    "print(f\"Syntax warnings\\n{syntax_warnings}\", file=sys.stderr)\n",
    "        \n",
    "# Combine results into a single DataFrame\n",
    "all_results = pd.concat(results)\n",
    "\n",
    "# Group by predictor and aggregate results\n",
    "final_results = (\n",
    "    all_results.groupby(\"Predictor\")[[\"Estimate\", \"Std Err\", \"P-value\"]]\n",
    "    .mean()\n",
    "    .rename(columns={\"Estimate\": \"Mean Estimate\", \"Std Err\": \"Mean Std Err\", \"P-value\": \"Mean P-value\"})\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "print(final_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "0d4208ff-d83d-4d30-924f-052d59f91726",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Predictor</th>\n",
       "      <th>Mean Estimate</th>\n",
       "      <th>Mean Std Err</th>\n",
       "      <th>Mean P-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1391</th>\n",
       "      <td>Ferrithrix</td>\n",
       "      <td>-5.621701e-02</td>\n",
       "      <td>3.395498e-02</td>\n",
       "      <td>0.137927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>664</th>\n",
       "      <td>CS_Pseudomonas_aeruginosa</td>\n",
       "      <td>4.515746e-01</td>\n",
       "      <td>9.320250e+10</td>\n",
       "      <td>0.151082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>852</th>\n",
       "      <td>Clavichlamydia</td>\n",
       "      <td>2.358970e+09</td>\n",
       "      <td>2.236434e+12</td>\n",
       "      <td>0.195146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2562</th>\n",
       "      <td>Neptunochlamydia</td>\n",
       "      <td>-8.927003e+02</td>\n",
       "      <td>2.115908e+12</td>\n",
       "      <td>0.215177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3671</th>\n",
       "      <td>Sphingomicrobium</td>\n",
       "      <td>-1.042638e+09</td>\n",
       "      <td>2.433856e+03</td>\n",
       "      <td>0.233021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3657</th>\n",
       "      <td>Sorangium</td>\n",
       "      <td>1.779438e-03</td>\n",
       "      <td>2.418180e-02</td>\n",
       "      <td>0.620888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>Cuspidothrix</td>\n",
       "      <td>-1.886660e-04</td>\n",
       "      <td>3.757258e-03</td>\n",
       "      <td>0.622521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2067</th>\n",
       "      <td>Lignipirellula</td>\n",
       "      <td>1.215341e-02</td>\n",
       "      <td>2.742302e+00</td>\n",
       "      <td>0.630413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2757</th>\n",
       "      <td>Paenihalocynthiibacter</td>\n",
       "      <td>1.290325e+28</td>\n",
       "      <td>1.016200e+27</td>\n",
       "      <td>0.630682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1641</th>\n",
       "      <td>Group Var</td>\n",
       "      <td>2.433753e+02</td>\n",
       "      <td>1.753329e+03</td>\n",
       "      <td>0.744505</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4311 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Predictor  Mean Estimate  Mean Std Err  Mean P-value\n",
       "1391                 Ferrithrix  -5.621701e-02  3.395498e-02      0.137927\n",
       "664   CS_Pseudomonas_aeruginosa   4.515746e-01  9.320250e+10      0.151082\n",
       "852              Clavichlamydia   2.358970e+09  2.236434e+12      0.195146\n",
       "2562           Neptunochlamydia  -8.927003e+02  2.115908e+12      0.215177\n",
       "3671           Sphingomicrobium  -1.042638e+09  2.433856e+03      0.233021\n",
       "...                         ...            ...           ...           ...\n",
       "3657                  Sorangium   1.779438e-03  2.418180e-02      0.620888\n",
       "972                Cuspidothrix  -1.886660e-04  3.757258e-03      0.622521\n",
       "2067             Lignipirellula   1.215341e-02  2.742302e+00      0.630413\n",
       "2757     Paenihalocynthiibacter   1.290325e+28  1.016200e+27      0.630682\n",
       "1641                  Group Var   2.433753e+02  1.753329e+03      0.744505\n",
       "\n",
       "[4311 rows x 4 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results.sort_values(by='Mean P-value')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "f508da75-f835-4de0-9bd5-ff6917c23d78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Predictor</th>\n",
       "      <th>Mean Estimate</th>\n",
       "      <th>Mean Std Err</th>\n",
       "      <th>Mean P-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Predictor, Mean Estimate, Mean Std Err, Mean P-value]\n",
       "Index: []"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_results[final_results['Mean P-value'] < 0.1].sort_values(by='Mean P-value')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "6e1a7f99-5b47-4ea0-b68e-00a6f757eec5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Estimate</th>\n",
       "      <th>Std Err</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Iteration</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Predictor</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Phycocomes</th>\n",
       "      <td>109</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Roseibaca</th>\n",
       "      <td>119</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pokkaliibacter</th>\n",
       "      <td>122</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gloeomargarita</th>\n",
       "      <td>122</td>\n",
       "      <td>107</td>\n",
       "      <td>107</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Oceaniovalibus</th>\n",
       "      <td>122</td>\n",
       "      <td>108</td>\n",
       "      <td>108</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thermicanus</th>\n",
       "      <td>210</td>\n",
       "      <td>209</td>\n",
       "      <td>209</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_sulfur_modification_cluster</th>\n",
       "      <td>222</td>\n",
       "      <td>221</td>\n",
       "      <td>221</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Group Var</th>\n",
       "      <td>7218</td>\n",
       "      <td>6669</td>\n",
       "      <td>6669</td>\n",
       "      <td>7218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intercept</th>\n",
       "      <td>7218</td>\n",
       "      <td>7174</td>\n",
       "      <td>7174</td>\n",
       "      <td>7218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CS_Pseudomonas_aeruginosa</th>\n",
       "      <td>7218</td>\n",
       "      <td>7189</td>\n",
       "      <td>7189</td>\n",
       "      <td>7218</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4311 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Estimate  Std Err  P-value  Iteration\n",
       "Predictor                                                             \n",
       "Phycocomes                            109       93       93        109\n",
       "Roseibaca                             119      100      100        119\n",
       "Pokkaliibacter                        122       98       98        122\n",
       "Gloeomargarita                        122      107      107        122\n",
       "Oceaniovalibus                        122      108      108        122\n",
       "...                                   ...      ...      ...        ...\n",
       "Thermicanus                           210      209      209        210\n",
       "DNA_sulfur_modification_cluster       222      221      221        222\n",
       "Group Var                            7218     6669     6669       7218\n",
       "Intercept                            7218     7174     7174       7218\n",
       "CS_Pseudomonas_aeruginosa            7218     7189     7189       7218\n",
       "\n",
       "[4311 rows x 4 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_results.groupby(\"Predictor\").count().sort_values(by='Estimate')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "93698ad3-1be4-45dd-b099-9a8e82a1d58b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Predictor</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>Std Err</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Iteration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Intercept</td>\n",
       "      <td>1.686749</td>\n",
       "      <td>1.848951</td>\n",
       "      <td>0.361625</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zobellella</td>\n",
       "      <td>0.062825</td>\n",
       "      <td>0.061080</td>\n",
       "      <td>0.303685</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Swingsia</td>\n",
       "      <td>-0.363966</td>\n",
       "      <td>1.156607</td>\n",
       "      <td>0.753002</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mesoaciditoga</td>\n",
       "      <td>-0.254529</td>\n",
       "      <td>0.146950</td>\n",
       "      <td>0.083259</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mariniblastus</td>\n",
       "      <td>0.030389</td>\n",
       "      <td>0.034348</td>\n",
       "      <td>0.376308</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Cysteiniphilum</td>\n",
       "      <td>0.492660</td>\n",
       "      <td>0.373764</td>\n",
       "      <td>0.187468</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Leucobacter</td>\n",
       "      <td>0.002880</td>\n",
       "      <td>0.003073</td>\n",
       "      <td>0.348561</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Sphingobium</td>\n",
       "      <td>0.002870</td>\n",
       "      <td>0.010046</td>\n",
       "      <td>0.775100</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>CS_Pseudomonas_aeruginosa</td>\n",
       "      <td>0.891576</td>\n",
       "      <td>0.287592</td>\n",
       "      <td>0.001934</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Group Var</td>\n",
       "      <td>30.575261</td>\n",
       "      <td>261.104545</td>\n",
       "      <td>0.906781</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Predictor   Estimate     Std Err   P-value  Iteration\n",
       "0                    Intercept   1.686749    1.848951  0.361625       9998\n",
       "1                   Zobellella   0.062825    0.061080  0.303685       9998\n",
       "2                     Swingsia  -0.363966    1.156607  0.753002       9998\n",
       "3                Mesoaciditoga  -0.254529    0.146950  0.083259       9998\n",
       "4                Mariniblastus   0.030389    0.034348  0.376308       9998\n",
       "..                         ...        ...         ...       ...        ...\n",
       "98              Cysteiniphilum   0.492660    0.373764  0.187468       9998\n",
       "99                 Leucobacter   0.002880    0.003073  0.348561       9998\n",
       "100                Sphingobium   0.002870    0.010046  0.775100       9998\n",
       "101  CS_Pseudomonas_aeruginosa   0.891576    0.287592  0.001934       9998\n",
       "102                  Group Var  30.575261  261.104545  0.906781       9998\n",
       "\n",
       "[103 rows x 5 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "887155cd-126f-4eec-9c8b-efde510e4db0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(743454, 5)"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_results.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "0ba4eb78-6ea1-4c56-abb0-beaebf844653",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Predictor</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>Std Err</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Iteration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zobellella</td>\n",
       "      <td>0.062825</td>\n",
       "      <td>0.061080</td>\n",
       "      <td>0.303685</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Swingsia</td>\n",
       "      <td>-0.363966</td>\n",
       "      <td>1.156607</td>\n",
       "      <td>0.753002</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mesoaciditoga</td>\n",
       "      <td>-0.254529</td>\n",
       "      <td>0.146950</td>\n",
       "      <td>0.083259</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mariniblastus</td>\n",
       "      <td>0.030389</td>\n",
       "      <td>0.034348</td>\n",
       "      <td>0.376308</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Extracellular_matrix_proteins_PSL_involved_in_...</td>\n",
       "      <td>-0.000538</td>\n",
       "      <td>0.002094</td>\n",
       "      <td>0.797165</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Gilliamella</td>\n",
       "      <td>0.000335</td>\n",
       "      <td>0.010068</td>\n",
       "      <td>0.973478</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Cysteiniphilum</td>\n",
       "      <td>0.492660</td>\n",
       "      <td>0.373764</td>\n",
       "      <td>0.187468</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Leucobacter</td>\n",
       "      <td>0.002880</td>\n",
       "      <td>0.003073</td>\n",
       "      <td>0.348561</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Sphingobium</td>\n",
       "      <td>0.002870</td>\n",
       "      <td>0.010046</td>\n",
       "      <td>0.775100</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>CS_Pseudomonas_aeruginosa</td>\n",
       "      <td>0.891576</td>\n",
       "      <td>0.287592</td>\n",
       "      <td>0.001934</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Predictor  Estimate   Std Err  \\\n",
       "1                                           Zobellella  0.062825  0.061080   \n",
       "2                                             Swingsia -0.363966  1.156607   \n",
       "3                                        Mesoaciditoga -0.254529  0.146950   \n",
       "4                                        Mariniblastus  0.030389  0.034348   \n",
       "5    Extracellular_matrix_proteins_PSL_involved_in_... -0.000538  0.002094   \n",
       "..                                                 ...       ...       ...   \n",
       "97                                         Gilliamella  0.000335  0.010068   \n",
       "98                                      Cysteiniphilum  0.492660  0.373764   \n",
       "99                                         Leucobacter  0.002880  0.003073   \n",
       "100                                        Sphingobium  0.002870  0.010046   \n",
       "101                          CS_Pseudomonas_aeruginosa  0.891576  0.287592   \n",
       "\n",
       "      P-value  Iteration  \n",
       "1    0.303685       9998  \n",
       "2    0.753002       9998  \n",
       "3    0.083259       9998  \n",
       "4    0.376308       9998  \n",
       "5    0.797165       9998  \n",
       "..        ...        ...  \n",
       "97   0.973478       9998  \n",
       "98   0.187468       9998  \n",
       "99   0.348561       9998  \n",
       "100  0.775100       9998  \n",
       "101  0.001934       9998  \n",
       "\n",
       "[101 rows x 5 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14559183-6f2f-4461-a2d0-6f2592af0246",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "cb1c96b0-87b6-4d48-869f-23f6b9db479d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid\n",
      "2_O_alpha_mannosyl_D_glycerate_utilization\n",
      "2_aminophenol_Metabolism\n",
      "2_ketoacid_oxidoreductases_disambiguation\n",
      "2_oxoglutarate_dehydrogenase_\n",
      "2_phosphoglycolate_salvage\n",
      "3_amino_5_hydroxybenzoic_Acid_Synthesis\n",
      "4_hydroxybenzoyl_CoA_reductase\n",
      "5_methylaminomethyl_2_thiouridine\n"
     ]
    }
   ],
   "source": [
    "for c in df.columns:\n",
    "    if c not in set(all_results['Predictor'].values):\n",
    "        print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "9512ee2c-a78f-4ff0-9403-91d80629fac3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MGI',\n",
       " 'pwCF_ID',\n",
       " 'IP_vs_OP',\n",
       " 'Hospital',\n",
       " 'Age',\n",
       " 'Gender',\n",
       " 'Sample_Type',\n",
       " 'H2_Uncorrected',\n",
       " 'CH4_Uncorrected',\n",
       " 'CO2',\n",
       " 'CH4_H2_ratio_corrected',\n",
       " 'Corr_',\n",
       " 'NTM',\n",
       " 'IgE',\n",
       " 'Spec_IgE',\n",
       " 'Precipitins',\n",
       " 'FVC',\n",
       " 'FEV1_best_FEV1',\n",
       " 'FEV1_RATIO_SCORE',\n",
       " 'FEV1_Obstruction_Rank',\n",
       " 'Cystic_Fibrosis_related_diabetes_CFRD_',\n",
       " 'Pancreatic_insufficiency_PI_',\n",
       " 'CF_gene_1',\n",
       " 'CFLD',\n",
       " 'CS_mucoid',\n",
       " 'CS_non_mucoid',\n",
       " 'CS_Pseudomonas_aeruginosa',\n",
       " 'CS_Oral_flora',\n",
       " 'CS_Stenophotomonas_maltophilia',\n",
       " 'CS_Aspergillus_fumigatus',\n",
       " 'CS_Aspergillus_flavus',\n",
       " 'CS_Candida_albicans',\n",
       " 'CS_Mycobacteroides_abscessus',\n",
       " 'CS_Mycobacterium_intracellulare',\n",
       " 'CS_Staphylococcus_aureus',\n",
       " 'CS_Inquilinus_limosus',\n",
       " 'CS_Achromobacter_xylosoxidans',\n",
       " 'CS_Burkholderia_cepacia',\n",
       " 'CS_NTM__Smear_negative_',\n",
       " 'CS_Mycolicibacter_terrae',\n",
       " 'CS_Aspergillus_nidulans',\n",
       " 'CS_MAC__Smear_negative_',\n",
       " 'CS_Penicillium',\n",
       " 'CS_Aspergillus_niger',\n",
       " 'CS_Scedosporium_apiospermum',\n",
       " 'CS_Lomentospora_prolificans',\n",
       " 'CS_Acremonium_species',\n",
       " 'CS_MDR_Pseudomonas_aeruginosa',\n",
       " 'CS_Haemophilus_influenzae',\n",
       " 'CS_Scedosporium_apiospermum_1',\n",
       " 'P12M_0_Nothing',\n",
       " 'P12M_1_Staphylococcus_aureus',\n",
       " 'P12M_2_Pseudomonas_aeruginosa',\n",
       " 'P12M_3_Stenotrophomonas_maltophilia',\n",
       " 'P12M_4_Rapid_NTM',\n",
       " 'P12M_5_Slow_NTM',\n",
       " 'P12M_6_Aspergillus',\n",
       " 'P12M_7_Burkholderia_cepacia',\n",
       " 'P12M_8_Haemophilus_influenza',\n",
       " 'P12M_9_Streptococcus_pneumoniae',\n",
       " 'P12M_10_Achromobacter_xylosoxidans',\n",
       " 'P12M_11_Serratia_marcesens',\n",
       " 'P12M_12_Inquilinus_limosus',\n",
       " 'P12M_13_Scedosporum_apiospermum',\n",
       " 'P12M_14_Moraxella_catarrhalis',\n",
       " 'Positive_for_PA_within_next_12_months',\n",
       " 'N12M_Pseudomonas_aeruginosa',\n",
       " 'N12M_Pseudomonas_aeruginosa_multi_resistant_',\n",
       " 'N12M_mucoid',\n",
       " 'N12M_non_mucoid',\n",
       " 'N12M_MAC',\n",
       " 'N12M_M_intracellulare',\n",
       " 'N12M_Candida_albicans',\n",
       " 'N12M_Aspergillus_fumigatus',\n",
       " 'N12M_Stenotrophomonas_maltophilia',\n",
       " 'N12M_Staphylococcus_aureus',\n",
       " 'N12M_M_gordonae',\n",
       " 'N12M_M_chelonae',\n",
       " 'O_Penicillium_species',\n",
       " 'O_Enterobacter_cloacae',\n",
       " 'O_Aspergillus_Flavus',\n",
       " 'O_Mycobacterium_abscessus',\n",
       " 'O_Cupriavidus_gilardii',\n",
       " 'O_Mycobacterium_intracellular',\n",
       " 'O_Scedosporium_apiospermum',\n",
       " 'O_Pseudomonas_aeruginosa_multi_resistant_',\n",
       " 'O_Streptococcus_pyogenes',\n",
       " 'O_Pseudomonas_aeruginosa_non_mucoid_',\n",
       " 'O_Chryseobacterium_indologenes',\n",
       " 'O_Bordetella_species',\n",
       " 'O_Klebsiella_oxytoca',\n",
       " 'O_Mycobacterium_mucogenicum',\n",
       " 'O_Mycobacterium_scotochromogen',\n",
       " 'O_Acinetobacter_species',\n",
       " 'O_Enterobacter_species',\n",
       " 'O_Aspergillus_niger',\n",
       " 'O_Aspergillus_terreus',\n",
       " 'O_Pseudomonas_aeruginosa_mucoid_',\n",
       " 'O_Proteus_mirabilis',\n",
       " 'O_talaromyces_purpurogenus',\n",
       " 'O_Trichosporon_mycotoxinivorans',\n",
       " 'Antibiotics_YN',\n",
       " 'IV',\n",
       " 'INH',\n",
       " 'PO',\n",
       " '2_Amphotericin_B_Ambisome__INH',\n",
       " '4_Amikacin_IV',\n",
       " '2_Amikacin_INH',\n",
       " '2_Amoxicillin_Potassium_clavulanate_Aug_Duo__PO',\n",
       " '3_Azithromycin_IV',\n",
       " '2_Azithromycin_PO',\n",
       " '3_Aztreonam_IV',\n",
       " '3_Cefopime_IV',\n",
       " '4_Cefoxitin_IV',\n",
       " '3_Ceftazidime_IV',\n",
       " '2_Ceftazidime_INH',\n",
       " '1_Cephalexin_PO',\n",
       " '2_Ciprofloxacin_PO',\n",
       " '2_Clarithromycin_PO',\n",
       " '2_Clofazimine_PO',\n",
       " '4_Colistin_IV',\n",
       " '2_Colistin_IHN',\n",
       " '1_Flucloaxcillin_PO',\n",
       " '3_Imipenem',\n",
       " '1_Itraconazole_Lozenoc__PO',\n",
       " '3_Ivacaftor_Kalydeco_',\n",
       " '3_Meropenem_IV',\n",
       " '3_Methylpredinosolone_IV',\n",
       " '3_Omalizumab_SC',\n",
       " '3_piperacillin_sodium_tazobactam_sodium_Tazocin__IV',\n",
       " '2_prednisolone_PO',\n",
       " '1_Sulfamethoxazole_trimethoprim_Bactrim__PO',\n",
       " '3_tobramycin_IV',\n",
       " '2_tobramycin_INH',\n",
       " 'Sum_of_antibiotics',\n",
       " 'Sum_of_antifungals',\n",
       " 'Sum_of_steroid_mabs',\n",
       " 'DNA_extraction__conc',\n",
       " 'DNA_Conc_ng_ul_',\n",
       " 'Mean_Size_BP',\n",
       " 'Total_Clusters_Passing_Filter_Million_']"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(encoded_metadata.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d94e4b9e-28f4-4ca2-8277-9291480410e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f87f6a19-2d73-4881-b3cd-6afecdb67ec4",
   "metadata": {},
   "source": [
    "# Early trial\n",
    "\n",
    "Ignore the stuff down here, its probably wrong :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a29357e3-b502-41c1-94c7-2bb9ca8e7b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = smf.mixedlm(\n",
    "    'N12M_Pseudomonas_aeruginosa ~ abundance + Pseudomonas_Culture + CS_Pseudomonas_aeruginosa', \n",
    "    merged_data,\n",
    "    groups=merged_data[\"pwCF_ID\"] \n",
    ")\n",
    "result = model.fit()\n",
    "\n",
    "# Print the model summary\n",
    "print(result.summary())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ca2cc9f-0782-415b-a47d-52cbe48fedaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "if False:\n",
    "    model = smf.mixedlm(\n",
    "        'abundance ~ CS_Pseudomonas_aeruginosa + taxa_functions', \n",
    "        merged_data,\n",
    "        groups=merged_data[\"pwCF_ID\"] \n",
    "    )\n",
    "    result = model.fit()\n",
    "\n",
    "    # Print the model summary\n",
    "    print(result.summary())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce94b2ee-0be5-4869-9769-08caf6a6789d",
   "metadata": {},
   "source": [
    "## Write the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f413e87e-e2d0-4b90-8bbf-cdb0509b531b",
   "metadata": {},
   "outputs": [],
   "source": [
    "coefficients = pd.DataFrame({\n",
    "    \"Variable\": result.params.index,\n",
    "    \"Estimate\": result.params.values,\n",
    "    \"Std Error\": result.bse.values,\n",
    "    \"P-value\": result.pvalues.values\n",
    "})\n",
    "coefficients.to_csv(os.path.join('lmm', f'{taxa}_model_coefficients.tsv'), sep=\"\\t\", index=False)\n",
    "\n",
    "with open(os.path.join('lmm', f'{taxa}_model_results.tsv'), 'w') as f:\n",
    "    f.write(result.summary().as_text())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1a55639-6ff9-4e37-8dd8-4bdd90e575bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "# Extract fixed effect coefficients and confidence intervals\n",
    "params = result.fe_params  # Fixed effect coefficients\n",
    "conf_int = result.conf_int()  # Confidence intervals\n",
    "conf_int.columns = ['lower', 'upper']\n",
    "\n",
    "# Combine coefficients and confidence intervals\n",
    "coefficients = pd.DataFrame({\n",
    "    'coef': params,\n",
    "    'lower': conf_int['lower'],\n",
    "    'upper': conf_int['upper']\n",
    "})\n",
    "\n",
    "# Plot the coefficients with error bars\n",
    "plt.figure(figsize=(8, 6))\n",
    "plt.errorbar(coefficients.index, coefficients['coef'], \n",
    "             yerr=(coefficients['coef'] - coefficients['lower'], coefficients['upper'] - coefficients['coef']), \n",
    "             fmt='o', capsize=5)\n",
    "plt.axhline(0, color='gray', linestyle='--', linewidth=1)\n",
    "plt.xticks(rotation=45)\n",
    "plt.title(\"Fixed Effect Coefficients with Confidence Intervals\")\n",
    "plt.ylabel(\"Coefficient Value\")\n",
    "plt.xlabel(\"Predictor\")\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4e2dacf-2866-4264-aad6-71b688c12fdc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bioinformatics]",
   "language": "python",
   "name": "conda-env-bioinformatics-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
