{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a7a213c0-97e1-4338-b943-3166b9e1f91a",
   "metadata": {},
   "source": [
    "# Linear Mixed Models\n",
    "\n",
    "Build some linear mixed models about our data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b54c76e8-0aaf-4da7-be76-3b56a655b8d2",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "780ba312-1f7d-4751-b01f-0718dc08222a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from socket import gethostname\n",
    "\n",
    "hostname = gethostname()\n",
    "\n",
    "import re\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.collections import PatchCollection\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.dates as mdates\n",
    "from matplotlib.colors import ListedColormap\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import json\n",
    "\n",
    "from itertools import cycle\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder, OrdinalEncoder\n",
    "from sklearn.inspection import permutation_importance\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier\n",
    "from sklearn.metrics import mean_squared_error\n",
    "\n",
    "from scipy.stats import linregress\n",
    "\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "\n",
    "# there is a FutureWarning in sklearn StandardScalar which is really annoying. This ignores it.\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "try:\n",
    "  import google.colab\n",
    "  IN_COLAB = True\n",
    "  !pip install adjustText\n",
    "  from google.colab import drive\n",
    "  drive.mount('/content/drive')\n",
    "  datadir = '/content/drive/MyDrive/Projects/CF/Adelaide/CF_Data_Analysis'\n",
    "except ImportError:\n",
    "  IN_COLAB = False\n",
    "  datadir = '..'\n",
    "\n",
    "\n",
    "if hostname.startswith('hpc-node'):\n",
    "    IN_DEEPTHOUGHT = True\n",
    "    sys.path.append('..')\n",
    "else:\n",
    "    IN_DEEPTHOUGHT = False\n",
    "\n",
    "import cf_analysis_lib\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d320c78c-0cb0-4bcf-af63-3bb8668872dc",
   "metadata": {},
   "source": [
    "## Read the data frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ecd087d9-65e7-4859-801f-019893047960",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The subsystems df has shape: (127, 769)\n",
      "The taxonomy df has shape: (127, 3581)\n",
      "The metadata df has shape: (127, 166)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/edwa0468/GitHubs/CF_Data_Analysis/Jupyter/../cf_analysis_lib/read_data.py:125: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.\n",
      "  metadata[c] = pd.to_datetime(metadata[c])\n"
     ]
    }
   ],
   "source": [
    "sequence_type = \"MGI\"\n",
    "datadir = '..'\n",
    "#sslevel = 'level2_norm_ss.tsv.gz'\n",
    "sslevel = 'subsystems_norm_ss.tsv.gz'\n",
    "ss_df = cf_analysis_lib.read_subsystems(os.path.join(datadir, sequence_type, \"FunctionalAnalysis\", \"subsystems\", sslevel), sequence_type)\n",
    "ss_df = ss_df.T\n",
    "print(f\"The subsystems df has shape: {ss_df.shape}\")\n",
    "\n",
    "taxa = \"genus\"\n",
    "genus_otu = cf_analysis_lib.read_taxonomy(datadir, sequence_type, taxa)\n",
    "genus_otu = genus_otu.T\n",
    "print(f\"The taxonomy df has shape: {genus_otu.shape}\")\n",
    "metadata = cf_analysis_lib.read_metadata(datadir, sequence_type, categorise=True)\n",
    "print(f\"The metadata df has shape: {metadata.shape}\")\n",
    "\n",
    "df = ss_df.merge(genus_otu, left_index=True, right_index=True, how='inner')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11120eb9-bc3a-41b9-baaf-e2eafc646602",
   "metadata": {},
   "source": [
    "## Rename the columns\n",
    "\n",
    "Statsmodels can't handle column names with spaces or special characters, and so we rename them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a58a052-34b9-4257-b44b-fe0654c6b535",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = [re.sub(r'\\W+', '_', col) for col in df.columns]\n",
    "metadata.columns = [re.sub(r'\\W+', '_', col) for col in metadata.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4c2d8609-5ccc-421b-bed5-032b9bf66c9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid</th>\n",
       "      <th>2_O_alpha_mannosyl_D_glycerate_utilization</th>\n",
       "      <th>2_aminophenol_Metabolism</th>\n",
       "      <th>2_ketoacid_oxidoreductases_disambiguation</th>\n",
       "      <th>2_oxoglutarate_dehydrogenase_</th>\n",
       "      <th>2_phosphoglycolate_salvage</th>\n",
       "      <th>3_amino_5_hydroxybenzoic_Acid_Synthesis</th>\n",
       "      <th>4_hydroxybenzoyl_CoA_reductase</th>\n",
       "      <th>5_methylaminomethyl_2_thiouridine</th>\n",
       "      <th>A_Hypothetical_Protein_Related_to_Proline_Metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Oceaniferula</th>\n",
       "      <th>Persicirhabdus</th>\n",
       "      <th>Phragmitibacter</th>\n",
       "      <th>Prosthecobacter</th>\n",
       "      <th>Roseibacillus</th>\n",
       "      <th>Roseimicrobium</th>\n",
       "      <th>Sulfuriroseicoccus</th>\n",
       "      <th>Verrucomicrobium</th>\n",
       "      <th>Eremiobacter</th>\n",
       "      <th>Methylomirabilis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1068841_20180306_S</th>\n",
       "      <td>10.085904</td>\n",
       "      <td>2784.895948</td>\n",
       "      <td>516.160945</td>\n",
       "      <td>231.778018</td>\n",
       "      <td>122.573392</td>\n",
       "      <td>342.327431</td>\n",
       "      <td>783.041862</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4136.703664</td>\n",
       "      <td>827.439614</td>\n",
       "      <td>...</td>\n",
       "      <td>1.347606</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.539042</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.539042</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1447437_20171212_S</th>\n",
       "      <td>59.260325</td>\n",
       "      <td>1065.438272</td>\n",
       "      <td>543.947408</td>\n",
       "      <td>554.136026</td>\n",
       "      <td>428.171446</td>\n",
       "      <td>301.915763</td>\n",
       "      <td>679.310468</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3913.676651</td>\n",
       "      <td>732.956657</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1128691_20171206_S</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>426.619709</td>\n",
       "      <td>912.767750</td>\n",
       "      <td>49.606943</td>\n",
       "      <td>213.640568</td>\n",
       "      <td>277.798880</td>\n",
       "      <td>423.312580</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2093.412992</td>\n",
       "      <td>236.459761</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 4350 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid  \\\n",
       "1068841_20180306_S                                      10.085904   \n",
       "1447437_20171212_S                                      59.260325   \n",
       "1128691_20171206_S                                       0.000000   \n",
       "\n",
       "                    2_O_alpha_mannosyl_D_glycerate_utilization  \\\n",
       "1068841_20180306_S                                 2784.895948   \n",
       "1447437_20171212_S                                 1065.438272   \n",
       "1128691_20171206_S                                  426.619709   \n",
       "\n",
       "                    2_aminophenol_Metabolism  \\\n",
       "1068841_20180306_S                516.160945   \n",
       "1447437_20171212_S                543.947408   \n",
       "1128691_20171206_S                912.767750   \n",
       "\n",
       "                    2_ketoacid_oxidoreductases_disambiguation  \\\n",
       "1068841_20180306_S                                 231.778018   \n",
       "1447437_20171212_S                                 554.136026   \n",
       "1128691_20171206_S                                  49.606943   \n",
       "\n",
       "                    2_oxoglutarate_dehydrogenase_  2_phosphoglycolate_salvage  \\\n",
       "1068841_20180306_S                     122.573392                  342.327431   \n",
       "1447437_20171212_S                     428.171446                  301.915763   \n",
       "1128691_20171206_S                     213.640568                  277.798880   \n",
       "\n",
       "                    3_amino_5_hydroxybenzoic_Acid_Synthesis  \\\n",
       "1068841_20180306_S                               783.041862   \n",
       "1447437_20171212_S                               679.310468   \n",
       "1128691_20171206_S                               423.312580   \n",
       "\n",
       "                    4_hydroxybenzoyl_CoA_reductase  \\\n",
       "1068841_20180306_S                             0.0   \n",
       "1447437_20171212_S                             0.0   \n",
       "1128691_20171206_S                             0.0   \n",
       "\n",
       "                    5_methylaminomethyl_2_thiouridine  \\\n",
       "1068841_20180306_S                        4136.703664   \n",
       "1447437_20171212_S                        3913.676651   \n",
       "1128691_20171206_S                        2093.412992   \n",
       "\n",
       "                    A_Hypothetical_Protein_Related_to_Proline_Metabolism  ...  \\\n",
       "1068841_20180306_S                                         827.439614     ...   \n",
       "1447437_20171212_S                                         732.956657     ...   \n",
       "1128691_20171206_S                                         236.459761     ...   \n",
       "\n",
       "                    Oceaniferula  Persicirhabdus  Phragmitibacter  \\\n",
       "1068841_20180306_S      1.347606             0.0              0.0   \n",
       "1447437_20171212_S      0.000000             0.0              0.0   \n",
       "1128691_20171206_S      0.000000             0.0              0.0   \n",
       "\n",
       "                    Prosthecobacter  Roseibacillus  Roseimicrobium  \\\n",
       "1068841_20180306_S              0.0       0.539042             0.0   \n",
       "1447437_20171212_S              0.0       0.000000             0.0   \n",
       "1128691_20171206_S              0.0       0.000000             0.0   \n",
       "\n",
       "                    Sulfuriroseicoccus  Verrucomicrobium  Eremiobacter  \\\n",
       "1068841_20180306_S                 0.0          0.539042           0.0   \n",
       "1447437_20171212_S                 0.0          0.000000           0.0   \n",
       "1128691_20171206_S                 0.0          0.000000           0.0   \n",
       "\n",
       "                    Methylomirabilis  \n",
       "1068841_20180306_S               0.0  \n",
       "1447437_20171212_S               0.0  \n",
       "1128691_20171206_S               0.0  \n",
       "\n",
       "[3 rows x 4350 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dffe5996-5874-4ffd-8ac0-5f25f6d154b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minion</th>\n",
       "      <th>MGI</th>\n",
       "      <th>pwCF_ID</th>\n",
       "      <th>Sample_date</th>\n",
       "      <th>IP_vs_OP</th>\n",
       "      <th>Hospital</th>\n",
       "      <th>Room</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age_groups</th>\n",
       "      <th>Paediatric_vs_Adult</th>\n",
       "      <th>...</th>\n",
       "      <th>Sum_of_meds</th>\n",
       "      <th>Sum_of_antifungals</th>\n",
       "      <th>Sum_of_steroid_mabs</th>\n",
       "      <th>DNA_extraction__conc</th>\n",
       "      <th>SAGC_ULN</th>\n",
       "      <th>DNA_Conc_ng_ul_</th>\n",
       "      <th>Index_I7</th>\n",
       "      <th>Index_I5</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total_Clusters_Passing_Filter_Million_</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>634207_20180510_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>634207_20180510_S</td>\n",
       "      <td>634207</td>\n",
       "      <td>5/10/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Rm9</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>SAGCFN_22_01856</td>\n",
       "      <td>7.82</td>\n",
       "      <td>CGGACGATTC</td>\n",
       "      <td>CCACCACCTA</td>\n",
       "      <td>651</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634207_20180517_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>634207_20180517_S</td>\n",
       "      <td>634207</td>\n",
       "      <td>5/17/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Rm9</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.134</td>\n",
       "      <td>SAGCFN_22_01827</td>\n",
       "      <td>22.80</td>\n",
       "      <td>AGCGATAG</td>\n",
       "      <td>CCTATCCT</td>\n",
       "      <td>633</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715927_20180205_S</th>\n",
       "      <td>715927_20180205_S</td>\n",
       "      <td>715927_20180205_S</td>\n",
       "      <td>715927</td>\n",
       "      <td>2/05/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Level 6 DK Office</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.326</td>\n",
       "      <td>SAGCFN_22_01797</td>\n",
       "      <td>16.50</td>\n",
       "      <td>TAATGCGC</td>\n",
       "      <td>AGGCGAAG</td>\n",
       "      <td>516</td>\n",
       "      <td>3.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 166 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              minion                MGI  pwCF_ID Sample_date  \\\n",
       "NAME                                                                           \n",
       "634207_20180510_S                NaN  634207_20180510_S   634207   5/10/2018   \n",
       "634207_20180517_S                NaN  634207_20180517_S   634207   5/17/2018   \n",
       "715927_20180205_S  715927_20180205_S  715927_20180205_S   715927   2/05/2018   \n",
       "\n",
       "                  IP_vs_OP Hospital               Room  Age  Age_groups  \\\n",
       "NAME                                                                      \n",
       "634207_20180510_S       IP      WCH           Adol Rm9   17           3   \n",
       "634207_20180517_S       IP      WCH           Adol Rm9   17           3   \n",
       "715927_20180205_S       OP      WCH  Level 6 DK Office   13           3   \n",
       "\n",
       "                  Paediatric_vs_Adult  ... Sum_of_meds Sum_of_antifungals  \\\n",
       "NAME                                   ...                                  \n",
       "634207_20180510_S          Paediatric  ...           1                  0   \n",
       "634207_20180517_S          Paediatric  ...           1                  0   \n",
       "715927_20180205_S          Paediatric  ...           1                  0   \n",
       "\n",
       "                   Sum_of_steroid_mabs  DNA_extraction__conc         SAGC_ULN  \\\n",
       "NAME                                                                            \n",
       "634207_20180510_S                    0                 0.000  SAGCFN_22_01856   \n",
       "634207_20180517_S                    0                 0.134  SAGCFN_22_01827   \n",
       "715927_20180205_S                    0                 0.326  SAGCFN_22_01797   \n",
       "\n",
       "                   DNA_Conc_ng_ul_    Index_I7    Index_I5  Mean_Size_BP  \\\n",
       "NAME                                                                       \n",
       "634207_20180510_S             7.82  CGGACGATTC  CCACCACCTA           651   \n",
       "634207_20180517_S            22.80    AGCGATAG    CCTATCCT           633   \n",
       "715927_20180205_S            16.50    TAATGCGC    AGGCGAAG           516   \n",
       "\n",
       "                  Total_Clusters_Passing_Filter_Million_  \n",
       "NAME                                                      \n",
       "634207_20180510_S                                    2.9  \n",
       "634207_20180517_S                                    2.4  \n",
       "715927_20180205_S                                    3.4  \n",
       "\n",
       "[3 rows x 166 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b86206ad-b3e9-4a75-a72c-1aa8a6018b7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan 0.0 nan nan 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\" \".join(map(str, list(metadata[\"N12M_Pseudomonas_aeruginosa\"].values)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "997b5c90-a628-4d2a-be56-ad847b1b92dd",
   "metadata": {},
   "source": [
    "## Convert categories and objects to numeric forms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2b41c3cf-14a7-4acb-a8db-9a020f2330d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping minion: object\n",
      "Dropping Sample_date: object\n",
      "Dropping IP_vs_OP: object\n",
      "Dropping Hospital: object\n",
      "Dropping Room: object\n",
      "Dropping Paediatric_vs_Adult: object\n",
      "Dropping Gender: object\n",
      "Dropping Sample_Type: object\n",
      "Dropping Culture_Result_Matched_with_sequenced_sample_: object\n",
      "Dropping Cutured_in_previous_12_months: object\n",
      "Dropping Others_cultured: object\n",
      "Dropping Spec_IgG: object\n",
      "Dropping Cystic_Fibrosis_related_diabetes_CFRD_: object\n",
      "Dropping Pancreatic_insufficiency_PI_: object\n",
      "Dropping CF_gene_1: object\n",
      "Dropping CF_gene_2: object\n",
      "Dropping Notes: object\n",
      "Dropping CFLD: object\n",
      "Dropping Next_PA_Positive_Date: object\n",
      "Dropping Positive_for_PA_within_next_12_months: object\n",
      "Dropping Culture_for_next_sample: object\n",
      "Dropping Antibiotics_duration_: object\n",
      "Dropping Antibiotics_YN: object\n",
      "Dropping SAGC_ULN: object\n",
      "Dropping Index_I7: object\n",
      "Dropping Index_I5: object\n"
     ]
    }
   ],
   "source": [
    "# we make a copy of each data frame and then overwrite the columns. Internally this is more efficient than making a df one\n",
    "# column at a time.\n",
    "encoded_metadata = metadata.copy()\n",
    "to_delete = []\n",
    "for col in metadata:\n",
    "    if pd.api.types.is_numeric_dtype(metadata[col]):\n",
    "        continue\n",
    "    elif col == sequence_type:\n",
    "        continue\n",
    "    elif metadata[col].dtypes == 'category':\n",
    "        encoded_metadata[col] = metadata[col].cat.codes\n",
    "    else:\n",
    "        print(f'Dropping {col}: {metadata[col].dtypes}')\n",
    "        to_delete.append(col)\n",
    "\n",
    "encoded_metadata = encoded_metadata.drop(columns=to_delete)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "687d40de-2faf-4820-8f2c-641893438d8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NAME\n",
       "634207_20180510_S            NaN\n",
       "634207_20180517_S            NaN\n",
       "715927_20180205_S     18/03/2019\n",
       "715927_20180213_S     18/03/2019\n",
       "715927_20180226_S     18/03/2019\n",
       "                         ...    \n",
       "1651490_20180206_S     4/09/2018\n",
       "1651490_20171215_S     4/09/2018\n",
       "1658447_20171006_S     6/10/2017\n",
       "1664053_20180406_S     6/04/2018\n",
       "1690154_20180406_S     6/04/2018\n",
       "Name: Next_PA_Positive_Date, Length: 127, dtype: object"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['Next_PA_Positive_Date']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba6221a3-5165-46e8-81ae-991772bff65b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "54ead384-0cee-49d8-b3ef-0ecef29ee0e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Not sure about minion: object\n",
      "Not sure about MGI: object\n",
      "Not sure about pwCF_ID: int64\n",
      "Not sure about Sample_date: object\n",
      "Not sure about Room: object\n",
      "Not sure about Age: int64\n",
      "Not sure about H2_Uncorrected: float64\n",
      "Not sure about CH4_Uncorrected: float64\n",
      "Not sure about CO2: float64\n",
      "Not sure about H2_Corrected: float64\n",
      "Not sure about CH4_Corrected: float64\n",
      "Not sure about CH4_H2_ratio_corrected: float64\n",
      "Not sure about Corr_: float64\n",
      "Not sure about Culture_Result_Matched_with_sequenced_sample_: object\n",
      "Not sure about Pseudomonas_Culture: float64\n",
      "Not sure about Cutured_in_previous_12_months: object\n",
      "Not sure about Others_cultured: object\n",
      "Not sure about IgE: float64\n",
      "Not sure about Spec_IgE: float64\n",
      "Not sure about Spec_IgG: object\n",
      "Not sure about Precipitins: float64\n",
      "Not sure about FVC: float64\n",
      "Not sure about FEV1: float64\n",
      "Not sure about Best_FEV1: int64\n",
      "Not sure about FEV1_best_FEV1: float64\n",
      "Not sure about CF_gene_2: object\n",
      "Not sure about Notes: object\n",
      "Not sure about CS_Scedosporium_apiospermum_1: int64\n",
      "Not sure about Next_PA_Positive_Date: object\n",
      "Not sure about Culture_for_next_sample: object\n",
      "Not sure about Antibiotics_duration_: object\n",
      "Not sure about 3_piperacillin_sodium_tazobactam_sodium_Tazocin__IV: int64\n",
      "Not sure about DNA_extraction__conc: float64\n",
      "Not sure about SAGC_ULN: object\n",
      "Not sure about DNA_Conc_ng_ul_: float64\n",
      "Not sure about Index_I7: object\n",
      "Not sure about Index_I5: object\n",
      "Not sure about Mean_Size_BP: int64\n",
      "Not sure about Total_Clusters_Passing_Filter_Million_: float64\n"
     ]
    }
   ],
   "source": [
    "#encoder = LabelEncoder()\n",
    "encoder = OrdinalEncoder()\n",
    "\n",
    "# we make a copy of each data frame and then overwrite the columns. Internally this is more efficient than making a df one\n",
    "# column at a time.\n",
    "encoded_metadata = metadata.copy()\n",
    "for col in metadata:\n",
    "    if not pd.api.types.is_numeric_dtype(metadata[col]):\n",
    "        encoded_metadata[col] = metadata[col].cat.codes\n",
    "\n",
    "encoded_df = df.copy()\n",
    "for col in df:\n",
    "    if not pd.api.types.is_numeric_dtype(df[col]):\n",
    "        encoded_df[col] = encoder.fit_transform([metadata[col].values])[0]  # Ensure all values are strings\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "30a6e99c-7976-4417-97b6-c1af34bf18cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan 0.0 nan nan 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\" \".join(map(str, list(encoded_metadata[\"N12M_Pseudomonas_aeruginosa\"].values)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "114be50c-3f43-4d77-9f70-7ab0dc67189b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a88ceb42-5a0b-4b70-9496-9a6b89d20dc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NAME\n",
       "634207_20180510_S     0\n",
       "634207_20180517_S     0\n",
       "715927_20180205_S     0\n",
       "715927_20180213_S     0\n",
       "715927_20180226_S     0\n",
       "                     ..\n",
       "1651490_20180206_S    1\n",
       "1651490_20171215_S    1\n",
       "1658447_20171006_S    1\n",
       "1664053_20180406_S    1\n",
       "1690154_20180406_S    1\n",
       "Name: N12M_Pseudomonas_aeruginosa, Length: 110, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[\"N12M_Pseudomonas_aeruginosa\"].dropna().astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa47e822-1a1d-4b5a-8f8d-95deabbd2e8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f5e7efe-1302-4143-b09d-d01b4d68b277",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "400cd129-09ea-4793-bb44-1867a196b658",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fb50e6c-bfa2-40f9-be4b-a25ac5d888a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "220248b4-7cea-4117-b614-42b49cba9320",
   "metadata": {},
   "source": [
    "# Predict something \"simple\"\n",
    "\n",
    "Here, we are just trying to predict Pseudomonas positive in the next 12 months!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2bbff15f-b949-43ac-a018-b84d95d544e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df: (127, 4350)\n",
      "metadata: (127, 166)\n",
      "df_combined: (110, 4517)\n"
     ]
    }
   ],
   "source": [
    "df_combined = encoded_df.merge(encoded_metadata, left_index=True, right_index=True, how='inner')\n",
    "df_combined = df_combined.dropna(subset=['N12M_Pseudomonas_aeruginosa'])\n",
    "df_combined = df_combined.dropna(subset=[\"pwCF_ID\"])\n",
    "df_combined[\"N12M_Pseudomonas_aeruginosa\"] = df_combined[\"N12M_Pseudomonas_aeruginosa\"].astype(int)\n",
    "df_combined = df_combined.reset_index()\n",
    "\n",
    "print(f\"df: {df.shape}\")\n",
    "print(f\"metadata: {metadata.shape}\")\n",
    "print(f\"df_combined: {df_combined.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f518c0d1-7b76-4e75-ab89-fa0737849e0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N12M_Pseudomonas_aeruginosa ~ Pedobacter + CS_Pseudomonas_aeruginosa + Pseudomonas_Culture\n"
     ]
    }
   ],
   "source": [
    "random_column_number = 100\n",
    "#random_columns = np.random.choice(df.columns, size=random_column_number, replace=False)\n",
    "random_columns = ['Pedobacter']\n",
    "predictors = \" + \".join(random_columns)  # Concatenate all bacterial abundance column names\n",
    "formula = f\"N12M_Pseudomonas_aeruginosa ~ {predictors} + CS_Pseudomonas_aeruginosa + Pseudomonas_Culture\"\n",
    "print(formula)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "02cf9ce5-c05d-47fc-93cc-fadd85d49b8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Limosilactobacillus', 'Pedobacter', 'CS_Pseudomonas_aeruginosa', 'Pseudomonas_Culture'}\n"
     ]
    }
   ],
   "source": [
    "print(set(formula.split(\" ~ \")[1].replace(\"+\", \"\").split()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "56e1a0d4-7b3d-42ff-97cc-46871ac2d37c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Aegiribacteria', 'Kapaibacterium', 'Stahlbacteria', 'Vermiphilus',\n",
       "       'Babela', 'Chromulinivorax', 'Ozemobacter', 'Desulfacyla',\n",
       "       'Dissulfurimicrobium', 'Dissulfurirhabdus',\n",
       "       ...\n",
       "       'Oceaniferula', 'Persicirhabdus', 'Phragmitibacter', 'Prosthecobacter',\n",
       "       'Roseibacillus', 'Roseimicrobium', 'Sulfuriroseicoccus',\n",
       "       'Verrucomicrobium', 'Eremiobacter', 'Methylomirabilis'],\n",
       "      dtype='object', name='taxonomy', length=3581)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genus_otu.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e0b72dac-2ca7-455c-840c-d06b2abc0c96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(110, 4517)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined.dropna(subset=random_columns).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e72b0c13-182d-4751-8445-dd713b88a129",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Pedobacter',\n",
       " 'CS_Pseudomonas_aeruginosa',\n",
       " 'Pseudomonas_Culture',\n",
       " 'N12M_Pseudomonas_aeruginosa']"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_columns = random_columns + ['CS_Pseudomonas_aeruginosa' ,'Pseudomonas_Culture', 'N12M_Pseudomonas_aeruginosa']\n",
    "all_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8a1d83b2-6e2f-459c-9303-ded8b1575485",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid</th>\n",
       "      <th>2_O_alpha_mannosyl_D_glycerate_utilization</th>\n",
       "      <th>2_aminophenol_Metabolism</th>\n",
       "      <th>2_ketoacid_oxidoreductases_disambiguation</th>\n",
       "      <th>2_oxoglutarate_dehydrogenase_</th>\n",
       "      <th>2_phosphoglycolate_salvage</th>\n",
       "      <th>3_amino_5_hydroxybenzoic_Acid_Synthesis</th>\n",
       "      <th>4_hydroxybenzoyl_CoA_reductase</th>\n",
       "      <th>5_methylaminomethyl_2_thiouridine</th>\n",
       "      <th>...</th>\n",
       "      <th>Sum_of_meds</th>\n",
       "      <th>Sum_of_antifungals</th>\n",
       "      <th>Sum_of_steroid_mabs</th>\n",
       "      <th>DNA_extraction__conc</th>\n",
       "      <th>SAGC_ULN</th>\n",
       "      <th>DNA_Conc_ng_ul_</th>\n",
       "      <th>Index_I7</th>\n",
       "      <th>Index_I5</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total_Clusters_Passing_Filter_Million_</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>10.085904</td>\n",
       "      <td>2784.895948</td>\n",
       "      <td>516.160945</td>\n",
       "      <td>231.778018</td>\n",
       "      <td>122.573392</td>\n",
       "      <td>342.327431</td>\n",
       "      <td>783.041862</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4136.703664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1447437_20171212_S</td>\n",
       "      <td>59.260325</td>\n",
       "      <td>1065.438272</td>\n",
       "      <td>543.947408</td>\n",
       "      <td>554.136026</td>\n",
       "      <td>428.171446</td>\n",
       "      <td>301.915763</td>\n",
       "      <td>679.310468</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3913.676651</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>498</td>\n",
       "      <td>4.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1128691_20171206_S</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>426.619709</td>\n",
       "      <td>912.767750</td>\n",
       "      <td>49.606943</td>\n",
       "      <td>213.640568</td>\n",
       "      <td>277.798880</td>\n",
       "      <td>423.312580</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2093.412992</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.750</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>535</td>\n",
       "      <td>4.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1128691_20171218_S</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>659.087578</td>\n",
       "      <td>864.137047</td>\n",
       "      <td>139.140711</td>\n",
       "      <td>235.318676</td>\n",
       "      <td>355.174973</td>\n",
       "      <td>538.254855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2050.494687</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>352</td>\n",
       "      <td>4.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1128691_20180116_S</td>\n",
       "      <td>14.478968</td>\n",
       "      <td>159.268644</td>\n",
       "      <td>593.637673</td>\n",
       "      <td>48.263225</td>\n",
       "      <td>358.354449</td>\n",
       "      <td>202.705547</td>\n",
       "      <td>348.701804</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1570.967988</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.708</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>500</td>\n",
       "      <td>3.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>877469_20190115_S</td>\n",
       "      <td>6.849038</td>\n",
       "      <td>4314.893843</td>\n",
       "      <td>1041.053753</td>\n",
       "      <td>22.830126</td>\n",
       "      <td>67.120571</td>\n",
       "      <td>352.725449</td>\n",
       "      <td>402.951727</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4232.705389</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.332</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>383</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>892355_20180123_S</td>\n",
       "      <td>11.073270</td>\n",
       "      <td>227.002043</td>\n",
       "      <td>586.883332</td>\n",
       "      <td>606.261555</td>\n",
       "      <td>683.036230</td>\n",
       "      <td>243.611949</td>\n",
       "      <td>653.784340</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4210.611072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.890</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>349</td>\n",
       "      <td>5.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>895293_20180502_S</td>\n",
       "      <td>11.343610</td>\n",
       "      <td>250.504720</td>\n",
       "      <td>1278.046721</td>\n",
       "      <td>264.054032</td>\n",
       "      <td>86.589556</td>\n",
       "      <td>229.708102</td>\n",
       "      <td>753.089661</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3424.824904</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.900</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>584</td>\n",
       "      <td>4.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>980574_20180403_S</td>\n",
       "      <td>15.281074</td>\n",
       "      <td>1205.624053</td>\n",
       "      <td>444.731949</td>\n",
       "      <td>383.080719</td>\n",
       "      <td>119.209942</td>\n",
       "      <td>228.425711</td>\n",
       "      <td>929.423029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3243.276243</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.216</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>364</td>\n",
       "      <td>2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>983493_20180123_S</td>\n",
       "      <td>490.780540</td>\n",
       "      <td>556.217945</td>\n",
       "      <td>1145.154592</td>\n",
       "      <td>812.514449</td>\n",
       "      <td>257.278065</td>\n",
       "      <td>498.960215</td>\n",
       "      <td>452.608720</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2574.961898</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.698</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>469</td>\n",
       "      <td>4.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108 rows Ã— 4517 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  index  2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid  \\\n",
       "0    1068841_20180306_S                                      10.085904   \n",
       "1    1447437_20171212_S                                      59.260325   \n",
       "2    1128691_20171206_S                                       0.000000   \n",
       "3    1128691_20171218_S                                       0.000000   \n",
       "4    1128691_20180116_S                                      14.478968   \n",
       "..                  ...                                            ...   \n",
       "105   877469_20190115_S                                       6.849038   \n",
       "106   892355_20180123_S                                      11.073270   \n",
       "107   895293_20180502_S                                      11.343610   \n",
       "108   980574_20180403_S                                      15.281074   \n",
       "109   983493_20180123_S                                     490.780540   \n",
       "\n",
       "     2_O_alpha_mannosyl_D_glycerate_utilization  2_aminophenol_Metabolism  \\\n",
       "0                                   2784.895948                516.160945   \n",
       "1                                   1065.438272                543.947408   \n",
       "2                                    426.619709                912.767750   \n",
       "3                                    659.087578                864.137047   \n",
       "4                                    159.268644                593.637673   \n",
       "..                                          ...                       ...   \n",
       "105                                 4314.893843               1041.053753   \n",
       "106                                  227.002043                586.883332   \n",
       "107                                  250.504720               1278.046721   \n",
       "108                                 1205.624053                444.731949   \n",
       "109                                  556.217945               1145.154592   \n",
       "\n",
       "     2_ketoacid_oxidoreductases_disambiguation  2_oxoglutarate_dehydrogenase_  \\\n",
       "0                                   231.778018                     122.573392   \n",
       "1                                   554.136026                     428.171446   \n",
       "2                                    49.606943                     213.640568   \n",
       "3                                   139.140711                     235.318676   \n",
       "4                                    48.263225                     358.354449   \n",
       "..                                         ...                            ...   \n",
       "105                                  22.830126                      67.120571   \n",
       "106                                 606.261555                     683.036230   \n",
       "107                                 264.054032                      86.589556   \n",
       "108                                 383.080719                     119.209942   \n",
       "109                                 812.514449                     257.278065   \n",
       "\n",
       "     2_phosphoglycolate_salvage  3_amino_5_hydroxybenzoic_Acid_Synthesis  \\\n",
       "0                    342.327431                               783.041862   \n",
       "1                    301.915763                               679.310468   \n",
       "2                    277.798880                               423.312580   \n",
       "3                    355.174973                               538.254855   \n",
       "4                    202.705547                               348.701804   \n",
       "..                          ...                                      ...   \n",
       "105                  352.725449                               402.951727   \n",
       "106                  243.611949                               653.784340   \n",
       "107                  229.708102                               753.089661   \n",
       "108                  228.425711                               929.423029   \n",
       "109                  498.960215                               452.608720   \n",
       "\n",
       "     4_hydroxybenzoyl_CoA_reductase  5_methylaminomethyl_2_thiouridine  ...  \\\n",
       "0                               0.0                        4136.703664  ...   \n",
       "1                               0.0                        3913.676651  ...   \n",
       "2                               0.0                        2093.412992  ...   \n",
       "3                               0.0                        2050.494687  ...   \n",
       "4                               0.0                        1570.967988  ...   \n",
       "..                              ...                                ...  ...   \n",
       "105                             0.0                        4232.705389  ...   \n",
       "106                             0.0                        4210.611072  ...   \n",
       "107                             0.0                        3424.824904  ...   \n",
       "108                             0.0                        3243.276243  ...   \n",
       "109                             0.0                        2574.961898  ...   \n",
       "\n",
       "     Sum_of_meds  Sum_of_antifungals  Sum_of_steroid_mabs  \\\n",
       "0            0.0                 0.0                  0.0   \n",
       "1            0.0                 0.0                  0.0   \n",
       "2            0.0                 0.0                  0.0   \n",
       "3            0.0                 0.0                  0.0   \n",
       "4            0.0                 0.0                  0.0   \n",
       "..           ...                 ...                  ...   \n",
       "105          0.0                 0.0                  0.0   \n",
       "106          0.0                 0.0                  0.0   \n",
       "107          0.0                 0.0                  0.0   \n",
       "108          0.0                 0.0                  0.0   \n",
       "109          0.0                 0.0                  0.0   \n",
       "\n",
       "     DNA_extraction__conc  SAGC_ULN  DNA_Conc_ng_ul_  Index_I7  Index_I5  \\\n",
       "0                   1.070       0.0           42.800       0.0       0.0   \n",
       "1                   1.510       0.0           32.600       0.0       0.0   \n",
       "2                   1.750       0.0           33.400       0.0       0.0   \n",
       "3                   3.400       0.0           30.800       0.0       0.0   \n",
       "4                   0.708       0.0           27.400       0.0       0.0   \n",
       "..                    ...       ...              ...       ...       ...   \n",
       "105                 0.000       0.0            0.332       0.0       0.0   \n",
       "106                 2.890       0.0           21.200       0.0       0.0   \n",
       "107                 0.442       0.0           16.900       0.0       0.0   \n",
       "108                 0.216       0.0           34.600       0.0       0.0   \n",
       "109                 0.698       0.0           21.600       0.0       0.0   \n",
       "\n",
       "     Mean_Size_BP  Total_Clusters_Passing_Filter_Million_  \n",
       "0             417                                     2.8  \n",
       "1             498                                     4.7  \n",
       "2             535                                     4.5  \n",
       "3             352                                     4.8  \n",
       "4             500                                     3.8  \n",
       "..            ...                                     ...  \n",
       "105           383                                     0.2  \n",
       "106           349                                     5.1  \n",
       "107           584                                     4.6  \n",
       "108           364                                     2.2  \n",
       "109           469                                     4.4  \n",
       "\n",
       "[108 rows x 4517 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6b6c1ca-d660-4743-9e00-386a866e11e3",
   "metadata": {},
   "source": [
    "### Check for colinearity\n",
    "\n",
    "If VIF > 10, the variable is highly collinear. Remove or combine collinear predictors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "42771bf1-4511-4724-ad21-bdae5289c0d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.outliers_influence import variance_inflation_factor\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9ceb2080-12b2-4543-beca-637eb3d5a73c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      Variable       VIF\n",
      "0                   Pedobacter  1.004945\n",
      "1    CS_Pseudomonas_aeruginosa       NaN\n",
      "2          Pseudomonas_Culture  1.004945\n",
      "3  N12M_Pseudomonas_aeruginosa       NaN\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/edwa0468/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/regression/linear_model.py:1738: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  return 1 - self.ssr/self.uncentered_tss\n"
     ]
    }
   ],
   "source": [
    "X = df_combined[all_columns]\n",
    "vif_data = pd.DataFrame()\n",
    "vif_data[\"Variable\"] = X.columns\n",
    "vif_data[\"VIF\"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]\n",
    "\n",
    "print(vif_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c285f117-9737-4f3e-bbfd-2853e309be50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56\n"
     ]
    }
   ],
   "source": [
    "print(df_combined[\"pwCF_ID\"].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "35f1f6ec-b52d-4f8a-a7f0-01e1ceb934a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index                                             True\n",
       "2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid     True\n",
       "2_O_alpha_mannosyl_D_glycerate_utilization        True\n",
       "2_aminophenol_Metabolism                          True\n",
       "2_ketoacid_oxidoreductases_disambiguation         True\n",
       "                                                 ...  \n",
       "DNA_Conc_ng_ul_                                   True\n",
       "Index_I7                                         False\n",
       "Index_I5                                         False\n",
       "Mean_Size_BP                                      True\n",
       "Total_Clusters_Passing_Filter_Million_            True\n",
       "Length: 4517, dtype: bool"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined.nunique() > 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "79e6f987-417e-4a78-a361-d6d71a68fb0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Rhodonellum', 'Fluviicoccus', 'Pseudotabrizicola', 'Halocynthiibacter', 'Haematobacter', 'DNA_phosphorothioation_dependent_restriction_system_DptFGH', 'Albibacterium', 'Geminisphaera', 'Spirillospora', 'Neptunochlamydia', 'Georhizobium', 'Brevefilum', 'Rhodobaca', 'Lasso_peptide_maturation_cluster', 'Vitiosangium', 'Geopsychrobacter', 'Chroococcus', 'Hephaestia', 'Mycobacterial_cell_wall_virulence_lipid_phthiocerol_dimycocerosate_PDIM_', 'Sporosalibacterium', 'Inmirania', 'Hongsoonwoonella', 'Luteithermobacter', 'Emcibacter', 'Methylohalobius', 'Salibaculum', 'Hydrogenobacter', 'Toluene_4_monooxygenase_T4MO_', 'Igneacidithiobacillus'}\n"
     ]
    }
   ],
   "source": [
    "df_combined3 = df_combined.loc[:, df_combined.nunique() > 1].dropna()\n",
    "print(set(df.columns) - set(df_combined3.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "6b227aa7-74dd-43b2-8006-0374562a581d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined[\"N12M_Pseudomonas_aeruginosa\"].astype(int).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc44cac9-d42d-472c-a773-8c659debcd1d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9158fe32-9d24-4e94-bdfe-1055f056bce4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2a921f9-f391-48fd-b181-af55bc703286",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aa3ad0e-3e30-4240-b945-9f4dd667ef87",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c864b5c1-c03f-448e-a591-702c5791a760",
   "metadata": {},
   "source": [
    "## Use PCA to reduce the dimensionality of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f850070e-19d2-4b98-8527-7f96782c995c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid</th>\n",
       "      <th>2_O_alpha_mannosyl_D_glycerate_utilization</th>\n",
       "      <th>2_aminophenol_Metabolism</th>\n",
       "      <th>2_ketoacid_oxidoreductases_disambiguation</th>\n",
       "      <th>2_oxoglutarate_dehydrogenase_</th>\n",
       "      <th>2_phosphoglycolate_salvage</th>\n",
       "      <th>3_amino_5_hydroxybenzoic_Acid_Synthesis</th>\n",
       "      <th>4_hydroxybenzoyl_CoA_reductase</th>\n",
       "      <th>5_methylaminomethyl_2_thiouridine</th>\n",
       "      <th>...</th>\n",
       "      <th>DNA_Conc_ng_ul_</th>\n",
       "      <th>Index_I7</th>\n",
       "      <th>Index_I5</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total_Clusters_Passing_Filter_Million_</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>10.085904</td>\n",
       "      <td>2784.895948</td>\n",
       "      <td>516.160945</td>\n",
       "      <td>231.778018</td>\n",
       "      <td>122.573392</td>\n",
       "      <td>342.327431</td>\n",
       "      <td>783.041862</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4136.703664</td>\n",
       "      <td>...</td>\n",
       "      <td>42.800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "      <td>105232.294537</td>\n",
       "      <td>29481.534204</td>\n",
       "      <td>3009.236199</td>\n",
       "      <td>-748.874391</td>\n",
       "      <td>1393.844533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1447437_20171212_S</td>\n",
       "      <td>59.260325</td>\n",
       "      <td>1065.438272</td>\n",
       "      <td>543.947408</td>\n",
       "      <td>554.136026</td>\n",
       "      <td>428.171446</td>\n",
       "      <td>301.915763</td>\n",
       "      <td>679.310468</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3913.676651</td>\n",
       "      <td>...</td>\n",
       "      <td>32.600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>498</td>\n",
       "      <td>4.7</td>\n",
       "      <td>-68177.360588</td>\n",
       "      <td>-173244.331191</td>\n",
       "      <td>150955.115109</td>\n",
       "      <td>58275.741094</td>\n",
       "      <td>15805.072273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1128691_20171206_S</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>426.619709</td>\n",
       "      <td>912.767750</td>\n",
       "      <td>49.606943</td>\n",
       "      <td>213.640568</td>\n",
       "      <td>277.798880</td>\n",
       "      <td>423.312580</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2093.412992</td>\n",
       "      <td>...</td>\n",
       "      <td>33.400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>535</td>\n",
       "      <td>4.5</td>\n",
       "      <td>-211025.239559</td>\n",
       "      <td>-151897.294836</td>\n",
       "      <td>-320966.178592</td>\n",
       "      <td>262834.328944</td>\n",
       "      <td>74715.555125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1128691_20171218_S</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>659.087578</td>\n",
       "      <td>864.137047</td>\n",
       "      <td>139.140711</td>\n",
       "      <td>235.318676</td>\n",
       "      <td>355.174973</td>\n",
       "      <td>538.254855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2050.494687</td>\n",
       "      <td>...</td>\n",
       "      <td>30.800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>352</td>\n",
       "      <td>4.8</td>\n",
       "      <td>-188035.291051</td>\n",
       "      <td>-141834.133045</td>\n",
       "      <td>-266846.807833</td>\n",
       "      <td>226853.551490</td>\n",
       "      <td>69220.178950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1128691_20180116_S</td>\n",
       "      <td>14.478968</td>\n",
       "      <td>159.268644</td>\n",
       "      <td>593.637673</td>\n",
       "      <td>48.263225</td>\n",
       "      <td>358.354449</td>\n",
       "      <td>202.705547</td>\n",
       "      <td>348.701804</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1570.967988</td>\n",
       "      <td>...</td>\n",
       "      <td>27.400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>500</td>\n",
       "      <td>3.8</td>\n",
       "      <td>-231769.880818</td>\n",
       "      <td>-155931.482253</td>\n",
       "      <td>-328612.917308</td>\n",
       "      <td>256193.671691</td>\n",
       "      <td>79085.256264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>877469_20190115_S</td>\n",
       "      <td>6.849038</td>\n",
       "      <td>4314.893843</td>\n",
       "      <td>1041.053753</td>\n",
       "      <td>22.830126</td>\n",
       "      <td>67.120571</td>\n",
       "      <td>352.725449</td>\n",
       "      <td>402.951727</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4232.705389</td>\n",
       "      <td>...</td>\n",
       "      <td>0.332</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>383</td>\n",
       "      <td>0.2</td>\n",
       "      <td>433871.879019</td>\n",
       "      <td>-50971.981671</td>\n",
       "      <td>-5320.962053</td>\n",
       "      <td>24773.058456</td>\n",
       "      <td>24120.785059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>892355_20180123_S</td>\n",
       "      <td>11.073270</td>\n",
       "      <td>227.002043</td>\n",
       "      <td>586.883332</td>\n",
       "      <td>606.261555</td>\n",
       "      <td>683.036230</td>\n",
       "      <td>243.611949</td>\n",
       "      <td>653.784340</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4210.611072</td>\n",
       "      <td>...</td>\n",
       "      <td>21.200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>349</td>\n",
       "      <td>5.1</td>\n",
       "      <td>-242787.205324</td>\n",
       "      <td>-340573.238777</td>\n",
       "      <td>401656.123059</td>\n",
       "      <td>192192.110121</td>\n",
       "      <td>-6890.326181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>895293_20180502_S</td>\n",
       "      <td>11.343610</td>\n",
       "      <td>250.504720</td>\n",
       "      <td>1278.046721</td>\n",
       "      <td>264.054032</td>\n",
       "      <td>86.589556</td>\n",
       "      <td>229.708102</td>\n",
       "      <td>753.089661</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3424.824904</td>\n",
       "      <td>...</td>\n",
       "      <td>16.900</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>584</td>\n",
       "      <td>4.6</td>\n",
       "      <td>95694.320998</td>\n",
       "      <td>22992.264925</td>\n",
       "      <td>-6039.669020</td>\n",
       "      <td>-3148.567876</td>\n",
       "      <td>13543.206780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>980574_20180403_S</td>\n",
       "      <td>15.281074</td>\n",
       "      <td>1205.624053</td>\n",
       "      <td>444.731949</td>\n",
       "      <td>383.080719</td>\n",
       "      <td>119.209942</td>\n",
       "      <td>228.425711</td>\n",
       "      <td>929.423029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3243.276243</td>\n",
       "      <td>...</td>\n",
       "      <td>34.600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>364</td>\n",
       "      <td>2.2</td>\n",
       "      <td>-4008.324996</td>\n",
       "      <td>95743.541072</td>\n",
       "      <td>17895.588170</td>\n",
       "      <td>2995.813400</td>\n",
       "      <td>23541.421539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>983493_20180123_S</td>\n",
       "      <td>490.780540</td>\n",
       "      <td>556.217945</td>\n",
       "      <td>1145.154592</td>\n",
       "      <td>812.514449</td>\n",
       "      <td>257.278065</td>\n",
       "      <td>498.960215</td>\n",
       "      <td>452.608720</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2574.961898</td>\n",
       "      <td>...</td>\n",
       "      <td>21.600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>469</td>\n",
       "      <td>4.4</td>\n",
       "      <td>-47651.404957</td>\n",
       "      <td>-78848.180154</td>\n",
       "      <td>-122746.333045</td>\n",
       "      <td>51511.251975</td>\n",
       "      <td>32122.109092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108 rows Ã— 4522 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  index  2_3_diacetamido_2_3_dideoxy_d_mannuronic_acid  \\\n",
       "0    1068841_20180306_S                                      10.085904   \n",
       "1    1447437_20171212_S                                      59.260325   \n",
       "2    1128691_20171206_S                                       0.000000   \n",
       "3    1128691_20171218_S                                       0.000000   \n",
       "4    1128691_20180116_S                                      14.478968   \n",
       "..                  ...                                            ...   \n",
       "103   877469_20190115_S                                       6.849038   \n",
       "104   892355_20180123_S                                      11.073270   \n",
       "105   895293_20180502_S                                      11.343610   \n",
       "106   980574_20180403_S                                      15.281074   \n",
       "107   983493_20180123_S                                     490.780540   \n",
       "\n",
       "     2_O_alpha_mannosyl_D_glycerate_utilization  2_aminophenol_Metabolism  \\\n",
       "0                                   2784.895948                516.160945   \n",
       "1                                   1065.438272                543.947408   \n",
       "2                                    426.619709                912.767750   \n",
       "3                                    659.087578                864.137047   \n",
       "4                                    159.268644                593.637673   \n",
       "..                                          ...                       ...   \n",
       "103                                 4314.893843               1041.053753   \n",
       "104                                  227.002043                586.883332   \n",
       "105                                  250.504720               1278.046721   \n",
       "106                                 1205.624053                444.731949   \n",
       "107                                  556.217945               1145.154592   \n",
       "\n",
       "     2_ketoacid_oxidoreductases_disambiguation  2_oxoglutarate_dehydrogenase_  \\\n",
       "0                                   231.778018                     122.573392   \n",
       "1                                   554.136026                     428.171446   \n",
       "2                                    49.606943                     213.640568   \n",
       "3                                   139.140711                     235.318676   \n",
       "4                                    48.263225                     358.354449   \n",
       "..                                         ...                            ...   \n",
       "103                                  22.830126                      67.120571   \n",
       "104                                 606.261555                     683.036230   \n",
       "105                                 264.054032                      86.589556   \n",
       "106                                 383.080719                     119.209942   \n",
       "107                                 812.514449                     257.278065   \n",
       "\n",
       "     2_phosphoglycolate_salvage  3_amino_5_hydroxybenzoic_Acid_Synthesis  \\\n",
       "0                    342.327431                               783.041862   \n",
       "1                    301.915763                               679.310468   \n",
       "2                    277.798880                               423.312580   \n",
       "3                    355.174973                               538.254855   \n",
       "4                    202.705547                               348.701804   \n",
       "..                          ...                                      ...   \n",
       "103                  352.725449                               402.951727   \n",
       "104                  243.611949                               653.784340   \n",
       "105                  229.708102                               753.089661   \n",
       "106                  228.425711                               929.423029   \n",
       "107                  498.960215                               452.608720   \n",
       "\n",
       "     4_hydroxybenzoyl_CoA_reductase  5_methylaminomethyl_2_thiouridine  ...  \\\n",
       "0                               0.0                        4136.703664  ...   \n",
       "1                               0.0                        3913.676651  ...   \n",
       "2                               0.0                        2093.412992  ...   \n",
       "3                               0.0                        2050.494687  ...   \n",
       "4                               0.0                        1570.967988  ...   \n",
       "..                              ...                                ...  ...   \n",
       "103                             0.0                        4232.705389  ...   \n",
       "104                             0.0                        4210.611072  ...   \n",
       "105                             0.0                        3424.824904  ...   \n",
       "106                             0.0                        3243.276243  ...   \n",
       "107                             0.0                        2574.961898  ...   \n",
       "\n",
       "     DNA_Conc_ng_ul_  Index_I7  Index_I5  Mean_Size_BP  \\\n",
       "0             42.800       0.0       0.0           417   \n",
       "1             32.600       0.0       0.0           498   \n",
       "2             33.400       0.0       0.0           535   \n",
       "3             30.800       0.0       0.0           352   \n",
       "4             27.400       0.0       0.0           500   \n",
       "..               ...       ...       ...           ...   \n",
       "103            0.332       0.0       0.0           383   \n",
       "104           21.200       0.0       0.0           349   \n",
       "105           16.900       0.0       0.0           584   \n",
       "106           34.600       0.0       0.0           364   \n",
       "107           21.600       0.0       0.0           469   \n",
       "\n",
       "     Total_Clusters_Passing_Filter_Million_            PC1            PC2  \\\n",
       "0                                       2.8  105232.294537   29481.534204   \n",
       "1                                       4.7  -68177.360588 -173244.331191   \n",
       "2                                       4.5 -211025.239559 -151897.294836   \n",
       "3                                       4.8 -188035.291051 -141834.133045   \n",
       "4                                       3.8 -231769.880818 -155931.482253   \n",
       "..                                      ...            ...            ...   \n",
       "103                                     0.2  433871.879019  -50971.981671   \n",
       "104                                     5.1 -242787.205324 -340573.238777   \n",
       "105                                     4.6   95694.320998   22992.264925   \n",
       "106                                     2.2   -4008.324996   95743.541072   \n",
       "107                                     4.4  -47651.404957  -78848.180154   \n",
       "\n",
       "               PC3            PC4           PC5  \n",
       "0      3009.236199    -748.874391   1393.844533  \n",
       "1    150955.115109   58275.741094  15805.072273  \n",
       "2   -320966.178592  262834.328944  74715.555125  \n",
       "3   -266846.807833  226853.551490  69220.178950  \n",
       "4   -328612.917308  256193.671691  79085.256264  \n",
       "..             ...            ...           ...  \n",
       "103   -5320.962053   24773.058456  24120.785059  \n",
       "104  401656.123059  192192.110121  -6890.326181  \n",
       "105   -6039.669020   -3148.567876  13543.206780  \n",
       "106   17895.588170    2995.813400  23541.421539  \n",
       "107 -122746.333045   51511.251975  32122.109092  \n",
       "\n",
       "[108 rows x 4522 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "\n",
    "# Apply PCA to reduce bacterial abundance columns to 5 components\n",
    "pca = PCA(n_components=5)\n",
    "pca_components = pca.fit_transform(df_combined[df.columns])  # Add bacterial abundance columns\n",
    "pca_df = pd.DataFrame(pca_components, columns=[\"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\"])\n",
    "\n",
    "# Add PCA components to the combined DataFrame\n",
    "df_combined2 = pd.concat([df_combined.reset_index(drop=True), pca_df], axis=1)\n",
    "df_combined2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ad01662a-ccea-4d6b-951e-5fa715c8ffed",
   "metadata": {},
   "outputs": [
    {
     "ename": "PatsyError",
     "evalue": "Error evaluating factor: NameError: name 'N12M_Pseudomonas_aeruginosa' is not defined\n    N12M_Pseudomonas_aeruginosa ~ Pedobacter + CS_Pseudomonas_aeruginosa + Pseudomonas_Culture\n    ^^^^^^^^^^^^^^^^^^^^^^^^^^^",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/compat.py:36\u001b[0m, in \u001b[0;36mcall_and_wrap_exc\u001b[0;34m(msg, origin, f, *args, **kwargs)\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 36\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     37\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/eval.py:165\u001b[0m, in \u001b[0;36mEvalEnvironment.eval\u001b[0;34m(self, expr, source_name, inner_namespace)\u001b[0m\n\u001b[1;32m    164\u001b[0m code \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mcompile\u001b[39m(expr, source_name, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124meval\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mflags, \u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m--> 165\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43meval\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mcode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mVarLookupDict\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43minner_namespace\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    166\u001b[0m \u001b[43m                                    \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_namespaces\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m<string>:1\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'N12M_Pseudomonas_aeruginosa' is not defined",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mPatsyError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[57], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m formula \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mN12M_Pseudomonas_aeruginosa ~ Pedobacter + CS_Pseudomonas_aeruginosa + Pseudomonas_Culture\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# Fit the GLMM\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m model \u001b[38;5;241m=\u001b[39m \u001b[43msmf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmixedlm\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mformula\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mformula\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdf_combined3\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroups\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdf_combined3\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpwCF_ID\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# Grouping by pwCF_ID for random effects\u001b[39;49;00m\n\u001b[1;32m      7\u001b[0m \u001b[43m)\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;66;03m#family=sm.families.Binomial()  # Logistic link function\u001b[39;00m\n\u001b[1;32m      9\u001b[0m result \u001b[38;5;241m=\u001b[39m model\u001b[38;5;241m.\u001b[39mfit()\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/regression/mixed_linear_model.py:1046\u001b[0m, in \u001b[0;36mMixedLM.from_formula\u001b[0;34m(cls, formula, data, re_formula, vc_formula, subset, use_sparse, missing, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1044\u001b[0m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexog_vc\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m exog_vc\n\u001b[1;32m   1045\u001b[0m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgroups\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m groups\n\u001b[0;32m-> 1046\u001b[0m mod \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mMixedLM\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_formula\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1047\u001b[0m \u001b[43m    \u001b[49m\u001b[43mformula\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1049\u001b[0m \u001b[38;5;66;03m# expand re names to account for pairs of RE\u001b[39;00m\n\u001b[1;32m   1050\u001b[0m (param_names,\n\u001b[1;32m   1051\u001b[0m  exog_re_names,\n\u001b[1;32m   1052\u001b[0m  exog_re_names_full) \u001b[38;5;241m=\u001b[39m mod\u001b[38;5;241m.\u001b[39m_make_param_names(exog_re_names)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/model.py:200\u001b[0m, in \u001b[0;36mModel.from_formula\u001b[0;34m(cls, formula, data, subset, drop_cols, *args, **kwargs)\u001b[0m\n\u001b[1;32m    197\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m missing \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m:  \u001b[38;5;66;03m# with patsy it's drop or raise. let's raise.\u001b[39;00m\n\u001b[1;32m    198\u001b[0m     missing \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mraise\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m--> 200\u001b[0m tmp \u001b[38;5;241m=\u001b[39m \u001b[43mhandle_formula_data\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mformula\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdepth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43meval_env\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    201\u001b[0m \u001b[43m                          \u001b[49m\u001b[43mmissing\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmissing\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    202\u001b[0m ((endog, exog), missing_idx, design_info) \u001b[38;5;241m=\u001b[39m tmp\n\u001b[1;32m    203\u001b[0m max_endog \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m_formula_max_endog\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/formula/formulatools.py:63\u001b[0m, in \u001b[0;36mhandle_formula_data\u001b[0;34m(Y, X, formula, depth, missing)\u001b[0m\n\u001b[1;32m     61\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     62\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m data_util\u001b[38;5;241m.\u001b[39m_is_using_pandas(Y, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m---> 63\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[43mdmatrices\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformula\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mY\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdepth\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreturn_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdataframe\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     64\u001b[0m \u001b[43m                           \u001b[49m\u001b[43mNA_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_action\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     65\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     66\u001b[0m         result \u001b[38;5;241m=\u001b[39m dmatrices(formula, Y, depth, return_type\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdataframe\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     67\u001b[0m                            NA_action\u001b[38;5;241m=\u001b[39mna_action)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/highlevel.py:309\u001b[0m, in \u001b[0;36mdmatrices\u001b[0;34m(formula_like, data, eval_env, NA_action, return_type)\u001b[0m\n\u001b[1;32m    299\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Construct two design matrices given a formula_like and data.\u001b[39;00m\n\u001b[1;32m    300\u001b[0m \n\u001b[1;32m    301\u001b[0m \u001b[38;5;124;03mThis function is identical to :func:`dmatrix`, except that it requires\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    306\u001b[0m \u001b[38;5;124;03mSee :func:`dmatrix` for details.\u001b[39;00m\n\u001b[1;32m    307\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    308\u001b[0m eval_env \u001b[38;5;241m=\u001b[39m EvalEnvironment\u001b[38;5;241m.\u001b[39mcapture(eval_env, reference\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[0;32m--> 309\u001b[0m (lhs, rhs) \u001b[38;5;241m=\u001b[39m \u001b[43m_do_highlevel_design\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformula_like\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43meval_env\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    310\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43mNA_action\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreturn_type\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    311\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m lhs\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    312\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PatsyError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodel is missing required outcome variables\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/highlevel.py:164\u001b[0m, in \u001b[0;36m_do_highlevel_design\u001b[0;34m(formula_like, data, eval_env, NA_action, return_type)\u001b[0m\n\u001b[1;32m    162\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdata_iter_maker\u001b[39m():\n\u001b[1;32m    163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28miter\u001b[39m([data])\n\u001b[0;32m--> 164\u001b[0m design_infos \u001b[38;5;241m=\u001b[39m \u001b[43m_try_incr_builders\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformula_like\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata_iter_maker\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43meval_env\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    165\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43mNA_action\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    166\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m design_infos \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    167\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m build_design_matrices(design_infos, data,\n\u001b[1;32m    168\u001b[0m                                  NA_action\u001b[38;5;241m=\u001b[39mNA_action,\n\u001b[1;32m    169\u001b[0m                                  return_type\u001b[38;5;241m=\u001b[39mreturn_type)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/highlevel.py:66\u001b[0m, in \u001b[0;36m_try_incr_builders\u001b[0;34m(formula_like, data_iter_maker, eval_env, NA_action)\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(formula_like, ModelDesc):\n\u001b[1;32m     65\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(eval_env, EvalEnvironment)\n\u001b[0;32m---> 66\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdesign_matrix_builders\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mformula_like\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlhs_termlist\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     67\u001b[0m \u001b[43m                                   \u001b[49m\u001b[43mformula_like\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrhs_termlist\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     68\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43mdata_iter_maker\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     69\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43meval_env\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     70\u001b[0m \u001b[43m                                  \u001b[49m\u001b[43mNA_action\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     72\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/build.py:693\u001b[0m, in \u001b[0;36mdesign_matrix_builders\u001b[0;34m(termlists, data_iter_maker, eval_env, NA_action)\u001b[0m\n\u001b[1;32m    689\u001b[0m factor_states \u001b[38;5;241m=\u001b[39m _factors_memorize(all_factors, data_iter_maker, eval_env)\n\u001b[1;32m    690\u001b[0m \u001b[38;5;66;03m# Now all the factors have working eval methods, so we can evaluate them\u001b[39;00m\n\u001b[1;32m    691\u001b[0m \u001b[38;5;66;03m# on some data to find out what type of data they return.\u001b[39;00m\n\u001b[1;32m    692\u001b[0m (num_column_counts,\n\u001b[0;32m--> 693\u001b[0m  cat_levels_contrasts) \u001b[38;5;241m=\u001b[39m \u001b[43m_examine_factor_types\u001b[49m\u001b[43m(\u001b[49m\u001b[43mall_factors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    694\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mfactor_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    695\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mdata_iter_maker\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    696\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mNA_action\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    697\u001b[0m \u001b[38;5;66;03m# Now we need the factor infos, which encapsulate the knowledge of\u001b[39;00m\n\u001b[1;32m    698\u001b[0m \u001b[38;5;66;03m# how to turn any given factor into a chunk of data:\u001b[39;00m\n\u001b[1;32m    699\u001b[0m factor_infos \u001b[38;5;241m=\u001b[39m {}\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/build.py:443\u001b[0m, in \u001b[0;36m_examine_factor_types\u001b[0;34m(factors, factor_states, data_iter_maker, NA_action)\u001b[0m\n\u001b[1;32m    441\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m data \u001b[38;5;129;01min\u001b[39;00m data_iter_maker():\n\u001b[1;32m    442\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m factor \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mlist\u001b[39m(examine_needed):\n\u001b[0;32m--> 443\u001b[0m         value \u001b[38;5;241m=\u001b[39m \u001b[43mfactor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfactor_states\u001b[49m\u001b[43m[\u001b[49m\u001b[43mfactor\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    444\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m factor \u001b[38;5;129;01min\u001b[39;00m cat_sniffers \u001b[38;5;129;01mor\u001b[39;00m guess_categorical(value):\n\u001b[1;32m    445\u001b[0m             \u001b[38;5;28;01mif\u001b[39;00m factor \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m cat_sniffers:\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/eval.py:564\u001b[0m, in \u001b[0;36mEvalFactor.eval\u001b[0;34m(self, memorize_state, data)\u001b[0m\n\u001b[1;32m    563\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21meval\u001b[39m(\u001b[38;5;28mself\u001b[39m, memorize_state, data):\n\u001b[0;32m--> 564\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_eval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmemorize_state\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43meval_code\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    565\u001b[0m \u001b[43m                      \u001b[49m\u001b[43mmemorize_state\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    566\u001b[0m \u001b[43m                      \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/eval.py:547\u001b[0m, in \u001b[0;36mEvalFactor._eval\u001b[0;34m(self, code, memorize_state, data)\u001b[0m\n\u001b[1;32m    545\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_eval\u001b[39m(\u001b[38;5;28mself\u001b[39m, code, memorize_state, data):\n\u001b[1;32m    546\u001b[0m     inner_namespace \u001b[38;5;241m=\u001b[39m VarLookupDict([data, memorize_state[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtransforms\u001b[39m\u001b[38;5;124m\"\u001b[39m]])\n\u001b[0;32m--> 547\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mcall_and_wrap_exc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mError evaluating factor\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    548\u001b[0m \u001b[43m                             \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    549\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mmemorize_state\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43meval_env\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    550\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mcode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    551\u001b[0m \u001b[43m                             \u001b[49m\u001b[43minner_namespace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minner_namespace\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/patsy/compat.py:43\u001b[0m, in \u001b[0;36mcall_and_wrap_exc\u001b[0;34m(msg, origin, f, *args, **kwargs)\u001b[0m\n\u001b[1;32m     39\u001b[0m     new_exc \u001b[38;5;241m=\u001b[39m PatsyError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     40\u001b[0m                          \u001b[38;5;241m%\u001b[39m (msg, e\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m, e),\n\u001b[1;32m     41\u001b[0m                          origin)\n\u001b[1;32m     42\u001b[0m     \u001b[38;5;66;03m# Use 'exec' to hide this syntax from the Python 2 parser:\u001b[39;00m\n\u001b[0;32m---> 43\u001b[0m     \u001b[43mexec\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mraise new_exc from e\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     44\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     45\u001b[0m     \u001b[38;5;66;03m# In python 2, we just let the original exception escape -- better\u001b[39;00m\n\u001b[1;32m     46\u001b[0m     \u001b[38;5;66;03m# than destroying the traceback. But if it's a PatsyError, we can\u001b[39;00m\n\u001b[1;32m     47\u001b[0m     \u001b[38;5;66;03m# at least set the origin properly.\u001b[39;00m\n\u001b[1;32m     48\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e, PatsyError):\n",
      "File \u001b[0;32m<string>:1\u001b[0m\n",
      "\u001b[0;31mPatsyError\u001b[0m: Error evaluating factor: NameError: name 'N12M_Pseudomonas_aeruginosa' is not defined\n    N12M_Pseudomonas_aeruginosa ~ Pedobacter + CS_Pseudomonas_aeruginosa + Pseudomonas_Culture\n    ^^^^^^^^^^^^^^^^^^^^^^^^^^^"
     ]
    }
   ],
   "source": [
    "formula = \"N12M_Pseudomonas_aeruginosa ~ Pedobacter + CS_Pseudomonas_aeruginosa + Pseudomonas_Culture\"\n",
    "# Fit the GLMM\n",
    "model = smf.mixedlm(\n",
    "    formula=formula,\n",
    "    data=df_combined3,\n",
    "    groups=df_combined3[\"pwCF_ID\"],  # Grouping by pwCF_ID for random effects\n",
    ")\n",
    "#family=sm.families.Binomial()  # Logistic link function\n",
    "result = model.fit()\n",
    "\n",
    "# Print the model summary\n",
    "print(result.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "556ea72b-a5dc-4b8d-b56d-5047dee8bd16",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "684020ba-8b14-4179-89ae-22401744b494",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "04378ee0-b921-42dc-b5ec-315d6d775f7a",
   "metadata": {},
   "outputs": [
    {
     "ename": "LinAlgError",
     "evalue": "Singular matrix",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLinAlgError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[40], line 9\u001b[0m\n\u001b[1;32m      3\u001b[0m model \u001b[38;5;241m=\u001b[39m smf\u001b[38;5;241m.\u001b[39mmixedlm(\n\u001b[1;32m      4\u001b[0m     formula\u001b[38;5;241m=\u001b[39mformula,\n\u001b[1;32m      5\u001b[0m     data\u001b[38;5;241m=\u001b[39mdf_combined,\n\u001b[1;32m      6\u001b[0m     groups\u001b[38;5;241m=\u001b[39mdf_combined[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpwCF_ID\u001b[39m\u001b[38;5;124m\"\u001b[39m],  \u001b[38;5;66;03m# Grouping by pwCF_ID for random effects\u001b[39;00m\n\u001b[1;32m      7\u001b[0m )\n\u001b[1;32m      8\u001b[0m \u001b[38;5;66;03m#family=sm.families.Binomial()  # Logistic link function\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;66;03m# Print the model summary\u001b[39;00m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28mprint\u001b[39m(result\u001b[38;5;241m.\u001b[39msummary())\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/regression/mixed_linear_model.py:2191\u001b[0m, in \u001b[0;36mMixedLM.fit\u001b[0;34m(self, start_params, reml, niter_sa, do_cg, fe_pen, cov_pen, free, full_output, method, **fit_kwargs)\u001b[0m\n\u001b[1;32m   2189\u001b[0m \u001b[38;5;66;03m# Try optimizing one or more times\u001b[39;00m\n\u001b[1;32m   2190\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m j \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(method)):\n\u001b[0;32m-> 2191\u001b[0m     rslt \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mMixedLM\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstart_params\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpacked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2192\u001b[0m \u001b[43m                                    \u001b[49m\u001b[43mskip_hessian\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   2193\u001b[0m \u001b[43m                                    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m[\u001b[49m\u001b[43mj\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2194\u001b[0m \u001b[43m                                    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mfit_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2195\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m rslt\u001b[38;5;241m.\u001b[39mmle_retvals[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mconverged\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n\u001b[1;32m   2196\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/model.py:563\u001b[0m, in \u001b[0;36mLikelihoodModel.fit\u001b[0;34m(self, start_params, method, maxiter, full_output, disp, fargs, callback, retall, skip_hessian, **kwargs)\u001b[0m\n\u001b[1;32m    560\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muse_t\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    562\u001b[0m optimizer \u001b[38;5;241m=\u001b[39m Optimizer()\n\u001b[0;32m--> 563\u001b[0m xopt, retvals, optim_settings \u001b[38;5;241m=\u001b[39m \u001b[43moptimizer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscore\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart_params\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    564\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mfargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    565\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mhessian\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhess\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    566\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    567\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mdisp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdisp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    568\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mmaxiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaxiter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    569\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mcallback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    570\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mretall\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretall\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    571\u001b[0m \u001b[43m                                               \u001b[49m\u001b[43mfull_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfull_output\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    572\u001b[0m \u001b[38;5;66;03m# Restore cov_type, cov_kwds and use_t\u001b[39;00m\n\u001b[1;32m    573\u001b[0m optim_settings\u001b[38;5;241m.\u001b[39mupdate(kwds)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/optimizer.py:241\u001b[0m, in \u001b[0;36mOptimizer._fit\u001b[0;34m(self, objective, gradient, start_params, fargs, kwargs, hessian, method, maxiter, full_output, disp, callback, retall)\u001b[0m\n\u001b[1;32m    238\u001b[0m     fit_funcs\u001b[38;5;241m.\u001b[39mupdate(extra_fit_funcs)\n\u001b[1;32m    240\u001b[0m func \u001b[38;5;241m=\u001b[39m fit_funcs[method]\n\u001b[0;32m--> 241\u001b[0m xopt, retvals \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobjective\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgradient\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart_params\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    242\u001b[0m \u001b[43m                     \u001b[49m\u001b[43mdisp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdisp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmaxiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaxiter\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    243\u001b[0m \u001b[43m                     \u001b[49m\u001b[43mretall\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretall\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfull_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfull_output\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m                     \u001b[49m\u001b[43mhess\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhessian\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    246\u001b[0m optim_settings \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124moptimizer\u001b[39m\u001b[38;5;124m'\u001b[39m: method, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstart_params\u001b[39m\u001b[38;5;124m'\u001b[39m: start_params,\n\u001b[1;32m    247\u001b[0m                   \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmaxiter\u001b[39m\u001b[38;5;124m'\u001b[39m: maxiter, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfull_output\u001b[39m\u001b[38;5;124m'\u001b[39m: full_output,\n\u001b[1;32m    248\u001b[0m                   \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdisp\u001b[39m\u001b[38;5;124m'\u001b[39m: disp, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfargs\u001b[39m\u001b[38;5;124m'\u001b[39m: fargs, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcallback\u001b[39m\u001b[38;5;124m'\u001b[39m: callback,\n\u001b[1;32m    249\u001b[0m                   \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mretall\u001b[39m\u001b[38;5;124m'\u001b[39m: retall, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mextra_fit_funcs\u001b[39m\u001b[38;5;124m\"\u001b[39m: extra_fit_funcs}\n\u001b[1;32m    250\u001b[0m optim_settings\u001b[38;5;241m.\u001b[39mupdate(kwargs)\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/optimizer.py:529\u001b[0m, in \u001b[0;36m_fit_bfgs\u001b[0;34m(f, score, start_params, fargs, kwargs, disp, maxiter, callback, retall, full_output, hess)\u001b[0m\n\u001b[1;32m    527\u001b[0m norm \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnorm\u001b[39m\u001b[38;5;124m'\u001b[39m, np\u001b[38;5;241m.\u001b[39mInf)\n\u001b[1;32m    528\u001b[0m epsilon \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mepsilon\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;241m1.4901161193847656e-08\u001b[39m)\n\u001b[0;32m--> 529\u001b[0m retvals \u001b[38;5;241m=\u001b[39m \u001b[43moptimize\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfmin_bfgs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart_params\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscore\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    530\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mgtol\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgtol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnorm\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnorm\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mepsilon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mepsilon\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    531\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mmaxiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaxiter\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfull_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfull_output\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    532\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mdisp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdisp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mretall\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretall\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    533\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m full_output:\n\u001b[1;32m    534\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m retall:\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/optimize.py:1143\u001b[0m, in \u001b[0;36mfmin_bfgs\u001b[0;34m(f, x0, fprime, args, gtol, norm, epsilon, maxiter, full_output, disp, retall, callback)\u001b[0m\n\u001b[1;32m   1037\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1038\u001b[0m \u001b[38;5;124;03mMinimize a function using the BFGS algorithm.\u001b[39;00m\n\u001b[1;32m   1039\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1134\u001b[0m \n\u001b[1;32m   1135\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1136\u001b[0m opts \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgtol\u001b[39m\u001b[38;5;124m'\u001b[39m: gtol,\n\u001b[1;32m   1137\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnorm\u001b[39m\u001b[38;5;124m'\u001b[39m: norm,\n\u001b[1;32m   1138\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124meps\u001b[39m\u001b[38;5;124m'\u001b[39m: epsilon,\n\u001b[1;32m   1139\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdisp\u001b[39m\u001b[38;5;124m'\u001b[39m: disp,\n\u001b[1;32m   1140\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmaxiter\u001b[39m\u001b[38;5;124m'\u001b[39m: maxiter,\n\u001b[1;32m   1141\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreturn_all\u001b[39m\u001b[38;5;124m'\u001b[39m: retall}\n\u001b[0;32m-> 1143\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43m_minimize_bfgs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mx0\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfprime\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallback\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mopts\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1145\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m full_output:\n\u001b[1;32m   1146\u001b[0m     retlist \u001b[38;5;241m=\u001b[39m (res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mx\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfun\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mjac\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhess_inv\u001b[39m\u001b[38;5;124m'\u001b[39m],\n\u001b[1;32m   1147\u001b[0m                res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnfev\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnjev\u001b[39m\u001b[38;5;124m'\u001b[39m], res[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstatus\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/optimize.py:1201\u001b[0m, in \u001b[0;36m_minimize_bfgs\u001b[0;34m(fun, x0, args, jac, callback, gtol, norm, eps, maxiter, disp, return_all, finite_diff_rel_step, **unknown_options)\u001b[0m\n\u001b[1;32m   1198\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m maxiter \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1199\u001b[0m     maxiter \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(x0) \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m200\u001b[39m\n\u001b[0;32m-> 1201\u001b[0m sf \u001b[38;5;241m=\u001b[39m \u001b[43m_prepare_scalar_function\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfun\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mx0\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mjac\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mepsilon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43meps\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1202\u001b[0m \u001b[43m                              \u001b[49m\u001b[43mfinite_diff_rel_step\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfinite_diff_rel_step\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1204\u001b[0m f \u001b[38;5;241m=\u001b[39m sf\u001b[38;5;241m.\u001b[39mfun\n\u001b[1;32m   1205\u001b[0m myfprime \u001b[38;5;241m=\u001b[39m sf\u001b[38;5;241m.\u001b[39mgrad\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/optimize.py:261\u001b[0m, in \u001b[0;36m_prepare_scalar_function\u001b[0;34m(fun, x0, jac, args, bounds, epsilon, finite_diff_rel_step, hess)\u001b[0m\n\u001b[1;32m    257\u001b[0m     bounds \u001b[38;5;241m=\u001b[39m (\u001b[38;5;241m-\u001b[39mnp\u001b[38;5;241m.\u001b[39minf, np\u001b[38;5;241m.\u001b[39minf)\n\u001b[1;32m    259\u001b[0m \u001b[38;5;66;03m# ScalarFunction caches. Reuse of fun(x) during grad\u001b[39;00m\n\u001b[1;32m    260\u001b[0m \u001b[38;5;66;03m# calculation reduces overall function evaluations.\u001b[39;00m\n\u001b[0;32m--> 261\u001b[0m sf \u001b[38;5;241m=\u001b[39m \u001b[43mScalarFunction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfun\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mx0\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgrad\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhess\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    262\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mfinite_diff_rel_step\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbounds\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mepsilon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mepsilon\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    264\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m sf\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/_differentiable_functions.py:140\u001b[0m, in \u001b[0;36mScalarFunction.__init__\u001b[0;34m(self, fun, x0, args, grad, hess, finite_diff_rel_step, finite_diff_bounds, epsilon)\u001b[0m\n\u001b[1;32m    137\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf \u001b[38;5;241m=\u001b[39m fun_wrapped(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mx)\n\u001b[1;32m    139\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_update_fun_impl \u001b[38;5;241m=\u001b[39m update_fun\n\u001b[0;32m--> 140\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_update_fun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    142\u001b[0m \u001b[38;5;66;03m# Gradient evaluation\u001b[39;00m\n\u001b[1;32m    143\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mcallable\u001b[39m(grad):\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/_differentiable_functions.py:233\u001b[0m, in \u001b[0;36mScalarFunction._update_fun\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    231\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_update_fun\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    232\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf_updated:\n\u001b[0;32m--> 233\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_update_fun_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    234\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf_updated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/_differentiable_functions.py:137\u001b[0m, in \u001b[0;36mScalarFunction.__init__.<locals>.update_fun\u001b[0;34m()\u001b[0m\n\u001b[1;32m    136\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mupdate_fun\u001b[39m():\n\u001b[0;32m--> 137\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf \u001b[38;5;241m=\u001b[39m \u001b[43mfun_wrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/scipy/optimize/_differentiable_functions.py:134\u001b[0m, in \u001b[0;36mScalarFunction.__init__.<locals>.fun_wrapped\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m    130\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnfev \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    131\u001b[0m \u001b[38;5;66;03m# Send a copy because the user may overwrite it.\u001b[39;00m\n\u001b[1;32m    132\u001b[0m \u001b[38;5;66;03m# Overwriting results in undefined behaviour because\u001b[39;00m\n\u001b[1;32m    133\u001b[0m \u001b[38;5;66;03m# fun(self.x) will change self.x, with the two no longer linked.\u001b[39;00m\n\u001b[0;32m--> 134\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/base/model.py:531\u001b[0m, in \u001b[0;36mLikelihoodModel.fit.<locals>.f\u001b[0;34m(params, *args)\u001b[0m\n\u001b[1;32m    530\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m(params, \u001b[38;5;241m*\u001b[39margs):\n\u001b[0;32m--> 531\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloglike\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;241m/\u001b[39m nobs\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/regression/mixed_linear_model.py:1498\u001b[0m, in \u001b[0;36mMixedLM.loglike\u001b[0;34m(self, params, profile_fe)\u001b[0m\n\u001b[1;32m   1496\u001b[0m \u001b[38;5;66;03m# Move to the profile set\u001b[39;00m\n\u001b[1;32m   1497\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m profile_fe:\n\u001b[0;32m-> 1498\u001b[0m     fe_params, sing \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_fe_params\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcov_re\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvcomp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1499\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m sing:\n\u001b[1;32m   1500\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cov_sing \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/bioinformatics/lib/python3.9/site-packages/statsmodels/regression/mixed_linear_model.py:1345\u001b[0m, in \u001b[0;36mMixedLM.get_fe_params\u001b[0;34m(self, cov_re, vcomp, tol)\u001b[0m\n\u001b[1;32m   1343\u001b[0m     fe_params \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mdot(np\u001b[38;5;241m.\u001b[39mlinalg\u001b[38;5;241m.\u001b[39mpinv(xtxy[:, \u001b[38;5;241m0\u001b[39m:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]), xtxy[:, \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m])\n\u001b[1;32m   1344\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1345\u001b[0m     fe_params \u001b[38;5;241m=\u001b[39m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlinalg\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msolve\u001b[49m\u001b[43m(\u001b[49m\u001b[43mxtxy\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m:\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mxtxy\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1347\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m fe_params, sing\n",
      "File \u001b[0;32m<__array_function__ internals>:180\u001b[0m, in \u001b[0;36msolve\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/numpy/linalg/linalg.py:393\u001b[0m, in \u001b[0;36msolve\u001b[0;34m(a, b)\u001b[0m\n\u001b[1;32m    391\u001b[0m signature \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDD->D\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m isComplexType(t) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdd->d\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    392\u001b[0m extobj \u001b[38;5;241m=\u001b[39m get_linalg_error_extobj(_raise_linalgerror_singular)\n\u001b[0;32m--> 393\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43mgufunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mb\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msignature\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msignature\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mextobj\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextobj\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    395\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m wrap(r\u001b[38;5;241m.\u001b[39mastype(result_t, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m))\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/numpy/linalg/linalg.py:88\u001b[0m, in \u001b[0;36m_raise_linalgerror_singular\u001b[0;34m(err, flag)\u001b[0m\n\u001b[1;32m     87\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_raise_linalgerror_singular\u001b[39m(err, flag):\n\u001b[0;32m---> 88\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m LinAlgError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSingular matrix\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mLinAlgError\u001b[0m: Singular matrix"
     ]
    }
   ],
   "source": [
    "df_combined = df_combined.dropna(subset=all_columns)\n",
    "# Fit the GLMM\n",
    "model = smf.mixedlm(\n",
    "    formula=formula,\n",
    "    data=df_combined,\n",
    "    groups=df_combined[\"pwCF_ID\"],  # Grouping by pwCF_ID for random effects\n",
    ")\n",
    "#family=sm.families.Binomial()  # Logistic link function\n",
    "result = model.fit()\n",
    "\n",
    "# Print the model summary\n",
    "print(result.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "d3871d7a-02db-4c2f-93cc-c235324df847",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NAME\n",
       "634207_20180510_S     0.0\n",
       "634207_20180517_S     0.0\n",
       "715927_20180205_S     0.0\n",
       "715927_20180213_S     0.0\n",
       "715927_20180226_S     0.0\n",
       "                     ... \n",
       "1651490_20180206_S    1.0\n",
       "1651490_20171215_S    1.0\n",
       "1658447_20171006_S    1.0\n",
       "1664053_20180406_S    1.0\n",
       "1690154_20180406_S    1.0\n",
       "Name: N12M_Pseudomonas_aeruginosa, Length: 127, dtype: category\n",
       "Categories (2, float64): [0.0, 1.0]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[\"N12M_Pseudomonas_aeruginosa\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "38a70d41-d2c1-4dbd-935b-64e24e4c3938",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NAME\n",
       "634207_20180510_S     0\n",
       "634207_20180517_S     0\n",
       "715927_20180205_S     0\n",
       "715927_20180213_S     0\n",
       "715927_20180226_S     0\n",
       "                     ..\n",
       "1651490_20180206_S    1\n",
       "1651490_20171215_S    1\n",
       "1658447_20171006_S    1\n",
       "1664053_20180406_S    1\n",
       "1690154_20180406_S    1\n",
       "Name: N12M_Pseudomonas_aeruginosa, Length: 127, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[\"N12M_Pseudomonas_aeruginosa\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72c95614-d245-4f5a-971c-5c3e1f5004bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6919d202-c096-403f-bb39-025e8cef7a1f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f87f6a19-2d73-4881-b3cd-6afecdb67ec4",
   "metadata": {},
   "source": [
    "# Early trial\n",
    "\n",
    "Ignore the stuff down here, its probably wrong :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61367d40-baba-41fc-90af-b67e61fcfb1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minion</th>\n",
       "      <th>MGI</th>\n",
       "      <th>pwCF_ID</th>\n",
       "      <th>Sample date</th>\n",
       "      <th>IP vs OP</th>\n",
       "      <th>Hospital</th>\n",
       "      <th>Room</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age groups</th>\n",
       "      <th>Paediatric vs Adult</th>\n",
       "      <th>...</th>\n",
       "      <th>Sum of meds</th>\n",
       "      <th>Sum of antifungals</th>\n",
       "      <th>Sum of steroid + mabs</th>\n",
       "      <th>DNA_extraction_ conc</th>\n",
       "      <th>SAGC ULN</th>\n",
       "      <th>DNA Conc. (ng/ul)</th>\n",
       "      <th>Index I7</th>\n",
       "      <th>Index I5</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total Clusters Passing Filter (Million)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>634207_20180510_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>634207_20180510_S</td>\n",
       "      <td>634207</td>\n",
       "      <td>5/10/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Rm9</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>SAGCFN_22_01856</td>\n",
       "      <td>7.82</td>\n",
       "      <td>CGGACGATTC</td>\n",
       "      <td>CCACCACCTA</td>\n",
       "      <td>651</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634207_20180517_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>634207_20180517_S</td>\n",
       "      <td>634207</td>\n",
       "      <td>5/17/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Rm9</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.134</td>\n",
       "      <td>SAGCFN_22_01827</td>\n",
       "      <td>22.80</td>\n",
       "      <td>AGCGATAG</td>\n",
       "      <td>CCTATCCT</td>\n",
       "      <td>633</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715927_20180205_S</th>\n",
       "      <td>715927_20180205_S</td>\n",
       "      <td>715927_20180205_S</td>\n",
       "      <td>715927</td>\n",
       "      <td>2/05/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Level 6 DK Office</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.326</td>\n",
       "      <td>SAGCFN_22_01797</td>\n",
       "      <td>16.50</td>\n",
       "      <td>TAATGCGC</td>\n",
       "      <td>AGGCGAAG</td>\n",
       "      <td>516</td>\n",
       "      <td>3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715927_20180213_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>715927_20180213_S</td>\n",
       "      <td>715927</td>\n",
       "      <td>2/13/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Room 11</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.234</td>\n",
       "      <td>SAGCFN_22_01811</td>\n",
       "      <td>31.00</td>\n",
       "      <td>TCCGCGAA</td>\n",
       "      <td>CCTATCCT</td>\n",
       "      <td>443</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715927_20180226_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>715927_20180226_S</td>\n",
       "      <td>715927</td>\n",
       "      <td>2/26/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>OPD 8</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.108</td>\n",
       "      <td>SAGCFN_22_01833</td>\n",
       "      <td>15.10</td>\n",
       "      <td>TAACTTGGTC</td>\n",
       "      <td>GATTCACGAC</td>\n",
       "      <td>510</td>\n",
       "      <td>2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651490_20180206_S</th>\n",
       "      <td>1651490_20180206_S</td>\n",
       "      <td>1651490_20180206_S</td>\n",
       "      <td>1651490</td>\n",
       "      <td>2/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 1</td>\n",
       "      <td>27</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.760</td>\n",
       "      <td>SAGCFN_22_01741</td>\n",
       "      <td>26.20</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>AGGCGAAG</td>\n",
       "      <td>507</td>\n",
       "      <td>6.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651490_20171215_S</th>\n",
       "      <td>1651490_20171215_S</td>\n",
       "      <td>1651490_20171215_S</td>\n",
       "      <td>1651490</td>\n",
       "      <td>12/15/2017</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 4</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7.760</td>\n",
       "      <td>SAGCFN_22_01738</td>\n",
       "      <td>34.20</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>564</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658447_20171006_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1658447_20171006_S</td>\n",
       "      <td>1658447</td>\n",
       "      <td>10/06/2017</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 3</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.098</td>\n",
       "      <td>SAGCFN_22_01837</td>\n",
       "      <td>13.10</td>\n",
       "      <td>CAGCAGGTCA</td>\n",
       "      <td>TACCTAAGTG</td>\n",
       "      <td>576</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664053_20180406_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1664053_20180406_S</td>\n",
       "      <td>1664053</td>\n",
       "      <td>4/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 1</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.159</td>\n",
       "      <td>SAGCFN_22_01822</td>\n",
       "      <td>30.80</td>\n",
       "      <td>TCTCGCGC</td>\n",
       "      <td>TAATCTTA</td>\n",
       "      <td>374</td>\n",
       "      <td>1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1690154_20180406_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1690154_20180406_S</td>\n",
       "      <td>1690154</td>\n",
       "      <td>4/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 4</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.276</td>\n",
       "      <td>SAGCFN_22_01804</td>\n",
       "      <td>35.00</td>\n",
       "      <td>CGGCTATG</td>\n",
       "      <td>GGCTCTGA</td>\n",
       "      <td>648</td>\n",
       "      <td>3.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows Ã— 166 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                minion                 MGI  pwCF_ID  \\\n",
       "NAME                                                                  \n",
       "634207_20180510_S                  NaN   634207_20180510_S   634207   \n",
       "634207_20180517_S                  NaN   634207_20180517_S   634207   \n",
       "715927_20180205_S    715927_20180205_S   715927_20180205_S   715927   \n",
       "715927_20180213_S                  NaN   715927_20180213_S   715927   \n",
       "715927_20180226_S                  NaN   715927_20180226_S   715927   \n",
       "...                                ...                 ...      ...   \n",
       "1651490_20180206_S  1651490_20180206_S  1651490_20180206_S  1651490   \n",
       "1651490_20171215_S  1651490_20171215_S  1651490_20171215_S  1651490   \n",
       "1658447_20171006_S                 NaN  1658447_20171006_S  1658447   \n",
       "1664053_20180406_S                 NaN  1664053_20180406_S  1664053   \n",
       "1690154_20180406_S                 NaN  1690154_20180406_S  1690154   \n",
       "\n",
       "                   Sample date IP vs OP Hospital               Room  Age  \\\n",
       "NAME                                                                       \n",
       "634207_20180510_S    5/10/2018       IP      WCH           Adol Rm9   17   \n",
       "634207_20180517_S    5/17/2018       IP      WCH           Adol Rm9   17   \n",
       "715927_20180205_S    2/05/2018       OP      WCH  Level 6 DK Office   13   \n",
       "715927_20180213_S    2/13/2018       IP      WCH       Adol Room 11   13   \n",
       "715927_20180226_S    2/26/2018       OP      WCH              OPD 8   13   \n",
       "...                        ...      ...      ...                ...  ...   \n",
       "1651490_20180206_S   2/06/2018       OP      RAH     Chest Clinic 1   27   \n",
       "1651490_20171215_S  12/15/2017       OP      RAH     Chest Clinic 4   26   \n",
       "1658447_20171006_S  10/06/2017       OP      RAH     Chest Clinic 3   26   \n",
       "1664053_20180406_S   4/06/2018       OP      RAH     Chest Clinic 1   26   \n",
       "1690154_20180406_S   4/06/2018       OP      RAH     Chest Clinic 4   25   \n",
       "\n",
       "                    Age groups Paediatric vs Adult  ... Sum of meds  \\\n",
       "NAME                                                ...               \n",
       "634207_20180510_S            3          Paediatric  ...           1   \n",
       "634207_20180517_S            3          Paediatric  ...           1   \n",
       "715927_20180205_S            3          Paediatric  ...           1   \n",
       "715927_20180213_S            3          Paediatric  ...           3   \n",
       "715927_20180226_S            3          Paediatric  ...           2   \n",
       "...                        ...                 ...  ...         ...   \n",
       "1651490_20180206_S           5               Adult  ...           1   \n",
       "1651490_20171215_S           5               Adult  ...           1   \n",
       "1658447_20171006_S           5               Adult  ...           1   \n",
       "1664053_20180406_S           5               Adult  ...           3   \n",
       "1690154_20180406_S           5               Adult  ...           0   \n",
       "\n",
       "                   Sum of antifungals  Sum of steroid + mabs  \\\n",
       "NAME                                                           \n",
       "634207_20180510_S                   0                      0   \n",
       "634207_20180517_S                   0                      0   \n",
       "715927_20180205_S                   0                      0   \n",
       "715927_20180213_S                   0                      0   \n",
       "715927_20180226_S                   0                      0   \n",
       "...                               ...                    ...   \n",
       "1651490_20180206_S                  0                      0   \n",
       "1651490_20171215_S                  0                      0   \n",
       "1658447_20171006_S                  0                      0   \n",
       "1664053_20180406_S                  1                      0   \n",
       "1690154_20180406_S                  0                      0   \n",
       "\n",
       "                    DNA_extraction_ conc         SAGC ULN  DNA Conc. (ng/ul)  \\\n",
       "NAME                                                                           \n",
       "634207_20180510_S                  0.000  SAGCFN_22_01856               7.82   \n",
       "634207_20180517_S                  0.134  SAGCFN_22_01827              22.80   \n",
       "715927_20180205_S                  0.326  SAGCFN_22_01797              16.50   \n",
       "715927_20180213_S                  0.234  SAGCFN_22_01811              31.00   \n",
       "715927_20180226_S                  0.108  SAGCFN_22_01833              15.10   \n",
       "...                                  ...              ...                ...   \n",
       "1651490_20180206_S                 4.760  SAGCFN_22_01741              26.20   \n",
       "1651490_20171215_S                 7.760  SAGCFN_22_01738              34.20   \n",
       "1658447_20171006_S                 0.098  SAGCFN_22_01837              13.10   \n",
       "1664053_20180406_S                 0.159  SAGCFN_22_01822              30.80   \n",
       "1690154_20180406_S                 0.276  SAGCFN_22_01804              35.00   \n",
       "\n",
       "                      Index I7    Index I5  Mean_Size_BP  \\\n",
       "NAME                                                       \n",
       "634207_20180510_S   CGGACGATTC  CCACCACCTA           651   \n",
       "634207_20180517_S     AGCGATAG    CCTATCCT           633   \n",
       "715927_20180205_S     TAATGCGC    AGGCGAAG           516   \n",
       "715927_20180213_S     TCCGCGAA    CCTATCCT           443   \n",
       "715927_20180226_S   TAACTTGGTC  GATTCACGAC           510   \n",
       "...                        ...         ...           ...   \n",
       "1651490_20180206_S    ATTACTCG    AGGCGAAG           507   \n",
       "1651490_20171215_S    ATTACTCG    ATAGAGGC           564   \n",
       "1658447_20171006_S  CAGCAGGTCA  TACCTAAGTG           576   \n",
       "1664053_20180406_S    TCTCGCGC    TAATCTTA           374   \n",
       "1690154_20180406_S    CGGCTATG    GGCTCTGA           648   \n",
       "\n",
       "                   Total Clusters Passing Filter (Million)  \n",
       "NAME                                                        \n",
       "634207_20180510_S                                      2.9  \n",
       "634207_20180517_S                                      2.4  \n",
       "715927_20180205_S                                      3.4  \n",
       "715927_20180213_S                                      2.7  \n",
       "715927_20180226_S                                      2.6  \n",
       "...                                                    ...  \n",
       "1651490_20180206_S                                     6.4  \n",
       "1651490_20171215_S                                     6.0  \n",
       "1658447_20171006_S                                     2.9  \n",
       "1664053_20180406_S                                     1.9  \n",
       "1690154_20180406_S                                     3.1  \n",
       "\n",
       "[127 rows x 166 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0c7f2287-dee4-4ce4-85e3-b6bfc09e710e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df: (127, 4350)\n",
      "metadata: (127, 166)\n",
      "df_combined: (127, 4516)\n"
     ]
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a2cc3d8d-eeaf-4827-9b14-acb8027ab8dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>taxa_functions</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>10.085904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1447437_20171212_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>59.260325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1128691_20171206_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1128691_20171218_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1128691_20180116_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>14.478968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552445</th>\n",
       "      <td>895293_20180502_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552446</th>\n",
       "      <td>896213_20180427_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552447</th>\n",
       "      <td>913873_20180417_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552448</th>\n",
       "      <td>980574_20180403_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552449</th>\n",
       "      <td>983493_20180123_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>552450 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_id                                 taxa_functions  \\\n",
       "0       1068841_20180306_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "1       1447437_20171212_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "2       1128691_20171206_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "3       1128691_20171218_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "4       1128691_20180116_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "...                    ...                                            ...   \n",
       "552445   895293_20180502_S                               Methylomirabilis   \n",
       "552446   896213_20180427_S                               Methylomirabilis   \n",
       "552447   913873_20180417_S                               Methylomirabilis   \n",
       "552448   980574_20180403_S                               Methylomirabilis   \n",
       "552449   983493_20180123_S                               Methylomirabilis   \n",
       "\n",
       "        abundance  \n",
       "0       10.085904  \n",
       "1       59.260325  \n",
       "2        0.000000  \n",
       "3        0.000000  \n",
       "4       14.478968  \n",
       "...           ...  \n",
       "552445   0.000000  \n",
       "552446   0.000000  \n",
       "552447   0.000000  \n",
       "552448   0.000000  \n",
       "552449   0.000000  \n",
       "\n",
       "[552450 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_long = df.reset_index().melt(id_vars='index', var_name='taxa_functions', value_name='abundance')\n",
    "df_long.rename(columns={'index': 'sample_id'}, inplace=True)\n",
    "df_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "37c1b0ef-1843-49be-af92-c96ec69f0699",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>taxa_functions</th>\n",
       "      <th>abundance</th>\n",
       "      <th>minion</th>\n",
       "      <th>MGI</th>\n",
       "      <th>pwCF_ID</th>\n",
       "      <th>Sample date</th>\n",
       "      <th>IP vs OP</th>\n",
       "      <th>Hospital</th>\n",
       "      <th>Room</th>\n",
       "      <th>...</th>\n",
       "      <th>Sum of meds</th>\n",
       "      <th>Sum of antifungals</th>\n",
       "      <th>Sum of steroid + mabs</th>\n",
       "      <th>DNA_extraction_ conc</th>\n",
       "      <th>SAGC ULN</th>\n",
       "      <th>DNA Conc. (ng/ul)</th>\n",
       "      <th>Index I7</th>\n",
       "      <th>Index I5</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total Clusters Passing Filter (Million)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>10.085904</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2-O-alpha-mannosyl-D-glycerate utilization</td>\n",
       "      <td>2784.895948</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2-aminophenol Metabolism</td>\n",
       "      <td>516.160945</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2-ketoacid oxidoreductases disambiguation</td>\n",
       "      <td>231.778018</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2-oxoglutarate dehydrogenase</td>\n",
       "      <td>122.573392</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 169 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_id                                 taxa_functions  \\\n",
       "0  1068841_20180306_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "1  1068841_20180306_S     2-O-alpha-mannosyl-D-glycerate utilization   \n",
       "2  1068841_20180306_S                       2-aminophenol Metabolism   \n",
       "3  1068841_20180306_S      2-ketoacid oxidoreductases disambiguation   \n",
       "4  1068841_20180306_S                  2-oxoglutarate dehydrogenase    \n",
       "\n",
       "     abundance              minion                 MGI  pwCF_ID Sample date  \\\n",
       "0    10.085904  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "1  2784.895948  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "2   516.160945  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "3   231.778018  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "4   122.573392  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "\n",
       "  IP vs OP Hospital            Room  ...  Sum of meds  Sum of antifungals  \\\n",
       "0       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "1       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "2       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "3       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "4       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "\n",
       "  Sum of steroid + mabs DNA_extraction_ conc         SAGC ULN  \\\n",
       "0                     0                 1.07  SAGCFN_22_01754   \n",
       "1                     0                 1.07  SAGCFN_22_01754   \n",
       "2                     0                 1.07  SAGCFN_22_01754   \n",
       "3                     0                 1.07  SAGCFN_22_01754   \n",
       "4                     0                 1.07  SAGCFN_22_01754   \n",
       "\n",
       "   DNA Conc. (ng/ul)  Index I7  Index I5  Mean_Size_BP  \\\n",
       "0               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "1               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "2               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "3               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "4               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "\n",
       "   Total Clusters Passing Filter (Million)  \n",
       "0                                      2.8  \n",
       "1                                      2.8  \n",
       "2                                      2.8  \n",
       "3                                      2.8  \n",
       "4                                      2.8  \n",
       "\n",
       "[5 rows x 169 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data =  pd.merge(df_long, metadata, left_on='sample_id', right_on=sequence_type)\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0c47cf3-104c-4eb6-a86d-b3b84dcae618",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns_no_spaces = {}\n",
    "for c in merged_data.columns:\n",
    "    if ' ' in c:\n",
    "        columns_no_spaces[c] = c.replace(' ', '_')\n",
    "        \n",
    "merged_data.rename(columns = columns_no_spaces, inplace=True)\n",
    "\n",
    "\n",
    "df_combined.rename(columns = columns_no_spaces, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f048ac6-d1fe-4053-bd6c-1c8db769b0d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['2,3-diacetamido-2,3-dideoxy-d-mannuronic acid',\n",
       " '2-O-alpha-mannosyl-D-glycerate utilization',\n",
       " '2-aminophenol Metabolism',\n",
       " '2-ketoacid oxidoreductases disambiguation',\n",
       " '2-oxoglutarate dehydrogenase ',\n",
       " '2-phosphoglycolate salvage',\n",
       " '3-amino-5-hydroxybenzoic Acid Synthesis',\n",
       " '4-hydroxybenzoyl-CoA reductase',\n",
       " '5-methylaminomethyl-2-thiouridine',\n",
       " 'A Hypothetical Protein Related to Proline Metabolism',\n",
       " 'A new toxin - antitoxin system',\n",
       " 'ABC transporter YeiABEF',\n",
       " 'ABC transporter YxeMNO',\n",
       " 'ABC transporter of unknown substrate X',\n",
       " 'ABC transporter tungstate (TC 3.A.1.6.2)',\n",
       " 'ABC-type iron transport system',\n",
       " 'AMP to 3-phosphoglycerate',\n",
       " 'ATP-dependent Nuclease',\n",
       " 'ATP-dependent RNA helicases, bacterial',\n",
       " 'AaeAB efflux system for hydroxylated, aromatic carboxylic acids',\n",
       " 'Accessory colonization factor',\n",
       " 'Acetoin, butanediol metabolism',\n",
       " 'Acetolactate synthase subunits',\n",
       " 'Acetophenone carboxylase 1',\n",
       " 'Acetyl-CoA Pathway Wood-Ljungdahl',\n",
       " 'Actinobacterial signal transduction system MtrAB-LpqB',\n",
       " 'Activation of fatty acids as acyl-adenylates cluster subgroup FadD10',\n",
       " 'Acyclic terpene utilization',\n",
       " 'Acyl carrier protein',\n",
       " 'Acyl-CoA thioesterase II',\n",
       " 'Adaptation to d-cysteine',\n",
       " 'Adenosyl nucleosidases',\n",
       " 'Adenylylsulfate reductase',\n",
       " 'Adhesins in Staphylococcus',\n",
       " 'Aerotolerance operon',\n",
       " 'Agrobacterium opine transport',\n",
       " 'Agrocinopine utilization',\n",
       " 'Alginate biosynthesis',\n",
       " 'Alkane biosynthesis in bacteria',\n",
       " 'Alkane synthesis in bacteria 2',\n",
       " 'Alpha-acetolactate operon',\n",
       " 'Amidase clustered with urea and nitrile hydratase functions',\n",
       " 'Amino acid racemase',\n",
       " 'Aminoglycoside modifying enzymes: N-acetyltransferases',\n",
       " 'Aminoglycoside modifying enzymes: O-nucleotidyltransferases',\n",
       " 'Aminoglycoside modifying enzymes: O-phosphotransferases',\n",
       " 'An Arabinose Sensor',\n",
       " 'Anaerobic Oxidative Degradation of L-Ornithine',\n",
       " 'Anaerobic dimethyl sulfoxide reductase (EC 1.8.5.3)',\n",
       " 'Anaerobic glycerol-3-phosphate dehydrogenase subunits',\n",
       " 'Anaerobic module of TCA',\n",
       " 'Anaerobic toluene and ethylbenzene degradation',\n",
       " 'Antibiotic anticapsin - bacilysin',\n",
       " 'Antibiotic targets in DNA processing',\n",
       " 'Antibiotic targets in cell wall biosynthesis',\n",
       " 'Antibiotic targets in metabolic pathways',\n",
       " 'Antibiotic targets in protein synthesis',\n",
       " 'Antibiotic targets in transcription',\n",
       " 'Archaeal membrane bound hydrogenases',\n",
       " 'Arginine biosynthesis',\n",
       " 'Arginine biosynthesis via N-acetyl-L-citrulline',\n",
       " 'Arginine decarboxylase and Agmatinase cluster',\n",
       " 'Arginine succinyltransferase pathway',\n",
       " 'Aromatic amino acid interconversions with aryl acids',\n",
       " 'Arsenic resistance',\n",
       " 'Aspartate to Homoserine module',\n",
       " 'Aspartate to Threonine Module',\n",
       " 'AttEFGH ABC Transport System',\n",
       " 'Autoinducer 2 (AI-2) transport and processing (lsrACDBFGE operon)',\n",
       " 'BOX elements in Streptococci',\n",
       " 'Bacillibactin Siderophore',\n",
       " 'Bacillithiol synthesis',\n",
       " 'Bacillus cannibalism system',\n",
       " 'Bacitracin resistance',\n",
       " 'BactLuciferaseComplex',\n",
       " 'Bacterial Sphingolipids',\n",
       " 'Bacterial cell division related cluster 2',\n",
       " 'Bacterial checkpoint-control-related cluster',\n",
       " 'Bacterial pathways for dimethylsulfoniopropionate and acrylate catabolism',\n",
       " 'Bacteriocin resistance related cluster',\n",
       " 'Bacteriocins in Lactobacilli',\n",
       " 'Bacteroides capsular polysaccharide transcription antitermination proteins',\n",
       " 'BarA-UvrY(SirA) two-component regulatory system',\n",
       " 'Barnase-barstar complex',\n",
       " 'Benzoyl-CoA pathway of anaerobic aromatic metabolism',\n",
       " 'Beta-lactamases Ambler class A',\n",
       " 'Beta-lactamases Ambler class B',\n",
       " 'Beta-lactamases Ambler class C',\n",
       " 'Beta-lactamases Ambler class D',\n",
       " 'Betaine lipids in bactera',\n",
       " 'Bicyclomycin resistance cluster',\n",
       " 'Bile hydrolysis',\n",
       " 'Biofilm Adhesin Biosynthesis',\n",
       " 'Biofilm formation in Staphylococcus',\n",
       " 'Biogenesis of c-type cytochromes',\n",
       " 'Biogenesis of cbb3-type cytochrome c oxidases',\n",
       " 'Biogenesis of cytochrome c oxidases',\n",
       " \"Biosynthesis of 4,4'-diapolycopene dialdehyde\",\n",
       " 'Biosynthesis of Arabinogalactan in Mycobacteria',\n",
       " 'Biosynthesis of thiamin antivitamin Bacimethrin',\n",
       " 'Biotin biosynthesis',\n",
       " 'Biotin synthesis & utilization',\n",
       " 'Biotin synthesis cluster',\n",
       " 'Box C-D RNA-guided RNA methyltransferase',\n",
       " 'Branched-Chain Amino Acid Biosynthesis',\n",
       " 'Branched-chain amino acids and alpha-keto acids utilization as energy sources',\n",
       " 'BtuS-PduT gene cluster for metalloporphyrine salvage',\n",
       " 'Butyrate kinase pathway',\n",
       " 'CMP-legionaminic acid synthesis',\n",
       " 'CMP-pseudaminic acid synthesis',\n",
       " 'Cadmium resistance',\n",
       " 'Calvin-Benson cycle',\n",
       " 'Carbon monoxide induced hydrogenase',\n",
       " 'Carbon storage regulator',\n",
       " 'Carbonic anhydrase',\n",
       " 'Carboxysome',\n",
       " 'Cardiolipin biosynthesis',\n",
       " 'Carnitine metabolism and transport',\n",
       " 'Carnitine operon',\n",
       " 'Carotenogenesis related cluster',\n",
       " 'Cell division cluster',\n",
       " 'Cell division initiation related cluster',\n",
       " 'Cell division related cluster',\n",
       " 'Cell division related cluster including coaD',\n",
       " 'Cell envelope-associated LytR-CpsA-Psr transcriptional attenuators',\n",
       " 'Cell invasion-related virulence operon MCE',\n",
       " 'Cell wall-associated cluster in Mycobacterium',\n",
       " 'Chaperones GroEL GroES and Thermosome',\n",
       " 'Chemotaxis in Escherichia coli and Bacillus subtilis',\n",
       " 'Chitobiose',\n",
       " 'Chloramphenicol resistance',\n",
       " 'Chlorophyll Biosynthesis in plants and prokaryotes',\n",
       " 'Chlorophyll Degradation',\n",
       " 'Chlorophyll a synthase and phytol kinase cluster',\n",
       " 'Chlorosome',\n",
       " 'Cholera toxin',\n",
       " 'Cholesterol catabolism',\n",
       " 'Choline uptake and conversion to betaine clusters',\n",
       " 'Chorismate Synthesis',\n",
       " 'Citrate lyase',\n",
       " 'Citrate uptake cluster',\n",
       " 'Cluster containing Glutathione synthetase',\n",
       " 'Cluster related to 2-thiouridine modification of tRNAs',\n",
       " 'Cluster related to PAS - para-aminosalicylic acid resistance',\n",
       " 'CoA disulfide thiol-disulfide redox system',\n",
       " 'Coat proteins CotJABC',\n",
       " 'Cob(I)alamin adenosyltransferase',\n",
       " 'Cobalamin synthesis',\n",
       " 'Cobalamin synthesis from Cob(II)yrinate a,c diamide',\n",
       " 'Cobalamin synthesis, Precorin-2 to Cob(II)yrinate a,c diamide, aerobic',\n",
       " 'Cobalamin synthesis, Precorin-2 to Cob(II)yrinate a,c diamide, anaerobic',\n",
       " 'Coenzyme A',\n",
       " 'Coenzyme A -- gjo',\n",
       " 'Coenzyme A Biosynthesis cluster',\n",
       " 'Coenzyme B synthesis',\n",
       " 'Coenzyme F420 synthesis',\n",
       " 'Coenzyme F420-H2 dehydrogenase (methanophenazine)',\n",
       " 'Coenzyme PQQ synthesis',\n",
       " 'Colanic acid synthesis',\n",
       " 'Cold shock proteins of CSP family',\n",
       " 'Colicin V and Bacteriocin Production Cluster',\n",
       " 'Conjugative transfer',\n",
       " 'Conjugative transposon, Bacteroidales',\n",
       " 'Control of cell elongation - division cycle in Bacilli',\n",
       " 'Copper Transport System',\n",
       " 'Copper homeostasis: copper tolerance',\n",
       " 'Copper transport and blue copper proteins',\n",
       " 'Copper uptake system CopCD',\n",
       " 'CtsR and MscAB regulation of protein degradation',\n",
       " 'Cyanate hydrolysis',\n",
       " 'Cyanobacterial bypass in the TCA',\n",
       " 'Cyanophycin Metabolism',\n",
       " 'Cyclic Di-GMP Turnover Proteins',\n",
       " 'Cysteine dioxygenase containing cluster 1',\n",
       " 'Cysteine synthesis',\n",
       " 'CytR regulation',\n",
       " 'Cytochrome b6-f complex in plants (plastidial) and cyanobacteria',\n",
       " 'Cytochrome bo ubiquinol oxidase ',\n",
       " 'Cytochrome c oxidase EC 1.9.3.1',\n",
       " 'Cytochrome d ubiquinol oxidase operon',\n",
       " 'Cytolysin and Lipase operon in Vibrio',\n",
       " 'D-3-phosphoglycerate to phosphoserine module',\n",
       " 'D-Galacturonate and D-Glucuronate Utilization',\n",
       " 'D-alanylation of teichoic acid',\n",
       " 'D-allose utilzation',\n",
       " 'D-altritol utilization',\n",
       " 'D-arabinitol utilization',\n",
       " 'D-arabinose utilization',\n",
       " 'D-galacturonate utilization, oxidative pathway',\n",
       " 'D-gluconate and ketogluconates metabolism',\n",
       " 'D-glucosaminate utilization',\n",
       " 'D-threonate, L-threonate and D-erythronate utilization',\n",
       " 'D-tyrosyl-tRNA(Tyr) deacylase',\n",
       " 'DAP (1,3-diaminopropane) production',\n",
       " 'DNA Helicase of Unknown Function',\n",
       " 'DNA Repair Base Excision',\n",
       " 'DNA internalization-related cluster',\n",
       " 'DNA phosphorothioation-dependent restriction system DptFGH',\n",
       " 'DNA processing cluster',\n",
       " 'DNA recombination, archaeal',\n",
       " 'DNA repair system including RecA, MutS and a hypothetical protein',\n",
       " 'DNA repair, UvrABC system',\n",
       " 'DNA repair, bacterial',\n",
       " 'DNA repair, bacterial DinG and relatives',\n",
       " 'DNA repair, bacterial MutHLS system',\n",
       " 'DNA repair, bacterial RecBCD pathway',\n",
       " 'DNA repair, bacterial RecFOR pathway',\n",
       " 'DNA repair, bacterial SbcCD exonuclease',\n",
       " 'DNA repair, bacterial UmuCD system',\n",
       " 'DNA repair, bacterial UvrD and related helicases',\n",
       " 'DNA repair, bacterial photolyase',\n",
       " 'DNA sulfur modification cluster',\n",
       " 'DNA topoisomerases, Type II, ATP-dependent',\n",
       " 'Damaged nicotinamide nucleotide NAD(P)HX repair',\n",
       " 'De Novo Pyrimidine Synthesis',\n",
       " 'DeNovo Purine Biosynthesis',\n",
       " 'Degradation of branched-chain amino acids and alpha-keto acids [Leu, Val]',\n",
       " 'Dehydrogenase complexes',\n",
       " 'Delta(6)-desaturase containing cluster',\n",
       " 'Denitrifying reductase gene clusters',\n",
       " 'Desulfovibrio high-molecular-weight cytochrome cluster',\n",
       " 'DevBCA efflux pump transporter (of ABC type) cyanobacterial',\n",
       " 'Diaminopimelate Synthesis ',\n",
       " 'Dihydropyrimidine dehydrogenase',\n",
       " 'Dihydroxyacetone kinases',\n",
       " 'Dioxygenolytic carbon monoxide-forming enzymes',\n",
       " 'Diphthamide biosynthesis',\n",
       " 'Diphtheria toxin',\n",
       " 'Dipicolinate Synthesis',\n",
       " 'Dissimilatory nitrite reductase',\n",
       " 'Dpp dipeptide ABC transport system',\n",
       " 'ESAT-6 protein secretion system in Mycobacteria (locus ESX-5)',\n",
       " 'ESAT-6 proteins secretion system in Firmicutes and Actinobacteria',\n",
       " 'ESAT-6 proteins secretion system in Firmicutes extended',\n",
       " 'ESAT-6 proteins secretion system in Mycobacteria (locus ESX-1)',\n",
       " 'ESAT-6 proteins secretion system in Mycobacteria (locus ESX-3)',\n",
       " 'ESAT-6 proteins secretion system in Mycobacteria (locus ESX-4)',\n",
       " 'EcsAB transporter affecting expression and secretion of secretory preproteins',\n",
       " 'Ectoine synthesis',\n",
       " 'Ectoine, hydroxyectoine uptake and catabolism',\n",
       " 'Efflux ABC transporter MdlAB',\n",
       " 'Efflux ABC transporter YadGH',\n",
       " 'Efflux ABC transporters BmrCD and LmrCD involved in multidrug resistance',\n",
       " 'Electron-bifurcating caffeyl-CoA reductase-Etf complex',\n",
       " 'Encapsulation of DyP-type peroxidase or ferritin-like protein oligomers',\n",
       " 'Energy conserving hydrogenase b, Methanococcales-Methanobacteriales-Methanopyrales',\n",
       " 'Energy-conserving hydrogenase (ferredoxin)',\n",
       " 'Enoyl-[ACP] reductases disambiguation',\n",
       " 'Entericidin',\n",
       " 'Enterobacterial common antigen (LPS O-antigen)',\n",
       " 'Entner-Doudoroff Pathway',\n",
       " 'EnvZ and OmpR regulon',\n",
       " 'Ergothioneine biosynthesis',\n",
       " 'Ethanolamine utilization',\n",
       " 'Ethylmalonyl-CoA pathway of C2 assimilation',\n",
       " 'Exosome',\n",
       " 'Exosporium',\n",
       " 'Extracellular matrix proteins (PEL) involved in glucose-rich biofilm formation in Pseudomonas',\n",
       " 'Extracellular matrix proteins (PSL) involved in mannose-rich biofilm formation in Pseudomonas',\n",
       " 'F0F1-type ATP synthase',\n",
       " 'Fatty Acid Biosynthesis FASII',\n",
       " 'Fatty Acid Biosynthesis cluster',\n",
       " 'Fatty acid catabolic operon fadN-fadA-fadE (yusJKL)',\n",
       " 'Fatty acid degradation',\n",
       " 'Fatty acid synthesis',\n",
       " 'Fe(III) respiration - Shewanella type',\n",
       " 'Fe-S cluster assembly',\n",
       " 'Fermentations: Lactate',\n",
       " 'Fermentations: Mixed acid',\n",
       " 'Ferrous iron transporter EfeUOB, low-pH-induced',\n",
       " 'Fet iron export system',\n",
       " 'Flagellum',\n",
       " 'Folate Biosynthesis',\n",
       " 'Folate biosynthesis cluster',\n",
       " 'Forespore to mother cell channel',\n",
       " 'Formaldehyde assimilation: Ribulose monophosphate pathway',\n",
       " 'Formate dehydrogenase',\n",
       " 'Fosfomycin resistance',\n",
       " 'Fructoselysine and glucoselysine',\n",
       " 'FtsEX cell-division-associated signaling system',\n",
       " 'Fusaric acid resistance cluster',\n",
       " 'Fusidic acid resistance',\n",
       " 'G-protein-coupled receptor (GPCR) system in Actinobacteria',\n",
       " 'GDP-fucose to mannose-1P module',\n",
       " 'GMP synthase',\n",
       " 'Galactosamine Substituent of Arabinogalactan in Mycobacteria',\n",
       " 'Galactose utilization',\n",
       " 'Gallic acid utilization',\n",
       " 'Gene Transfer Agent',\n",
       " 'General Secretion Pathway',\n",
       " 'Gentisate pathway',\n",
       " 'GlrKR two-component regulatory system',\n",
       " 'Glutaconate CoA-transferase or 3-oxoadipate CoA-transferase Subunits',\n",
       " 'Glutamate dehydrogenases',\n",
       " 'Glutamate fermentation',\n",
       " 'Glutamine synthetases',\n",
       " 'Glutathione analogs: mycothiol',\n",
       " 'Glutathione-dependent pathway of formaldehyde detoxification',\n",
       " 'Glutathione: Biosynthesis and gamma-glutamyl cycle',\n",
       " 'Glutathione: Non-redox reactions',\n",
       " 'Glutathione: Redox cycle',\n",
       " 'Glutathionylspermidine and Trypanothione',\n",
       " 'Glycerol ABC transporter',\n",
       " 'Glycerol fermentation to 1,3-propanediol',\n",
       " 'Glycerol utilization cluster',\n",
       " 'Glycerolipid and Glycerophospholipid Metabolism in Bacteria',\n",
       " 'Glycine betaine synthesis from choline',\n",
       " 'Glycine cleavage system',\n",
       " 'Glycine reductase, sarcosine reductase and betaine reductase',\n",
       " 'Glycogen metabolism',\n",
       " 'Glycolate, glyoxylate interconversions',\n",
       " 'Glycolysis and Gluconeogenesis',\n",
       " 'Glycolysis and Gluconeogenesis, including Archaeal enzymes',\n",
       " 'Glycyl-tRNA synthetase',\n",
       " 'Glyoxylate bypass',\n",
       " 'Gram-negative cluster',\n",
       " 'HMG CoA Synthesis',\n",
       " 'HPr catabolite repression system',\n",
       " 'HU-alpha and dsDNA mimic protein interaction',\n",
       " 'Head-to-head olefinic hydrocarbon biosynthesis',\n",
       " 'Heat shock dnaK gene cluster extended',\n",
       " 'Heme Biosynthesis: protoporphyrin-, coproporphyrin- and siroheme-dependent pathways',\n",
       " 'Heme O and Heme A biosynthesis',\n",
       " 'Heme O and Heme A biosynthesis (with selected terminal oxidases)',\n",
       " 'Heme and heme d1 biosynthesis from siroheme',\n",
       " 'Heme, hemin uptake and utilization systems in GramPositives',\n",
       " 'Heterocyst formation in cyanobacteria',\n",
       " 'Hexose phosphate transport system',\n",
       " 'Hfl operon',\n",
       " 'High affinity phosphate transporter and control of PHO regulon',\n",
       " 'HipAB system implicated in growth arrest, persistence and drug tolerance',\n",
       " 'Histidine Biosynthesis',\n",
       " 'Histidine Biosynthesis -- gjo',\n",
       " 'Histidine Degradation',\n",
       " 'Homogentisate pathway',\n",
       " 'Homoprotocatechuate degradative cluster',\n",
       " 'Hopanoid biosynthesis',\n",
       " 'Hyaluronate utilization',\n",
       " 'Hyaluronic Acid-containing Cell Walls',\n",
       " 'Hydrogen-sensing hydrogenase system',\n",
       " 'Hydrolysis of sphingomyelin',\n",
       " 'Hydroxy-fatty acid production as stress response',\n",
       " 'Hydroxyaromatic decarboxylase family',\n",
       " 'Hyperosmotic potassium uptake',\n",
       " 'IMP and Xanthine conversion to XMP and GMP module',\n",
       " 'IbrA and IbrB: co-activators of prophage gene expression',\n",
       " 'Indole-pyruvate oxidoreductase complex',\n",
       " 'Indoleacetamide metabolism',\n",
       " 'Inner membrane proteins of MarC family, not involved in antibiotic resistance',\n",
       " 'Inositol catabolism',\n",
       " 'Inositol-containing phospholipids',\n",
       " 'Integrons',\n",
       " 'Iron Scavenging cluster in Thermus',\n",
       " 'Iron acquisition in Streptococcus',\n",
       " 'Iron siderophore sensor & receptor system',\n",
       " 'Iron transport system including ABC transporter',\n",
       " 'Iron(III) dicitrate transport system Fec',\n",
       " 'Isoniazid inductible proteins cluster',\n",
       " 'Isoprenoid Biosynthesis: Interconversions',\n",
       " 'KDO2-Lipid A biosynthesis',\n",
       " 'L-2-amino-4-methoxy-trans-3-butenoic acid synthesis',\n",
       " 'L-2-amino-adipate to lysine module',\n",
       " 'L-Amino acid oxidases: lysine-epsilon oxidase LodA',\n",
       " 'L-ascorbate utilization (and related gene clusters)',\n",
       " 'L-fucose dissimilation cluster',\n",
       " 'L-fucose utilization',\n",
       " 'L-galactonate utilization',\n",
       " 'L-rhamnonate utilization cluster',\n",
       " 'L-serine dehydratase subunits',\n",
       " 'Lactose utilization',\n",
       " 'Lambdoid phage',\n",
       " 'Lasso peptide maturation cluster',\n",
       " 'Late competence',\n",
       " 'Lipid orphans',\n",
       " 'Lipid-linked oligosaccharide synthesis related cluster',\n",
       " 'Lipoic acid metabolism',\n",
       " 'Lipopolysaccharide in K12',\n",
       " 'Lipopolysaccharide transporter LptABC',\n",
       " 'Lipopolysaccharide-related cluster in Alphaproteobacteria',\n",
       " 'Lipoprotein transporter to outer membrane Lol',\n",
       " 'Lipoylated proteins',\n",
       " 'Lipoylation-related cluster',\n",
       " 'Listeria Pathogenicity Island LIPI-1 extended',\n",
       " 'Listeria surface proteins: Internalin-like proteins',\n",
       " 'Lpt lipopolysaccharide transport system',\n",
       " 'Lysine Biosynthesis via alpha-aminoadipate AAA',\n",
       " 'Lysine DAP biosynthetic pathway',\n",
       " 'Lysine fermentation',\n",
       " 'Lysine fermentation to crotonoyl-CoA',\n",
       " 'Lysine leader peptide',\n",
       " 'Macrolides, lincosamides, streptogramins, ketolides, oxazolidinones (MLSKO) resistance: efflux systems',\n",
       " 'Macrolides, lincosamides, streptogramins, ketolides, oxazolidinones (MLSKO) resistance: enzymatic degradation',\n",
       " 'Macrolides, lincosamides, streptogramins, ketolides, oxazolidinones (MLSKO) resistance: rRNA methylases',\n",
       " 'Macrolides, lincosamides, streptogramins, ketolides, oxazolidinones (MLSKO) resistance: ribosomal protection',\n",
       " 'Macromolecular synthesis operon',\n",
       " 'Magnesium transport',\n",
       " 'Maintenance of outer membrane lipid asymmetry',\n",
       " 'Malonate decarboxylase',\n",
       " 'Mannosylglucosylglycerate synthesis',\n",
       " 'MazEF toxin-antitoxing (programmed cell death) system',\n",
       " 'Mebrane bound hydrogenases',\n",
       " 'Mechanisms of PAS resistance',\n",
       " 'Membrane bound hydrogenase Mbh',\n",
       " 'Membrane bound hydrogenase Mbh -- gjo',\n",
       " 'Menaquinone biosynthesis from chorismate via 1,4-dihydroxy-2-naphthoate',\n",
       " 'Menaquinone biosynthesis from chorismate via 1,4-dihydroxy-6-naphthoate',\n",
       " 'Metabolite repair',\n",
       " 'Metal chelatases',\n",
       " 'Methanofuran biosynthesis',\n",
       " 'Methanogenesis',\n",
       " 'Methanogenesis from methylated compounds',\n",
       " 'Methionine Salvage',\n",
       " 'Methylenetetrahydrofolate reductase',\n",
       " 'Methylglyoxal Metabolism',\n",
       " 'Methylglyoxal resistance in Bacillus subtilis',\n",
       " 'Methylhydantoinase',\n",
       " 'Methylthiotransferases',\n",
       " 'Mevalonate metabolic pathway',\n",
       " 'MexXY System of Pseudomonas aeruginosa',\n",
       " 'Modification of eukaryotic initiation factor 5A',\n",
       " 'Molybdenum cofactor biosynthesis',\n",
       " 'Molybdopterin cytosine dinucleotide',\n",
       " 'Mu-like phage',\n",
       " 'Muconate lactonizing enzyme family',\n",
       " 'MukBEF Chromosome Condensation',\n",
       " 'Multi-subunit cation antiporter',\n",
       " 'Multidrug Resistance, Tripartite Systems Found in Gram Negative Bacteria TEMP',\n",
       " 'Multidrug resistance RND efflux system MexEFOprN',\n",
       " 'Multiple Antibiotic Resistance MAR locus',\n",
       " 'Mupirocin resistance',\n",
       " 'Murein hydrolase regulation and cell death',\n",
       " 'Mycobacterial FadD proteins (fatty acid CoA- and AMP- ligases)',\n",
       " 'Mycobacterial FadE proteins Acyl-CoA dehydrogenase',\n",
       " 'Mycobacterial HadABC proteins ((3R)-hydroxyacyl-ACP dehydrates of FASII)',\n",
       " 'Mycobacterial cell wall virulence lipid phthiocerol dimycocerosate (PDIM)',\n",
       " 'Mycobacterial gene cluster associated with resistance against FAS-II antibiotics',\n",
       " 'Mycobacterial heme acquisition system',\n",
       " 'Mycobacterial lipid and multidrug efflux system proteins MmpL, MmpS',\n",
       " 'Mycobacterial lipoproteins (LPR group)',\n",
       " 'Mycobacterial multidrug efflux systems implicated in AMR ',\n",
       " 'Mycobacterial signal transduction system MprAB required for persistent infections',\n",
       " 'Mycofactocin system cluster',\n",
       " 'Mycolic acid synthesis 2',\n",
       " 'Mycosporine synthesis cluster',\n",
       " \"N,N'-diacetylbacillosamine\",\n",
       " 'N-acetylneuraminate utilization',\n",
       " 'NAD and NADP cofactor biosynthesis bacterial',\n",
       " 'NAD and NADP cofactor biosynthesis global',\n",
       " 'NAD-reducing [NiFe] hydrogenase HoxFUYH(E)',\n",
       " 'NADH ubiquinone oxidoreductase',\n",
       " 'NADH ubiquinone oxidoreductase vs. multi-subunit cation antiporter',\n",
       " 'NADP-reducing [FeFe]-hydrogenase',\n",
       " 'Na(+) H(+) antiporter',\n",
       " 'Na-driven 2-hydroxyglutarate pathway',\n",
       " 'Na-translocating NADH-quinone oxidoreductase',\n",
       " 'NhaA, NhaD and Sodium-dependent phosphate transporters',\n",
       " 'NiFe hydrogenase maturation',\n",
       " 'Nickel-requiring',\n",
       " 'Nicotinate catabolism, anaerobic',\n",
       " 'Nicotinic acid utilization',\n",
       " 'Nitrilase',\n",
       " 'Nitrogen fixation',\n",
       " 'Nonmevalonate Branch of Isoprenoid Biosynthesis',\n",
       " 'Nucleoside ABC transporter cluster',\n",
       " 'Nucleoside triphosphate pyrophosphohydrolase MazG',\n",
       " 'Nucleoside uptake and degradation cluster',\n",
       " 'Nudix proteins (nucleoside triphosphate hydrolases)',\n",
       " 'O-methyl phosphoramidate modification in capsular polysaccharide',\n",
       " 'Oligopeptide degradation cluster',\n",
       " 'One-carbon metabolism by tetrahydropterines',\n",
       " 'Osmoregulation',\n",
       " 'Osmotic stress cluster',\n",
       " 'Outer membrane',\n",
       " 'Outer membrane porins in Mycobacteria',\n",
       " 'Outer membrane porins in Pseudomonas and Acinetobacter',\n",
       " 'Outer membrane proteases (Omptins)',\n",
       " 'Oxaloacetate decarboxylase Na-pump',\n",
       " 'Oxygen-sensing diguanylate cyclase and phosphodiesterase couple for c-di-GMP control',\n",
       " 'Oxygen-sensing two-component regulatory system NreABC',\n",
       " 'P2-like phage',\n",
       " 'PE multigene family in Mycobacterium tuberculosis complex (MTBC)',\n",
       " 'PEB1, the major cell-binding factor of Campylobacter jejuni',\n",
       " 'PPE multigene family in Mycobacterium tuberculosis complex (MTBC)',\n",
       " 'Particulate methane monooxygenase (pMMO)',\n",
       " 'Pentalenolactone Biosynthesis and Resistance',\n",
       " 'Pentose phosphate pathway',\n",
       " 'Peptidase clustering with DAP',\n",
       " 'Peptide ABC transport system Sap',\n",
       " 'Peptide methionine sulfoxide reductase',\n",
       " 'Peptidyl-prolyl cis-trans isomerase',\n",
       " 'Periplasmic disulfide interchange',\n",
       " 'Periplasmic nitrate reductase EC 1.7.99.4',\n",
       " 'Phage shock protein operon',\n",
       " 'Phd-Doc, YdcE-YdcD toxin-antitoxin (programmed cell death) systems',\n",
       " 'Phenazine biosynthesis',\n",
       " 'Phenol hydroxylase',\n",
       " 'Phenylacetyl-CoA catabolic pathway (core)',\n",
       " 'Phenylalanine and Tyrosine synthesis 1',\n",
       " 'Phosphatidylinositol mannosides biosynthesis related cluster',\n",
       " 'Phosphoenolpyruvate phosphomutase',\n",
       " 'Phospholipid and Fatty acid biosynthesis related cluster',\n",
       " 'Phosphonate (phosphite) dehydrogenase',\n",
       " 'Phosphonate metabolism',\n",
       " 'Phosphonoalanine utilization',\n",
       " 'Photorespiration (oxidative C2 cycle)',\n",
       " 'Photosystem I',\n",
       " 'Photosystem II',\n",
       " 'Phr peptides - Rap phosphatases signaling',\n",
       " 'Plasmid replication',\n",
       " 'Poly-gamma-glutamate biosynthesis',\n",
       " 'Polyadenylation, bacterial',\n",
       " 'Polyamine Metabolism',\n",
       " 'Polymyxin resistance',\n",
       " 'Polymyxin resistance, lipid A modifications with 4-amino-4-deoxy-L-arabinose',\n",
       " 'Polymyxin resistance, lipid A modifications with phosphoethanolamine',\n",
       " 'Polysulfide reductase EC 1.12.98.4 and Thiosulfate reductase',\n",
       " 'Polyunsaturated Fatty Acids synthesis',\n",
       " 'Pore-forming cytolytic toxins',\n",
       " 'Possible DNA repair related cluster',\n",
       " 'Possible RNA modification and stress response cluster',\n",
       " 'Possible heterocyst differentiation related cluster',\n",
       " 'Possible new toxin-antitoxin system including DivIC',\n",
       " 'Possible stress related actinobacterial cluster',\n",
       " 'Post-translational modification of proteins',\n",
       " 'Potassium-transporting ATPase',\n",
       " 'Potential DNA repair cluster',\n",
       " 'Predicted secretion system W clustering with cell division proteins',\n",
       " 'Predicted secretion system X',\n",
       " 'Prenylated indole derivatives in actinomycetes',\n",
       " 'Programmed cell death toxin-antitoxin PezAT',\n",
       " 'Programmed frameshift',\n",
       " 'Proline Synthesis',\n",
       " 'Proline biosynthesis (for review)',\n",
       " 'Proline, 4-hydroxyproline uptake and utilization',\n",
       " 'Propanediol utilization',\n",
       " 'Proteasome bacterial',\n",
       " 'Protection from Reactive Oxygen Species',\n",
       " 'Protein chaperones',\n",
       " 'Protein degradation',\n",
       " \"Protein-O-mannosyltransferase and 16S rRNA (cytidine(1402)-2'-O)-methyltransferase cluster\",\n",
       " 'Proteorhodopsin',\n",
       " 'Pterin synthesis related cluster',\n",
       " 'Purine catabolism in Bacillus subtilis',\n",
       " 'Purine nucleotide synthesis regulator',\n",
       " 'Putative TldE-TldD proteolytic complex',\n",
       " 'Putative oxidase COG2907',\n",
       " 'Putrescine utilization',\n",
       " 'Pyridoxin (Vitamin B6) Biosynthesis',\n",
       " 'Pyridoxin(Vitamin B6) Degradation Pathway',\n",
       " 'Pyrroloquinoline Quinone biosynthesis',\n",
       " 'Pyruvate Alanine Serine Interconversions',\n",
       " 'Pyruvate formate-lyase cluster',\n",
       " 'Pyruvate formate-lyase cluster with possible role in choline utilization',\n",
       " 'Pyruvate metabolism I: anaplerotic reactions, PEP',\n",
       " 'Pyruvate metabolism II: acetyl-CoA, acetogenesis from pyruvate',\n",
       " 'Pyruvate:ferredoxin oxidoreductase',\n",
       " 'Queuosine and archaeosine',\n",
       " 'Quinolone (Fluoroquinolone) resistance via Pentapeptide repeat proteins',\n",
       " 'Quinone oxidoreductase family',\n",
       " 'Quorum sensing in Yersinia',\n",
       " 'Quorum sensing regulation in Pseudomonas',\n",
       " 'Quorum-sensing related cluster',\n",
       " 'RNA polymerase, archaeal',\n",
       " 'RNA polymerase, archaeal initiation factors',\n",
       " 'RNA polymerase, bacterial',\n",
       " 'RNA processing and degradation, bacterial',\n",
       " 'RNA pseudouridine synthases',\n",
       " 'RSF1010 like IncQ broad-host-range plasmids',\n",
       " 'Rcn nickel and cobalt homeostasis system',\n",
       " 'Rcs two-component regulator of capsule synthesis',\n",
       " 'RecA and RecX',\n",
       " 'Repair of Iron Centers',\n",
       " 'Replication-associated recombination protein RarA',\n",
       " 'Resistance to Daptomycin',\n",
       " 'Resistance to Ethambutol',\n",
       " 'Resistance to Ethionamide and Isoniazid',\n",
       " 'Resistance to Triclosan',\n",
       " 'Resistance to Vancomycin and Teicoplanin',\n",
       " 'Resistance to chromium compounds',\n",
       " 'Resistance to the fluoroquinolones norfloxacin and ciprofloxacin',\n",
       " 'Respiratory Complex I',\n",
       " 'Respiratory Systems - Shewanella (updated)',\n",
       " 'Respiratory complex I in plants (plastidial) and cyanobacteria',\n",
       " 'Rhamnolipids in Pseudomonas',\n",
       " 'Ribitol utilization',\n",
       " 'Riboflavin, FMN and FAD biosynthesis in Archaea',\n",
       " 'Riboflavin, FMN and FAD metabolism with fusion events',\n",
       " 'Ribonuclease H',\n",
       " 'Ribonuclease J family',\n",
       " 'Ribonucleotide reduction',\n",
       " 'Ribosomal hibernation related cluster',\n",
       " 'Ribosomal protein S12p Asp methylthiotransferase',\n",
       " 'Ribosomal protein S5p acylation',\n",
       " 'Ribosomal proteins, single-copy',\n",
       " 'Ribosome LSU, bacterial',\n",
       " 'Ribosome LSU, eukaryotic and archaeal',\n",
       " 'Ribosome SSU, eukaryotic and archaeal',\n",
       " 'Ribosome activity modulation',\n",
       " 'Ribosome biogenesis, archaeal',\n",
       " 'RuBisCO operon',\n",
       " 'Rut pathway of pyrimidine degradation',\n",
       " 'RuvABC plus a hypothetical',\n",
       " 'S-Adenosyl-L-homocysteine recycling',\n",
       " 'S-methylmethionine',\n",
       " 'Salmochelin-mediated Iron Acquisition',\n",
       " 'Sarcosine oxidases, monomeric and heterotetrameric',\n",
       " 'SecY2-SecA2 Specialized Transport System',\n",
       " 'Selenocysteine metabolism',\n",
       " 'Selenoprotein O',\n",
       " 'Sex pheromones in Enterococcus faecalis and other Firmicutes',\n",
       " 'Shikimate kinase SK3 cluster',\n",
       " 'Siderophore Achromobactin',\n",
       " 'Siderophore Aerobactin',\n",
       " 'Siderophore Anthrachelin',\n",
       " 'Siderophore Enterobactin',\n",
       " 'Siderophore Mycobactin',\n",
       " 'Siderophore Pyoverdine',\n",
       " 'Siderophore Staphylobactin',\n",
       " 'Siderophore Yersiniabactin Biosynthesis',\n",
       " 'Siderophore assembly kit',\n",
       " 'Siderophore pyochelin',\n",
       " 'Sigma54-dependent transcription related gene cluster',\n",
       " 'Sigma54-dependent transcription related gene cluster 2',\n",
       " 'Signal peptidase',\n",
       " 'Small efflux transporters of SMR family',\n",
       " 'Sodium-transporting ATPase',\n",
       " 'Sox sulfur oxidation genes',\n",
       " 'Sox thiosulfate oxidation system',\n",
       " 'SpoVS protein family',\n",
       " 'Spore germinant receptors',\n",
       " 'Spore germination',\n",
       " 'Sporulation Cluster',\n",
       " 'Sporulation Cluster III A',\n",
       " 'Sporulation gene orphans',\n",
       " 'Sporulation proteins SigEG cluster',\n",
       " 'Sporulation proteins SpoIIIAA-SpoIIIAH',\n",
       " 'Sporulation proteins SpoVA cluster',\n",
       " 'Sporulation-associated proteins with broader functions',\n",
       " 'Streptococcal Hyaluronic Acid Capsule',\n",
       " 'Streptolysin S',\n",
       " 'Stress proteins YciF, YciE',\n",
       " 'Stress response and cell wall lysis cluster',\n",
       " 'Stringent Response, (p)ppGpp metabolism',\n",
       " 'Succinate dehydrogenase and Fumarate reductase cpmlexes',\n",
       " 'Sucrose to levan conversions',\n",
       " 'Sugar-phosphate stress regulation',\n",
       " 'Sulfate assimilation related cluster',\n",
       " 'Sulfate reduction-associated complexes',\n",
       " 'Sulfhydrogenase',\n",
       " 'Sulfoquinovose biosynthesis cluster',\n",
       " 'Sulfur metabolism related cluster',\n",
       " 'Sulfur transfer pathway CsdAEL',\n",
       " 'Sulfur transport related cluster',\n",
       " 'Sulfurtransferase and Ssec cluster',\n",
       " 'Sulphoglycolysis',\n",
       " 'TAM transport system',\n",
       " 'TCA Cycle',\n",
       " 'Tagatose utilization',\n",
       " 'Taurine and alkanesulfonate utilization as sulfur sources',\n",
       " 'Teichuronic acid biosynthesis',\n",
       " 'Teicoplanin resistance in Staphylococci',\n",
       " 'Tetracycline resistance, all mechanisms',\n",
       " 'Tetracycline resistance, ribosomal protection type',\n",
       " 'Tetrathionate respiration',\n",
       " 'The Chv two-component regulatory system',\n",
       " 'Thiamin, hydroxymethylpyrimidine selected transporters',\n",
       " 'Thiamin, thiazole, hydroxymethylpyrimidine salvage and uptake',\n",
       " 'Thiazole.oxazole-modified microcins',\n",
       " 'Thiocyanate hydrolase',\n",
       " 'Threonine anaerobic catabolism gene cluster',\n",
       " 'Threonine synthase cluster',\n",
       " 'Threonylcarbamoyladenosine',\n",
       " 'Tocopherols and tocotrienols (vitamin E) in plants and cyanobacteria',\n",
       " 'Tol-Pal Cell Envelope Complex',\n",
       " 'Tolerance to colicin E2 and CreBC signal transduction system',\n",
       " 'Toluene 4-monooxygenase (T4MO)',\n",
       " 'TonB-ExbBD energy transducing system',\n",
       " 'Toxin-Antitoxin MT1',\n",
       " 'Toxin-Antitoxin system HigAB',\n",
       " 'Toxin-antitoxin replicon stabilization systems',\n",
       " 'Toxins',\n",
       " 'Trans-envelope signaling system VreARI in Pseudomonas',\n",
       " 'Trans-translation by stalled ribosomes',\n",
       " 'Transcription factors, cyanobacterial RpoD-like sigma factors',\n",
       " 'Transcriptional regulator AfsR',\n",
       " 'Translation elongation factor G family',\n",
       " 'Translation elongation factors, bacterial',\n",
       " 'Translation elongation factors, eukaryotic and archaeal',\n",
       " 'Translation initiation factors, bacterial',\n",
       " 'Translation initiation factors, eukaryotic and archaeal',\n",
       " 'Translation termination factors, bacterial',\n",
       " 'Tricarballylate Utilization',\n",
       " 'Tricarboxylate transport system',\n",
       " 'Trimethylamine-N-oxide operon',\n",
       " 'Tripartite efflux system (of RND type) associated with Geranylgeranyl-PP synthase',\n",
       " 'Tripartite multidrug efflux systems (of MFS type) EmrKY-TolC and EmrAB-TolC',\n",
       " 'Tripartite multidrug efflux systems (of MFS type) in Gram-negative bacteria',\n",
       " 'Tripartite multidrug efflux systems (of RND type) AcrAB-TolC and related transporters',\n",
       " 'Tripartite multidrug efflux systems (of RND type) CmeABC and CmeDEF',\n",
       " 'Tripartite multidrug efflux systems (of RND type) MdtABC-TolC',\n",
       " 'Tripartite multidrug efflux systems (of RND type) in Pseudomonas',\n",
       " 'Trk and Ktr potassium uptake systems',\n",
       " 'Tryptophan synthesis',\n",
       " 'Twin-arginine translocation system',\n",
       " 'Twin-arginine translocation system related cluster',\n",
       " 'Two cell division clusters relating to chromosome partitioning',\n",
       " 'Two partner secretion pathway (TPS)',\n",
       " 'Type 4 conjugative transfer system, IncI1 type',\n",
       " 'Type 4 secretion and conjugative transfer',\n",
       " 'Type I restriction-modification systems',\n",
       " 'Type I secretion systems disambiguation',\n",
       " 'Type III secretion system',\n",
       " 'Type III secretion system orphans',\n",
       " 'Type III secretion systems',\n",
       " 'Type IV pilus',\n",
       " 'Type VI secretion system',\n",
       " 'Ubiquinone biosynthesis',\n",
       " 'Undecaprenyl-diphosphatases',\n",
       " 'Universal GTPases',\n",
       " 'Universal stress protein family',\n",
       " 'Urea carboxylase and Allophanate hydrolase cluster',\n",
       " 'Urea cycle',\n",
       " 'Urease subunits',\n",
       " 'Uridine to beta-Alanine and Thymine to 3-aminoisobutanoate module',\n",
       " 'Utilization systems for glycans and polysaccharides (PULs, SUS)',\n",
       " 'V-Type ATP synthase',\n",
       " 'Vancomycin resistance, D-Ala-D-Ser, D-Ala-Lac ligases',\n",
       " 'Vanillate and syringate utilization',\n",
       " 'Vibrioferrin synthesis',\n",
       " 'Vir-like type 4 secretion system',\n",
       " 'VraTSR and LiaFSR three-component regulatory systems',\n",
       " 'Wall polysaccharide pyruvylation',\n",
       " 'Widespread colonization island',\n",
       " 'Xanthine dehydrogenase subunits',\n",
       " 'YoeB-YefM toxin-antitoxin system',\n",
       " 'YrdC-YciO-Sua5 and associated protein families',\n",
       " 'ZraPRS-based zinc resistance',\n",
       " 'alpha-glucoside ABC transporter',\n",
       " 'beta-Barrel Assembly Machinery',\n",
       " 'dTDP-3-acetamido-3,6-dideoxy-alpha-D-galactose synthesis',\n",
       " 'dTDP-rhamnose synthesis',\n",
       " 'dTDP-rhamnose synthesis -- gjo',\n",
       " 'mRNA 3-end processing factor containing cluster',\n",
       " 'p-Aminobenzoyl-Glutamate Utilization',\n",
       " 'rRNA modification related cluster including tlyA',\n",
       " 'tRNA aminoacylation, Ala',\n",
       " 'tRNA aminoacylation, Arg',\n",
       " 'tRNA aminoacylation, Asn',\n",
       " 'tRNA aminoacylation, Asp',\n",
       " 'tRNA aminoacylation, Cys',\n",
       " 'tRNA aminoacylation, Gln',\n",
       " 'tRNA aminoacylation, Glu',\n",
       " 'tRNA aminoacylation, Gly',\n",
       " 'tRNA aminoacylation, His',\n",
       " 'tRNA aminoacylation, Ile',\n",
       " 'tRNA aminoacylation, Leu',\n",
       " 'tRNA aminoacylation, Lys',\n",
       " 'tRNA aminoacylation, Met',\n",
       " 'tRNA aminoacylation, Phe',\n",
       " 'tRNA aminoacylation, Pro',\n",
       " 'tRNA aminoacylation, Ser',\n",
       " 'tRNA aminoacylation, Thr',\n",
       " 'tRNA aminoacylation, Trp',\n",
       " 'tRNA aminoacylation, Tyr',\n",
       " 'tRNA aminoacylation, Val',\n",
       " 'tRNA modification, 5-oxyacetyl uridine (cmo5U34)',\n",
       " 'tRNA nucleotidyltransferase',\n",
       " 'tRNA splicing',\n",
       " 'tRNA thiolation',\n",
       " 'Aegiribacteria',\n",
       " 'Kapaibacterium',\n",
       " 'Stahlbacteria',\n",
       " 'Vermiphilus',\n",
       " 'Babela',\n",
       " 'Chromulinivorax',\n",
       " 'Ozemobacter',\n",
       " 'Desulfacyla',\n",
       " 'Dissulfurimicrobium',\n",
       " 'Dissulfurirhabdus',\n",
       " 'Bradymonas',\n",
       " 'Persicimonas',\n",
       " 'Lujinxingia',\n",
       " 'Microvenator',\n",
       " 'Acididesulfobacter',\n",
       " 'Acidulodesulfobacterium',\n",
       " 'Abditibacterium',\n",
       " 'Chloracidobacterium',\n",
       " 'Polarisedimenticola',\n",
       " 'Acanthopleuribacter',\n",
       " 'Sulfidibacter',\n",
       " 'Geothrix',\n",
       " 'Holophaga',\n",
       " 'Thermotomaculum',\n",
       " 'Angelobacter',\n",
       " 'Bryobacter',\n",
       " 'Paludibaculum',\n",
       " 'Solibacter',\n",
       " 'Sulfopaludibacter',\n",
       " 'Acidoferrum',\n",
       " 'Acidicapsa',\n",
       " 'Acidipila',\n",
       " 'Acidisarcina',\n",
       " 'Acidobacterium',\n",
       " 'Bryocella',\n",
       " 'Sulfotelmatomonas',\n",
       " 'Edaphobacter',\n",
       " 'Granulicella',\n",
       " 'Occallatibacter',\n",
       " 'Paracidobacterium',\n",
       " 'Pseudacidobacterium',\n",
       " 'Silvibacterium',\n",
       " 'Terracidiphilus',\n",
       " 'Terriglobus',\n",
       " 'Korobacter',\n",
       " 'Sulfotelmatobacter',\n",
       " 'Sulfomarinibacter',\n",
       " 'Thermoanaerobaculum',\n",
       " 'Luteitalea',\n",
       " 'Hakubella',\n",
       " 'Acidiferrimicrobium',\n",
       " 'Acidimicrobium',\n",
       " 'Aciditerrimonas',\n",
       " 'Acidithrix',\n",
       " 'Ferrimicrobium',\n",
       " 'Ferrithrix',\n",
       " 'Actinomarinicola',\n",
       " 'Aquihabitans',\n",
       " 'Dermatobacter',\n",
       " 'Iamia',\n",
       " 'Rhabdothermincola',\n",
       " 'Desertimonas',\n",
       " 'Ilumatobacter',\n",
       " 'Microthrix',\n",
       " 'Halopolyspora',\n",
       " 'Acidothermus',\n",
       " 'Actinobaculum',\n",
       " 'Actinomyces',\n",
       " 'Actinotignum',\n",
       " 'Ancrocorticia',\n",
       " 'Arcanobacterium',\n",
       " 'Boudabousia',\n",
       " 'Bowdeniella',\n",
       " 'Buchananella',\n",
       " 'Changpingibacter',\n",
       " 'Flaviflexus',\n",
       " 'Gleimia',\n",
       " 'Mobiluncus',\n",
       " 'Nanchangia',\n",
       " 'Neoactinobaculum',\n",
       " 'Pauljensenia',\n",
       " 'Peptidiphaga',\n",
       " 'Schaalia',\n",
       " 'Scrofimicrobium',\n",
       " 'Trueperella',\n",
       " 'Varibaculum',\n",
       " 'Winkia',\n",
       " 'Actinopolyspora',\n",
       " 'Aeriscardovia',\n",
       " 'Alloscardovia',\n",
       " 'Bifidobacterium',\n",
       " 'Bombiscardovia',\n",
       " 'Galliscardovia',\n",
       " 'Gardnerella',\n",
       " 'Parascardovia',\n",
       " 'Pseudoscardovia',\n",
       " 'Scardovia',\n",
       " 'Planktophila',\n",
       " 'Actinocrinis',\n",
       " 'Actinospica',\n",
       " 'Catenulispora',\n",
       " 'Cryptosporangium',\n",
       " 'Frankia',\n",
       " 'Parafrankia',\n",
       " 'Protofrankia',\n",
       " 'Pseudofrankia',\n",
       " 'Antricoccus',\n",
       " 'Blastococcus',\n",
       " 'Cumulibacter',\n",
       " 'Geodermatophilus',\n",
       " 'Goekera',\n",
       " 'Klenkia',\n",
       " 'Modestobacter',\n",
       " 'Trujillonella',\n",
       " 'Glycomyces',\n",
       " 'Haloglycomyces',\n",
       " 'Natronoglycomyces',\n",
       " 'Stackebrandtia',\n",
       " 'Jatrophihabitans',\n",
       " 'Haloactinopolyspora',\n",
       " 'Jiangella',\n",
       " 'Phytoactinopolyspora',\n",
       " 'Angustibacter',\n",
       " 'Kineococcus',\n",
       " 'Kineosporia',\n",
       " 'Pseudokineococcus',\n",
       " 'Quadrisphaera',\n",
       " 'Carbonactinospora',\n",
       " 'Actinacidiphila',\n",
       " 'Allostreptomyces',\n",
       " 'Embleya',\n",
       " 'Kitasatospora',\n",
       " 'Mangrovactinospora',\n",
       " 'Peterkaempfera',\n",
       " 'Phaeacidiphilus',\n",
       " 'Streptacidiphilus',\n",
       " 'Streptantibioticus',\n",
       " 'Streptomyces',\n",
       " 'Wenjunlia',\n",
       " 'Yinghuangia',\n",
       " 'Luteimicrobium',\n",
       " 'Pengzhenrongella',\n",
       " 'Timonella',\n",
       " 'Beutenbergia',\n",
       " 'Miniimonas',\n",
       " 'Salana',\n",
       " 'Serinibacter',\n",
       " 'Bogoriella',\n",
       " 'Georgenia',\n",
       " 'Brevibacterium',\n",
       " 'Sediminivirga',\n",
       " 'Spelaeicoccus',\n",
       " 'Actinotalea',\n",
       " 'Cellulomonas',\n",
       " 'Oerskovia',\n",
       " 'Paraoerskovia',\n",
       " 'Pseudactinotalea',\n",
       " 'Sediminihabitans',\n",
       " 'Demequina',\n",
       " 'Brachybacterium',\n",
       " 'Dermabacter',\n",
       " 'Devriesea',\n",
       " 'Helcobacillus',\n",
       " 'Allobranchiibius',\n",
       " 'Barrientosiimonas',\n",
       " 'Branchiibius',\n",
       " 'Calidifontibacter',\n",
       " 'Demetria',\n",
       " 'Dermacoccus',\n",
       " 'Flexivirga',\n",
       " 'Leekyejoonella',\n",
       " 'Luteipulveratus',\n",
       " 'Metallococcus',\n",
       " 'Rudaeicoccus',\n",
       " 'Yimella',\n",
       " 'Agilicoccus',\n",
       " 'Austwickia',\n",
       " 'Dermatophilus',\n",
       " 'Gephyromycinifex',\n",
       " 'Kineosphaera',\n",
       " 'Mobilicoccus',\n",
       " 'Piscicoccus',\n",
       " 'Arsenicicoccus',\n",
       " 'Humibacillus',\n",
       " 'Intrasporangium',\n",
       " 'Janibacter',\n",
       " 'Knoellia',\n",
       " 'Kribbia',\n",
       " 'Lapillicoccus',\n",
       " 'Marihabitans',\n",
       " 'Ornithinibacter',\n",
       " 'Ornithinicoccus',\n",
       " 'Oryzihumus',\n",
       " 'Pedococcus',\n",
       " 'Phycicoccus',\n",
       " 'Segeticoccus',\n",
       " 'Terrabacter',\n",
       " 'Terracoccus',\n",
       " 'Tetrasphaera',\n",
       " 'Flavimobilis',\n",
       " 'Jonesia',\n",
       " 'Kytococcus',\n",
       " 'Agreia',\n",
       " 'Agrococcus',\n",
       " 'Agromyces',\n",
       " 'Alpinimonas',\n",
       " 'Amnibacterium',\n",
       " 'Aquiluna',\n",
       " 'Aurantimicrobium',\n",
       " 'Lumbricidophila',\n",
       " 'Canibacter',\n",
       " 'Clavibacter',\n",
       " 'Cnuibacter',\n",
       " 'Compostimonas',\n",
       " 'Conyzicola',\n",
       " 'Cryobacterium',\n",
       " 'Curtobacterium',\n",
       " 'Diaminobutyricibacter',\n",
       " 'Diaminobutyricimonas',\n",
       " 'Frigoribacterium',\n",
       " 'Frondihabitans',\n",
       " 'Galbitalea',\n",
       " 'Glaciibacter',\n",
       " 'Glaciihabitans',\n",
       " 'Gryllotalpicola',\n",
       " 'Gulosibacter',\n",
       " 'Herbiconiux',\n",
       " 'Homoserinibacter',\n",
       " 'Homoserinimonas',\n",
       " 'Humibacter',\n",
       " 'Klugiella',\n",
       " ...]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_combined.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a29357e3-b502-41c1-94c7-2bb9ca8e7b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = smf.mixedlm(\n",
    "    'N12M_Pseudomonas_aeruginosa ~ abundance + Pseudomonas_Culture + CS_Pseudomonas_aeruginosa', \n",
    "    merged_data,\n",
    "    groups=merged_data[\"pwCF_ID\"] \n",
    ")\n",
    "result = model.fit()\n",
    "\n",
    "# Print the model summary\n",
    "print(result.summary())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ca2cc9f-0782-415b-a47d-52cbe48fedaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "if False:\n",
    "    model = smf.mixedlm(\n",
    "        'abundance ~ CS_Pseudomonas_aeruginosa + taxa_functions', \n",
    "        merged_data,\n",
    "        groups=merged_data[\"pwCF_ID\"] \n",
    "    )\n",
    "    result = model.fit()\n",
    "\n",
    "    # Print the model summary\n",
    "    print(result.summary())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce94b2ee-0be5-4869-9769-08caf6a6789d",
   "metadata": {},
   "source": [
    "## Write the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f413e87e-e2d0-4b90-8bbf-cdb0509b531b",
   "metadata": {},
   "outputs": [],
   "source": [
    "coefficients = pd.DataFrame({\n",
    "    \"Variable\": result.params.index,\n",
    "    \"Estimate\": result.params.values,\n",
    "    \"Std Error\": result.bse.values,\n",
    "    \"P-value\": result.pvalues.values\n",
    "})\n",
    "coefficients.to_csv(os.path.join('lmm', f'{taxa}_model_coefficients.tsv'), sep=\"\\t\", index=False)\n",
    "\n",
    "with open(os.path.join('lmm', f'{taxa}_model_results.tsv'), 'w') as f:\n",
    "    f.write(result.summary().as_text())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1a55639-6ff9-4e37-8dd8-4bdd90e575bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "# Extract fixed effect coefficients and confidence intervals\n",
    "params = result.fe_params  # Fixed effect coefficients\n",
    "conf_int = result.conf_int()  # Confidence intervals\n",
    "conf_int.columns = ['lower', 'upper']\n",
    "\n",
    "# Combine coefficients and confidence intervals\n",
    "coefficients = pd.DataFrame({\n",
    "    'coef': params,\n",
    "    'lower': conf_int['lower'],\n",
    "    'upper': conf_int['upper']\n",
    "})\n",
    "\n",
    "# Plot the coefficients with error bars\n",
    "plt.figure(figsize=(8, 6))\n",
    "plt.errorbar(coefficients.index, coefficients['coef'], \n",
    "             yerr=(coefficients['coef'] - coefficients['lower'], coefficients['upper'] - coefficients['coef']), \n",
    "             fmt='o', capsize=5)\n",
    "plt.axhline(0, color='gray', linestyle='--', linewidth=1)\n",
    "plt.xticks(rotation=45)\n",
    "plt.title(\"Fixed Effect Coefficients with Confidence Intervals\")\n",
    "plt.ylabel(\"Coefficient Value\")\n",
    "plt.xlabel(\"Predictor\")\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4e2dacf-2866-4264-aad6-71b688c12fdc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bioinformatics]",
   "language": "python",
   "name": "conda-env-bioinformatics-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
