{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a7a213c0-97e1-4338-b943-3166b9e1f91a",
   "metadata": {},
   "source": [
    "# Linear Mixed Models\n",
    "\n",
    "Build some linear mixed models about our data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b54c76e8-0aaf-4da7-be76-3b56a655b8d2",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "780ba312-1f7d-4751-b01f-0718dc08222a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import re\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.collections import PatchCollection\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.dates as mdates\n",
    "from matplotlib.colors import ListedColormap\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import json\n",
    "\n",
    "from itertools import cycle\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder\n",
    "from sklearn.inspection import permutation_importance\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier\n",
    "from sklearn.metrics import mean_squared_error\n",
    "\n",
    "from scipy.stats import linregress\n",
    "\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "\n",
    "# there is a FutureWarning in sklearn StandardScalar which is really annoying. This ignores it.\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "try:\n",
    "  import google.colab\n",
    "  IN_COLAB = True\n",
    "  !pip install adjustText\n",
    "  from google.colab import drive\n",
    "  drive.mount('/content/drive')\n",
    "  datadir = '/content/drive/MyDrive/Projects/CF/Adelaide/CF_Data_Analysis'\n",
    "except ImportError:\n",
    "  IN_COLAB = False\n",
    "  datadir = '..'\n",
    "\n",
    "\n",
    "sys.path.append('..')\n",
    "import cf_analysis_lib\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d320c78c-0cb0-4bcf-af63-3bb8668872dc",
   "metadata": {},
   "source": [
    "## Read the data frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ecd087d9-65e7-4859-801f-019893047960",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The subsystems df has shape: (127, 769)\n",
      "The taxonomy df has shape: (127, 3581)\n",
      "The metadata df has shape: (127, 166)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting IP vs OP to category\n",
      "Setting Hospital to category\n",
      "Setting Age groups to category\n",
      "Setting Paediatric vs Adult to category\n",
      "Setting Gender to category\n",
      "Setting Sample_Type to category\n",
      "Setting NTM to category\n",
      "Setting FEV1_RATIO_SCORE to category\n",
      "Setting FEV1_Obstruction_Rank to category\n",
      "Setting Cystic Fibrosis related diabetes (CFRD) to category\n",
      "Setting Pancreatic insufficiency (PI) to category\n",
      "Setting CF gene 1 to category\n",
      "Setting CFLD to category\n",
      "Setting CS_mucoid to category\n",
      "Setting CS_non-mucoid to category\n",
      "Setting CS_Pseudomonas aeruginosa to category\n",
      "Setting CS_Oral flora to category\n",
      "Setting CS_Stenophotomonas maltophilia to category\n",
      "Setting CS_Aspergillus fumigatus to category\n",
      "Setting CS_Aspergillus flavus to category\n",
      "Setting CS_Candida albicans to category\n",
      "Setting CS_Mycobacteroides abscessus to category\n",
      "Setting CS_Mycobacterium intracellulare to category\n",
      "Setting CS_Staphylococcus  aureus to category\n",
      "Setting CS_Inquilinus limosus to category\n",
      "Setting CS_Achromobacter xylosoxidans to category\n",
      "Setting CS_Burkholderia cepacia to category\n",
      "Setting CS_NTM_(Smear negative) to category\n",
      "Setting CS_Mycolicibacter terrae to category\n",
      "Setting CS_Aspergillus nidulans to category\n",
      "Setting CS_MAC_(Smear negative) to category\n",
      "Setting CS_Penicillium to category\n",
      "Setting CS_Aspergillus niger to category\n",
      "Setting CS_Aspergillus terreus to category\n",
      "Setting CS_Scedosporium apiospermum to category\n",
      "Setting CS_Lomentospora prolificans to category\n",
      "Setting CS_Acremonium species to category\n",
      "Setting CS_MDR_Pseudomonas aeruginosa to category\n",
      "Setting CS_Haemophilus influenzae to category\n",
      "Setting P12M_0_Nothing to category\n",
      "Setting P12M_1_Staphylococcus aureus to category\n",
      "Setting P12M_2_Pseudomonas aeruginosa to category\n",
      "Setting P12M_3_Stenotrophomonas maltophilia to category\n",
      "Setting P12M_4_Rapid NTM to category\n",
      "Setting P12M_5_Slow NTM to category\n",
      "Setting P12M_6_Aspergillus to category\n",
      "Setting P12M_7_Burkholderia cepacia to category\n",
      "Setting P12M_8_Haemophilus influenza to category\n",
      "Setting P12M_9_Streptococcus pneumoniae to category\n",
      "Setting P12M_10_Achromobacter xylosoxidans to category\n",
      "Setting P12M_11_Serratia marcesens to category\n",
      "Setting P12M_12_Inquilinus limosus to category\n",
      "Setting P12M_13_Scedosporum apiospermum to category\n",
      "Setting P12M_14_Moraxella catarrhalis to category\n",
      "Setting Positive for PA within next 12 months to category\n",
      "Setting N12M_Pseudomonas aeruginosa to category\n",
      "Setting N12M_Pseudomonas aeruginosa (multi-resistant) to category\n",
      "Setting N12M_mucoid to category\n",
      "Setting N12M_non_mucoid to category\n",
      "Setting N12M_MAC to category\n",
      "Setting N12M_M intracellulare to category\n",
      "Setting N12M_Candida albicans to category\n",
      "Setting N12M_Achromobacter xylosoxidans to category\n",
      "Setting N12M_Aspergillus fumigatus to category\n",
      "Setting N12M_Stenotrophomonas maltophilia to category\n",
      "Setting N12M_Staphylococcus aureus to category\n",
      "Setting N12M_M gordonae to category\n",
      "Setting N12M_M chelonae to category\n",
      "Setting O_Penicillium species to category\n",
      "Setting O_Enterobacter cloacae to category\n",
      "Setting O_Aspergillus Flavus to category\n",
      "Setting O_Mycobacterium abscessus to category\n",
      "Setting O_Cupriavidus gilardii to category\n",
      "Setting O_Mycobacterium intracellular to category\n",
      "Setting O_Scedosporium apiospermum to category\n",
      "Setting O_Pseudomonas aeruginosa (multi-resistant) to category\n",
      "Setting O_Streptococcus pyogenes to category\n",
      "Setting O_Pseudomonas aeruginosa (non-mucoid) to category\n",
      "Setting O_Chryseobacterium indologenes to category\n",
      "Setting O_Bordetella species to category\n",
      "Setting O_Klebsiella oxytoca to category\n",
      "Setting O_Mycobacterium mucogenicum to category\n",
      "Setting O_Mycobacterium scotochromogen to category\n",
      "Setting O_Acinetobacter species to category\n",
      "Setting O_Enterobacter species to category\n",
      "Setting O_Aspergillus niger to category\n",
      "Setting O_Aspergillus terreus to category\n",
      "Setting O_Pseudomonas aeruginosa (mucoid) to category\n",
      "Setting O_Proteus mirabilis to category\n",
      "Setting O_talaromyces purpurogenus to category\n",
      "Setting O_Trichosporon mycotoxinivorans to category\n",
      "Setting Antibiotics_YN to category\n",
      "Setting IV to category\n",
      "Setting INH to category\n",
      "Setting PO to category\n",
      "Setting 2 Amphotericin B (Ambisome)_INH to category\n",
      "Setting 4 Amikacin_IV to category\n",
      "Setting 2 Amikacin_INH to category\n",
      "Setting 2 Amoxicillin & Potassium clavulanate (Aug Duo)_PO to category\n",
      "Setting 3 Azithromycin_IV to category\n",
      "Setting 2 Azithromycin_PO to category\n",
      "Setting 3 Aztreonam_IV to category\n",
      "Setting 3 Cefopime_IV to category\n",
      "Setting 4 Cefoxitin_IV to category\n",
      "Setting 3 Ceftazidime_IV to category\n",
      "Setting 2 Ceftazidime_INH to category\n",
      "Setting 1 Cephalexin_PO to category\n",
      "Setting 2 Ciprofloxacin_PO to category\n",
      "Setting 2 Clarithromycin_PO to category\n",
      "Setting 2 Clofazimine PO to category\n",
      "Setting 4 Colistin_IV to category\n",
      "Setting 2 Colistin_IHN to category\n",
      "Setting 1 Flucloaxcillin_PO to category\n",
      "Setting 3 Imipenem to category\n",
      "Setting 1 Itraconazole (Lozenoc)_PO to category\n",
      "Setting 3 Ivacaftor (Kalydeco) to category\n",
      "Setting 3 Meropenem_IV to category\n",
      "Setting 3 Methylpredinosolone_IV to category\n",
      "Setting 3 Omalizumab_SC to category\n",
      "Setting 2 prednisolone_PO to category\n",
      "Setting 1 Sulfamethoxazole_trimethoprim (Bactrim)_PO to category\n",
      "Setting 3 tobramycin_IV to category\n",
      "Setting 2 tobramycin_INH to category\n",
      "Setting Sum of antibiotics to category\n",
      "Setting Sum of meds to category\n",
      "Setting Sum of antifungals to category\n",
      "Setting Sum of steroid + mabs to category\n"
     ]
    }
   ],
   "source": [
    "sequence_type = \"MGI\"\n",
    "datadir = '..'\n",
    "#sslevel = 'level2_norm_ss.tsv.gz'\n",
    "sslevel = 'subsystems_norm_ss.tsv.gz'\n",
    "ss_df = cf_analysis_lib.read_subsystems(os.path.join(datadir, sequence_type, \"FunctionalAnalysis\", \"subsystems\", sslevel), sequence_type)\n",
    "ss_df = ss_df.T\n",
    "print(f\"The subsystems df has shape: {ss_df.shape}\")\n",
    "\n",
    "taxa = \"genus\"\n",
    "genus_otu = cf_analysis_lib.read_taxonomy(datadir, sequence_type, taxa)\n",
    "genus_otu = genus_otu.T\n",
    "print(f\"The taxonomy df has shape: {genus_otu.shape}\")\n",
    "metadata = cf_analysis_lib.read_metadata(datadir, sequence_type)\n",
    "print(f\"The metadata df has shape: {metadata.shape}\")\n",
    "\n",
    "df = ss_df.merge(genus_otu, left_index=True, right_index=True, how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "28c279d2-eb7f-4261-86cf-c8eeb1f80a0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "minion                                       object\n",
       "MGI                                          object\n",
       "pwCF_ID                                       int64\n",
       "Sample date                                  object\n",
       "IP vs OP                                   category\n",
       "                                             ...   \n",
       "DNA Conc. (ng/ul)                           float64\n",
       "Index I7                                     object\n",
       "Index I5                                     object\n",
       "Mean_Size_BP                                  int64\n",
       "Total Clusters Passing Filter (Million)     float64\n",
       "Length: 166, dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a1ca03d-45b4-4e5e-bba4-5d95fbf163ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "minion                                       object\n",
       "MGI                                          object\n",
       "pwCF_ID                                       int64\n",
       "Sample date                                  object\n",
       "IP vs OP                                   category\n",
       "                                             ...   \n",
       "Index I7                                     object\n",
       "Index I5                                     object\n",
       "Mean_Size_BP                                  int64\n",
       "Total Clusters Passing Filter (Million)     float64\n",
       "IPc                                        category\n",
       "Length: 167, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['IPc'] = metadata['IP vs OP'].astype('category')\n",
    "metadata.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ad806f22-9f9b-4b8f-9c97-3163ca6a345e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Column header': 'Type of data',\n",
       " 'NAME': 'Text',\n",
       " 'minion': 'Text',\n",
       " 'MGI': 'Text',\n",
       " 'pwCF_ID': 'Text',\n",
       " 'Sample date': 'Date',\n",
       " 'IP vs OP': 'Categorical',\n",
       " 'Hospital': 'Categorical',\n",
       " 'Room': 'Text',\n",
       " 'Age': 'Numeric',\n",
       " 'Age groups': 'Categorical',\n",
       " 'Paediatric vs Adult': 'Categorical',\n",
       " 'Gender': 'Categorical',\n",
       " 'Sample_Type': 'Categorical',\n",
       " 'H2_Uncorrected': 'Numeric',\n",
       " 'CH4_Uncorrected': 'Numeric',\n",
       " 'CO2': 'Numeric',\n",
       " 'H2_Corrected': 'Numeric',\n",
       " 'CH4_Corrected': 'Numeric',\n",
       " 'CH4/H2 ratio_corrected': 'Numeric',\n",
       " 'Corr.': 'Numeric',\n",
       " 'Culture Result (Matched with sequenced sample)': 'Text',\n",
       " 'NTM': 'Categorical',\n",
       " 'Pseudomonas': 'Categorical',\n",
       " 'Cutured in previous 12 months': 'Text',\n",
       " 'Others cultured': 'Text',\n",
       " 'IgE': 'Numeric',\n",
       " 'Spec IgE': 'Numeric',\n",
       " 'Spec IgG': 'Numeric',\n",
       " 'Precipitins': 'Numeric',\n",
       " 'FVC': 'Numeric',\n",
       " 'FEV1': 'Numeric',\n",
       " 'Best FEV1': 'Numeric',\n",
       " 'FEV1/best FEV1': 'Numeric',\n",
       " 'FEV1_RATIO_SCORE': 'Categorical',\n",
       " 'FEV1_Obstruction_Rank': 'Categorical',\n",
       " 'Cystic Fibrosis related diabetes (CFRD)': 'Categorical',\n",
       " 'Pancreatic insufficiency (PI)': 'Categorical',\n",
       " 'CF gene 1': 'Categorical',\n",
       " 'CF gene 2': 'Text',\n",
       " 'Notes': 'Text',\n",
       " 'CFLD': 'Categorical',\n",
       " 'CS_mucoid': 'Categorical',\n",
       " 'CS_non-mucoid': 'Categorical',\n",
       " 'CS_Pseudomonas aeruginosa': 'Categorical',\n",
       " 'CS_Oral flora': 'Categorical',\n",
       " 'CS_Stenophotomonas maltophilia': 'Categorical',\n",
       " 'CS_Aspergillus fumigatus': 'Categorical',\n",
       " 'CS_Aspergillus flavus': 'Categorical',\n",
       " 'CS_Candida albicans': 'Categorical',\n",
       " 'CS_Mycobacteroides abscessus': 'Categorical',\n",
       " 'CS_Mycobacterium intracellulare': 'Categorical',\n",
       " 'CS_Staphylococcus  aureus': 'Categorical',\n",
       " 'CS_Inquilinus limosus': 'Categorical',\n",
       " 'CS_Achromobacter xylosoxidans': 'Categorical',\n",
       " 'CS_Burkholderia cepacia': 'Categorical',\n",
       " 'CS_NTM_(Smear negative)': 'Categorical',\n",
       " 'CS_Mycolicibacter terrae': 'Categorical',\n",
       " 'CS_Aspergillus nidulans': 'Categorical',\n",
       " 'CS_MAC_(Smear negative)': 'Categorical',\n",
       " 'CS_Penicillium': 'Categorical',\n",
       " 'CS_Aspergillus niger': 'Categorical',\n",
       " 'CS_Aspergillus terreus': 'Categorical',\n",
       " 'CS_Lomentospora prolificans': 'Categorical',\n",
       " 'CS_Acremonium species': 'Categorical',\n",
       " 'CS_MDR_Pseudomonas aeruginosa': 'Categorical',\n",
       " 'CS_Haemophilus influenzae': 'Categorical',\n",
       " 'CS_Scedosporium apiospermum': 'Categorical',\n",
       " 'P12M_0_Nothing': 'Categorical',\n",
       " 'P12M_1_Staphylococcus aureus': 'Categorical',\n",
       " 'P12M_2_Pseudomonas aeruginosa': 'Categorical',\n",
       " 'P12M_3_Stenotrophomonas maltophilia': 'Categorical',\n",
       " 'P12M_4_Rapid NTM': 'Categorical',\n",
       " 'P12M_5_Slow NTM': 'Categorical',\n",
       " 'P12M_6_Aspergillus': 'Categorical',\n",
       " 'P12M_7_Burkholderia cepacia': 'Categorical',\n",
       " 'P12M_8_Haemophilus influenza': 'Categorical',\n",
       " 'P12M_9_Streptococcus pneumoniae': 'Categorical',\n",
       " 'P12M_10_Achromobacter xylosoxidans': 'Categorical',\n",
       " 'P12M_11_Serratia marcesens': 'Categorical',\n",
       " 'P12M_12_Inquilinus limosus': 'Categorical',\n",
       " 'P12M_13_Scedosporum apiospermum': 'Categorical',\n",
       " 'P12M_14_Moraxella catarrhalis': 'Categorical',\n",
       " 'Next PA Positive Date': 'Date',\n",
       " 'Positive for PA within next 12 months': 'Categorical',\n",
       " 'Culture for next sample': 'Text',\n",
       " 'N12M_Pseudomonas aeruginosa': 'Categorical',\n",
       " 'N12M_Pseudomonas aeruginosa (multi-resistant)': 'Categorical',\n",
       " 'N12M_mucoid': 'Categorical',\n",
       " 'N12M_non_mucoid': 'Categorical',\n",
       " 'N12M_MAC': 'Categorical',\n",
       " 'N12M_M intracellulare': 'Categorical',\n",
       " 'N12M_Candida albicans': 'Categorical',\n",
       " 'N12M_Achromobacter xylosoxidans': 'Categorical',\n",
       " 'N12M_Aspergillus fumigatus': 'Categorical',\n",
       " 'N12M_Stenotrophomonas maltophilia': 'Categorical',\n",
       " 'N12M_Staphylococcus aureus': 'Categorical',\n",
       " 'N12M_M gordonae': 'Categorical',\n",
       " 'N12M_M chelonae': 'Categorical',\n",
       " 'O_Penicillium species': 'Categorical',\n",
       " 'O_Enterobacter cloacae': 'Categorical',\n",
       " 'O_Aspergillus Flavus': 'Categorical',\n",
       " 'O_Mycobacterium abscessus': 'Categorical',\n",
       " 'O_Cupriavidus gilardii': 'Categorical',\n",
       " 'O_Mycobacterium intracellular': 'Categorical',\n",
       " 'O_Scedosporium apiospermum': 'Categorical',\n",
       " 'O_Pseudomonas aeruginosa (multi-resistant)': 'Categorical',\n",
       " 'O_Streptococcus pyogenes': 'Categorical',\n",
       " 'O_Pseudomonas aeruginosa (non-mucoid)': 'Categorical',\n",
       " 'O_Chryseobacterium indologenes': 'Categorical',\n",
       " 'O_Bordetella species': 'Categorical',\n",
       " 'O_Klebsiella oxytoca': 'Categorical',\n",
       " 'O_Mycobacterium mucogenicum': 'Categorical',\n",
       " 'O_Mycobacterium scotochromogen': 'Categorical',\n",
       " 'O_Acinetobacter species': 'Categorical',\n",
       " 'O_Enterobacter species': 'Categorical',\n",
       " 'O_Aspergillus niger': 'Categorical',\n",
       " 'O_Aspergillus terreus': 'Categorical',\n",
       " 'O_Pseudomonas aeruginosa (mucoid)': 'Categorical',\n",
       " 'O_Proteus mirabilis': 'Categorical',\n",
       " 'O_talaromyces purpurogenus': 'Categorical',\n",
       " 'O_Trichosporon mycotoxinivorans': 'Categorical',\n",
       " 'Antibiotics (duration)': 'Text',\n",
       " 'Antibiotics_YN': 'Categorical',\n",
       " 'IV': 'Categorical',\n",
       " 'INH': 'Categorical',\n",
       " 'PO': 'Categorical',\n",
       " '2 Amphotericin B (Ambisome)_INH': 'Categorical',\n",
       " '4 Amikacin_IV': 'Categorical',\n",
       " '2 Amikacin_INH': 'Categorical',\n",
       " '2 Amoxicillin & Potassium clavulanate (Aug Duo)_PO': 'Categorical',\n",
       " '3 Azithromycin_IV': 'Categorical',\n",
       " '2 Azithromycin_PO': 'Categorical',\n",
       " '3 Aztreonam_IV': 'Categorical',\n",
       " '3 Cefopime_IV': 'Categorical',\n",
       " '4 Cefoxitin_IV': 'Categorical',\n",
       " '3 Ceftazidime_IV': 'Categorical',\n",
       " '2 Ceftazidime_INH': 'Categorical',\n",
       " '1 Cephalexin_PO': 'Categorical',\n",
       " '2 Ciprofloxacin_PO': 'Categorical',\n",
       " '2 Clarithromycin_PO': 'Categorical',\n",
       " '2 Clofazimine PO': 'Categorical',\n",
       " '4 Colistin_IV': 'Categorical',\n",
       " '2 Colistin_IHN': 'Categorical',\n",
       " '1 Flucloaxcillin_PO': 'Categorical',\n",
       " '3 Imipenem': 'Categorical',\n",
       " '1 Itraconazole (Lozenoc)_PO': 'Categorical',\n",
       " '3 Ivacaftor (Kalydeco)': 'Categorical',\n",
       " '3 Meropenem_IV': 'Categorical',\n",
       " '3 Methylpredinosolone_IV': 'Categorical',\n",
       " '3 Omalizumab_SC': 'Categorical',\n",
       " '\"3 piperacillin sodium, tazobactam sodium (Tazocin)_IV\"': 'Categorical',\n",
       " '2 prednisolone_PO': 'Categorical',\n",
       " '1 Sulfamethoxazole_trimethoprim (Bactrim)_PO': 'Categorical',\n",
       " '3 tobramycin_IV': 'Categorical',\n",
       " '2 tobramycin_INH': 'Categorical',\n",
       " 'Sum of antibiotics': 'Categorical',\n",
       " 'Sum of meds': 'Categorical',\n",
       " 'Sum of antifungals': 'Categorical',\n",
       " 'Sum of steroid + mabs': 'Categorical',\n",
       " 'DNA_extraction_ conc': 'Numeric',\n",
       " 'SAGC ULN': 'Text',\n",
       " 'DNA Conc. (ng/ul)': 'Numeric',\n",
       " 'Index I7': 'Text',\n",
       " 'Index I5': 'Text',\n",
       " 'Mean_Size_BP': 'Numeric',\n",
       " 'Total Clusters Passing Filter (Million)': 'Numeric'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "mdx_types = cf_analysis_lib.metadata_definitions()\n",
    "mdx_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11120eb9-bc3a-41b9-baaf-e2eafc646602",
   "metadata": {},
   "source": [
    "## Get rid of the spaces!\n",
    "\n",
    "For statsmodels, it really helps if we don't have spaces. We can get rid of them here and then things become easier later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b4b9a16f-6b7c-4736-9950-cf8b19091b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dcolumns_no_spaces = {}\n",
    "for c in df.columns:\n",
    "    if ' ' in c:\n",
    "        dcolumns_no_spaces[c] = c.replace(' ', '_')\n",
    "df.rename(columns=dcolumns_no_spaces, inplace=True)\n",
    "\n",
    "mcolumns_no_spaces = {}\n",
    "for c in metadata.columns:\n",
    "    if ' ' in c:\n",
    "        mcolumns_no_spaces[c] = c.replace(' ', '_')\n",
    "metadata.rename(columns=mcolumns_no_spaces, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4c2d8609-5ccc-421b-bed5-032b9bf66c9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2,3-diacetamido-2,3-dideoxy-d-mannuronic_acid</th>\n",
       "      <th>2-O-alpha-mannosyl-D-glycerate_utilization</th>\n",
       "      <th>2-aminophenol_Metabolism</th>\n",
       "      <th>2-ketoacid_oxidoreductases_disambiguation</th>\n",
       "      <th>2-oxoglutarate_dehydrogenase_</th>\n",
       "      <th>2-phosphoglycolate_salvage</th>\n",
       "      <th>3-amino-5-hydroxybenzoic_Acid_Synthesis</th>\n",
       "      <th>4-hydroxybenzoyl-CoA_reductase</th>\n",
       "      <th>5-methylaminomethyl-2-thiouridine</th>\n",
       "      <th>A_Hypothetical_Protein_Related_to_Proline_Metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Oceaniferula</th>\n",
       "      <th>Persicirhabdus</th>\n",
       "      <th>Phragmitibacter</th>\n",
       "      <th>Prosthecobacter</th>\n",
       "      <th>Roseibacillus</th>\n",
       "      <th>Roseimicrobium</th>\n",
       "      <th>Sulfuriroseicoccus</th>\n",
       "      <th>Verrucomicrobium</th>\n",
       "      <th>Eremiobacter</th>\n",
       "      <th>Methylomirabilis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1068841_20180306_S</th>\n",
       "      <td>10.085904</td>\n",
       "      <td>2784.895948</td>\n",
       "      <td>516.160945</td>\n",
       "      <td>231.778018</td>\n",
       "      <td>122.573392</td>\n",
       "      <td>342.327431</td>\n",
       "      <td>783.041862</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4136.703664</td>\n",
       "      <td>827.439614</td>\n",
       "      <td>...</td>\n",
       "      <td>1.347606</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.539042</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.539042</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1447437_20171212_S</th>\n",
       "      <td>59.260325</td>\n",
       "      <td>1065.438272</td>\n",
       "      <td>543.947408</td>\n",
       "      <td>554.136026</td>\n",
       "      <td>428.171446</td>\n",
       "      <td>301.915763</td>\n",
       "      <td>679.310468</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3913.676651</td>\n",
       "      <td>732.956657</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1128691_20171206_S</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>426.619709</td>\n",
       "      <td>912.767750</td>\n",
       "      <td>49.606943</td>\n",
       "      <td>213.640568</td>\n",
       "      <td>277.798880</td>\n",
       "      <td>423.312580</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2093.412992</td>\n",
       "      <td>236.459761</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 4350 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    2,3-diacetamido-2,3-dideoxy-d-mannuronic_acid  \\\n",
       "1068841_20180306_S                                      10.085904   \n",
       "1447437_20171212_S                                      59.260325   \n",
       "1128691_20171206_S                                       0.000000   \n",
       "\n",
       "                    2-O-alpha-mannosyl-D-glycerate_utilization  \\\n",
       "1068841_20180306_S                                 2784.895948   \n",
       "1447437_20171212_S                                 1065.438272   \n",
       "1128691_20171206_S                                  426.619709   \n",
       "\n",
       "                    2-aminophenol_Metabolism  \\\n",
       "1068841_20180306_S                516.160945   \n",
       "1447437_20171212_S                543.947408   \n",
       "1128691_20171206_S                912.767750   \n",
       "\n",
       "                    2-ketoacid_oxidoreductases_disambiguation  \\\n",
       "1068841_20180306_S                                 231.778018   \n",
       "1447437_20171212_S                                 554.136026   \n",
       "1128691_20171206_S                                  49.606943   \n",
       "\n",
       "                    2-oxoglutarate_dehydrogenase_  2-phosphoglycolate_salvage  \\\n",
       "1068841_20180306_S                     122.573392                  342.327431   \n",
       "1447437_20171212_S                     428.171446                  301.915763   \n",
       "1128691_20171206_S                     213.640568                  277.798880   \n",
       "\n",
       "                    3-amino-5-hydroxybenzoic_Acid_Synthesis  \\\n",
       "1068841_20180306_S                               783.041862   \n",
       "1447437_20171212_S                               679.310468   \n",
       "1128691_20171206_S                               423.312580   \n",
       "\n",
       "                    4-hydroxybenzoyl-CoA_reductase  \\\n",
       "1068841_20180306_S                             0.0   \n",
       "1447437_20171212_S                             0.0   \n",
       "1128691_20171206_S                             0.0   \n",
       "\n",
       "                    5-methylaminomethyl-2-thiouridine  \\\n",
       "1068841_20180306_S                        4136.703664   \n",
       "1447437_20171212_S                        3913.676651   \n",
       "1128691_20171206_S                        2093.412992   \n",
       "\n",
       "                    A_Hypothetical_Protein_Related_to_Proline_Metabolism  ...  \\\n",
       "1068841_20180306_S                                         827.439614     ...   \n",
       "1447437_20171212_S                                         732.956657     ...   \n",
       "1128691_20171206_S                                         236.459761     ...   \n",
       "\n",
       "                    Oceaniferula  Persicirhabdus  Phragmitibacter  \\\n",
       "1068841_20180306_S      1.347606             0.0              0.0   \n",
       "1447437_20171212_S      0.000000             0.0              0.0   \n",
       "1128691_20171206_S      0.000000             0.0              0.0   \n",
       "\n",
       "                    Prosthecobacter  Roseibacillus  Roseimicrobium  \\\n",
       "1068841_20180306_S              0.0       0.539042             0.0   \n",
       "1447437_20171212_S              0.0       0.000000             0.0   \n",
       "1128691_20171206_S              0.0       0.000000             0.0   \n",
       "\n",
       "                    Sulfuriroseicoccus  Verrucomicrobium  Eremiobacter  \\\n",
       "1068841_20180306_S                 0.0          0.539042           0.0   \n",
       "1447437_20171212_S                 0.0          0.000000           0.0   \n",
       "1128691_20171206_S                 0.0          0.000000           0.0   \n",
       "\n",
       "                    Methylomirabilis  \n",
       "1068841_20180306_S               0.0  \n",
       "1447437_20171212_S               0.0  \n",
       "1128691_20171206_S               0.0  \n",
       "\n",
       "[3 rows x 4350 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dffe5996-5874-4ffd-8ac0-5f25f6d154b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minion</th>\n",
       "      <th>MGI</th>\n",
       "      <th>pwCF_ID</th>\n",
       "      <th>Sample_date</th>\n",
       "      <th>IP_vs_OP</th>\n",
       "      <th>Hospital</th>\n",
       "      <th>Room</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age_groups</th>\n",
       "      <th>Paediatric_vs_Adult</th>\n",
       "      <th>...</th>\n",
       "      <th>Sum_of_meds</th>\n",
       "      <th>Sum_of_antifungals</th>\n",
       "      <th>Sum_of_steroid_+_mabs</th>\n",
       "      <th>DNA_extraction__conc</th>\n",
       "      <th>SAGC_ULN</th>\n",
       "      <th>DNA_Conc._(ng/ul)</th>\n",
       "      <th>Index_I7</th>\n",
       "      <th>Index_I5</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total_Clusters_Passing_Filter_(Million)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>634207_20180510_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>634207_20180510_S</td>\n",
       "      <td>634207</td>\n",
       "      <td>5/10/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Rm9</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>SAGCFN_22_01856</td>\n",
       "      <td>7.82</td>\n",
       "      <td>CGGACGATTC</td>\n",
       "      <td>CCACCACCTA</td>\n",
       "      <td>651</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634207_20180517_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>634207_20180517_S</td>\n",
       "      <td>634207</td>\n",
       "      <td>5/17/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Rm9</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.134</td>\n",
       "      <td>SAGCFN_22_01827</td>\n",
       "      <td>22.80</td>\n",
       "      <td>AGCGATAG</td>\n",
       "      <td>CCTATCCT</td>\n",
       "      <td>633</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715927_20180205_S</th>\n",
       "      <td>715927_20180205_S</td>\n",
       "      <td>715927_20180205_S</td>\n",
       "      <td>715927</td>\n",
       "      <td>2/05/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Level 6 DK Office</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.326</td>\n",
       "      <td>SAGCFN_22_01797</td>\n",
       "      <td>16.50</td>\n",
       "      <td>TAATGCGC</td>\n",
       "      <td>AGGCGAAG</td>\n",
       "      <td>516</td>\n",
       "      <td>3.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 166 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              minion                MGI  pwCF_ID Sample_date  \\\n",
       "NAME                                                                           \n",
       "634207_20180510_S                NaN  634207_20180510_S   634207   5/10/2018   \n",
       "634207_20180517_S                NaN  634207_20180517_S   634207   5/17/2018   \n",
       "715927_20180205_S  715927_20180205_S  715927_20180205_S   715927   2/05/2018   \n",
       "\n",
       "                  IP_vs_OP Hospital               Room  Age  Age_groups  \\\n",
       "NAME                                                                      \n",
       "634207_20180510_S       IP      WCH           Adol Rm9   17           3   \n",
       "634207_20180517_S       IP      WCH           Adol Rm9   17           3   \n",
       "715927_20180205_S       OP      WCH  Level 6 DK Office   13           3   \n",
       "\n",
       "                  Paediatric_vs_Adult  ... Sum_of_meds Sum_of_antifungals  \\\n",
       "NAME                                   ...                                  \n",
       "634207_20180510_S          Paediatric  ...           1                  0   \n",
       "634207_20180517_S          Paediatric  ...           1                  0   \n",
       "715927_20180205_S          Paediatric  ...           1                  0   \n",
       "\n",
       "                   Sum_of_steroid_+_mabs  DNA_extraction__conc  \\\n",
       "NAME                                                             \n",
       "634207_20180510_S                      0                 0.000   \n",
       "634207_20180517_S                      0                 0.134   \n",
       "715927_20180205_S                      0                 0.326   \n",
       "\n",
       "                          SAGC_ULN  DNA_Conc._(ng/ul)    Index_I7    Index_I5  \\\n",
       "NAME                                                                            \n",
       "634207_20180510_S  SAGCFN_22_01856               7.82  CGGACGATTC  CCACCACCTA   \n",
       "634207_20180517_S  SAGCFN_22_01827              22.80    AGCGATAG    CCTATCCT   \n",
       "715927_20180205_S  SAGCFN_22_01797              16.50    TAATGCGC    AGGCGAAG   \n",
       "\n",
       "                   Mean_Size_BP Total_Clusters_Passing_Filter_(Million)  \n",
       "NAME                                                                     \n",
       "634207_20180510_S           651                                     2.9  \n",
       "634207_20180517_S           633                                     2.4  \n",
       "715927_20180205_S           516                                     3.4  \n",
       "\n",
       "[3 rows x 166 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "248d3941-af21-4a9a-9e7d-c2249d17ecb4",
   "metadata": {},
   "source": [
    "# Convert our metadata to categories"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "220248b4-7cea-4117-b614-42b49cba9320",
   "metadata": {},
   "source": [
    "# Predict something \"simple\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f518c0d1-7b76-4e75-ab89-fa0737849e0b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04378ee0-b921-42dc-b5ec-315d6d775f7a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3871d7a-02db-4c2f-93cc-c235324df847",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a70d41-d2c1-4dbd-935b-64e24e4c3938",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72c95614-d245-4f5a-971c-5c3e1f5004bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6919d202-c096-403f-bb39-025e8cef7a1f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f87f6a19-2d73-4881-b3cd-6afecdb67ec4",
   "metadata": {},
   "source": [
    "# Early trial\n",
    "\n",
    "Ignore the stuff down here, its probably wrong :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61367d40-baba-41fc-90af-b67e61fcfb1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minion</th>\n",
       "      <th>MGI</th>\n",
       "      <th>pwCF_ID</th>\n",
       "      <th>Sample date</th>\n",
       "      <th>IP vs OP</th>\n",
       "      <th>Hospital</th>\n",
       "      <th>Room</th>\n",
       "      <th>Age</th>\n",
       "      <th>Age groups</th>\n",
       "      <th>Paediatric vs Adult</th>\n",
       "      <th>...</th>\n",
       "      <th>Sum of meds</th>\n",
       "      <th>Sum of antifungals</th>\n",
       "      <th>Sum of steroid + mabs</th>\n",
       "      <th>DNA_extraction_ conc</th>\n",
       "      <th>SAGC ULN</th>\n",
       "      <th>DNA Conc. (ng/ul)</th>\n",
       "      <th>Index I7</th>\n",
       "      <th>Index I5</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total Clusters Passing Filter (Million)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>634207_20180510_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>634207_20180510_S</td>\n",
       "      <td>634207</td>\n",
       "      <td>5/10/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Rm9</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>SAGCFN_22_01856</td>\n",
       "      <td>7.82</td>\n",
       "      <td>CGGACGATTC</td>\n",
       "      <td>CCACCACCTA</td>\n",
       "      <td>651</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634207_20180517_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>634207_20180517_S</td>\n",
       "      <td>634207</td>\n",
       "      <td>5/17/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Rm9</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.134</td>\n",
       "      <td>SAGCFN_22_01827</td>\n",
       "      <td>22.80</td>\n",
       "      <td>AGCGATAG</td>\n",
       "      <td>CCTATCCT</td>\n",
       "      <td>633</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715927_20180205_S</th>\n",
       "      <td>715927_20180205_S</td>\n",
       "      <td>715927_20180205_S</td>\n",
       "      <td>715927</td>\n",
       "      <td>2/05/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Level 6 DK Office</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.326</td>\n",
       "      <td>SAGCFN_22_01797</td>\n",
       "      <td>16.50</td>\n",
       "      <td>TAATGCGC</td>\n",
       "      <td>AGGCGAAG</td>\n",
       "      <td>516</td>\n",
       "      <td>3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715927_20180213_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>715927_20180213_S</td>\n",
       "      <td>715927</td>\n",
       "      <td>2/13/2018</td>\n",
       "      <td>IP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>Adol Room 11</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.234</td>\n",
       "      <td>SAGCFN_22_01811</td>\n",
       "      <td>31.00</td>\n",
       "      <td>TCCGCGAA</td>\n",
       "      <td>CCTATCCT</td>\n",
       "      <td>443</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715927_20180226_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>715927_20180226_S</td>\n",
       "      <td>715927</td>\n",
       "      <td>2/26/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>WCH</td>\n",
       "      <td>OPD 8</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>Paediatric</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.108</td>\n",
       "      <td>SAGCFN_22_01833</td>\n",
       "      <td>15.10</td>\n",
       "      <td>TAACTTGGTC</td>\n",
       "      <td>GATTCACGAC</td>\n",
       "      <td>510</td>\n",
       "      <td>2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651490_20180206_S</th>\n",
       "      <td>1651490_20180206_S</td>\n",
       "      <td>1651490_20180206_S</td>\n",
       "      <td>1651490</td>\n",
       "      <td>2/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 1</td>\n",
       "      <td>27</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4.760</td>\n",
       "      <td>SAGCFN_22_01741</td>\n",
       "      <td>26.20</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>AGGCGAAG</td>\n",
       "      <td>507</td>\n",
       "      <td>6.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651490_20171215_S</th>\n",
       "      <td>1651490_20171215_S</td>\n",
       "      <td>1651490_20171215_S</td>\n",
       "      <td>1651490</td>\n",
       "      <td>12/15/2017</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 4</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7.760</td>\n",
       "      <td>SAGCFN_22_01738</td>\n",
       "      <td>34.20</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>564</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658447_20171006_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1658447_20171006_S</td>\n",
       "      <td>1658447</td>\n",
       "      <td>10/06/2017</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 3</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.098</td>\n",
       "      <td>SAGCFN_22_01837</td>\n",
       "      <td>13.10</td>\n",
       "      <td>CAGCAGGTCA</td>\n",
       "      <td>TACCTAAGTG</td>\n",
       "      <td>576</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664053_20180406_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1664053_20180406_S</td>\n",
       "      <td>1664053</td>\n",
       "      <td>4/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 1</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.159</td>\n",
       "      <td>SAGCFN_22_01822</td>\n",
       "      <td>30.80</td>\n",
       "      <td>TCTCGCGC</td>\n",
       "      <td>TAATCTTA</td>\n",
       "      <td>374</td>\n",
       "      <td>1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1690154_20180406_S</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1690154_20180406_S</td>\n",
       "      <td>1690154</td>\n",
       "      <td>4/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 4</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "      <td>Adult</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.276</td>\n",
       "      <td>SAGCFN_22_01804</td>\n",
       "      <td>35.00</td>\n",
       "      <td>CGGCTATG</td>\n",
       "      <td>GGCTCTGA</td>\n",
       "      <td>648</td>\n",
       "      <td>3.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 166 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                minion                 MGI  pwCF_ID  \\\n",
       "NAME                                                                  \n",
       "634207_20180510_S                  NaN   634207_20180510_S   634207   \n",
       "634207_20180517_S                  NaN   634207_20180517_S   634207   \n",
       "715927_20180205_S    715927_20180205_S   715927_20180205_S   715927   \n",
       "715927_20180213_S                  NaN   715927_20180213_S   715927   \n",
       "715927_20180226_S                  NaN   715927_20180226_S   715927   \n",
       "...                                ...                 ...      ...   \n",
       "1651490_20180206_S  1651490_20180206_S  1651490_20180206_S  1651490   \n",
       "1651490_20171215_S  1651490_20171215_S  1651490_20171215_S  1651490   \n",
       "1658447_20171006_S                 NaN  1658447_20171006_S  1658447   \n",
       "1664053_20180406_S                 NaN  1664053_20180406_S  1664053   \n",
       "1690154_20180406_S                 NaN  1690154_20180406_S  1690154   \n",
       "\n",
       "                   Sample date IP vs OP Hospital               Room  Age  \\\n",
       "NAME                                                                       \n",
       "634207_20180510_S    5/10/2018       IP      WCH           Adol Rm9   17   \n",
       "634207_20180517_S    5/17/2018       IP      WCH           Adol Rm9   17   \n",
       "715927_20180205_S    2/05/2018       OP      WCH  Level 6 DK Office   13   \n",
       "715927_20180213_S    2/13/2018       IP      WCH       Adol Room 11   13   \n",
       "715927_20180226_S    2/26/2018       OP      WCH              OPD 8   13   \n",
       "...                        ...      ...      ...                ...  ...   \n",
       "1651490_20180206_S   2/06/2018       OP      RAH     Chest Clinic 1   27   \n",
       "1651490_20171215_S  12/15/2017       OP      RAH     Chest Clinic 4   26   \n",
       "1658447_20171006_S  10/06/2017       OP      RAH     Chest Clinic 3   26   \n",
       "1664053_20180406_S   4/06/2018       OP      RAH     Chest Clinic 1   26   \n",
       "1690154_20180406_S   4/06/2018       OP      RAH     Chest Clinic 4   25   \n",
       "\n",
       "                    Age groups Paediatric vs Adult  ... Sum of meds  \\\n",
       "NAME                                                ...               \n",
       "634207_20180510_S            3          Paediatric  ...           1   \n",
       "634207_20180517_S            3          Paediatric  ...           1   \n",
       "715927_20180205_S            3          Paediatric  ...           1   \n",
       "715927_20180213_S            3          Paediatric  ...           3   \n",
       "715927_20180226_S            3          Paediatric  ...           2   \n",
       "...                        ...                 ...  ...         ...   \n",
       "1651490_20180206_S           5               Adult  ...           1   \n",
       "1651490_20171215_S           5               Adult  ...           1   \n",
       "1658447_20171006_S           5               Adult  ...           1   \n",
       "1664053_20180406_S           5               Adult  ...           3   \n",
       "1690154_20180406_S           5               Adult  ...           0   \n",
       "\n",
       "                   Sum of antifungals  Sum of steroid + mabs  \\\n",
       "NAME                                                           \n",
       "634207_20180510_S                   0                      0   \n",
       "634207_20180517_S                   0                      0   \n",
       "715927_20180205_S                   0                      0   \n",
       "715927_20180213_S                   0                      0   \n",
       "715927_20180226_S                   0                      0   \n",
       "...                               ...                    ...   \n",
       "1651490_20180206_S                  0                      0   \n",
       "1651490_20171215_S                  0                      0   \n",
       "1658447_20171006_S                  0                      0   \n",
       "1664053_20180406_S                  1                      0   \n",
       "1690154_20180406_S                  0                      0   \n",
       "\n",
       "                    DNA_extraction_ conc         SAGC ULN  DNA Conc. (ng/ul)  \\\n",
       "NAME                                                                           \n",
       "634207_20180510_S                  0.000  SAGCFN_22_01856               7.82   \n",
       "634207_20180517_S                  0.134  SAGCFN_22_01827              22.80   \n",
       "715927_20180205_S                  0.326  SAGCFN_22_01797              16.50   \n",
       "715927_20180213_S                  0.234  SAGCFN_22_01811              31.00   \n",
       "715927_20180226_S                  0.108  SAGCFN_22_01833              15.10   \n",
       "...                                  ...              ...                ...   \n",
       "1651490_20180206_S                 4.760  SAGCFN_22_01741              26.20   \n",
       "1651490_20171215_S                 7.760  SAGCFN_22_01738              34.20   \n",
       "1658447_20171006_S                 0.098  SAGCFN_22_01837              13.10   \n",
       "1664053_20180406_S                 0.159  SAGCFN_22_01822              30.80   \n",
       "1690154_20180406_S                 0.276  SAGCFN_22_01804              35.00   \n",
       "\n",
       "                      Index I7    Index I5  Mean_Size_BP  \\\n",
       "NAME                                                       \n",
       "634207_20180510_S   CGGACGATTC  CCACCACCTA           651   \n",
       "634207_20180517_S     AGCGATAG    CCTATCCT           633   \n",
       "715927_20180205_S     TAATGCGC    AGGCGAAG           516   \n",
       "715927_20180213_S     TCCGCGAA    CCTATCCT           443   \n",
       "715927_20180226_S   TAACTTGGTC  GATTCACGAC           510   \n",
       "...                        ...         ...           ...   \n",
       "1651490_20180206_S    ATTACTCG    AGGCGAAG           507   \n",
       "1651490_20171215_S    ATTACTCG    ATAGAGGC           564   \n",
       "1658447_20171006_S  CAGCAGGTCA  TACCTAAGTG           576   \n",
       "1664053_20180406_S    TCTCGCGC    TAATCTTA           374   \n",
       "1690154_20180406_S    CGGCTATG    GGCTCTGA           648   \n",
       "\n",
       "                   Total Clusters Passing Filter (Million)  \n",
       "NAME                                                        \n",
       "634207_20180510_S                                      2.9  \n",
       "634207_20180517_S                                      2.4  \n",
       "715927_20180205_S                                      3.4  \n",
       "715927_20180213_S                                      2.7  \n",
       "715927_20180226_S                                      2.6  \n",
       "...                                                    ...  \n",
       "1651490_20180206_S                                     6.4  \n",
       "1651490_20171215_S                                     6.0  \n",
       "1658447_20171006_S                                     2.9  \n",
       "1664053_20180406_S                                     1.9  \n",
       "1690154_20180406_S                                     3.1  \n",
       "\n",
       "[127 rows x 166 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0c7f2287-dee4-4ce4-85e3-b6bfc09e710e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df: (127, 4350)\n",
      "metadata: (127, 166)\n",
      "df_combined: (127, 4516)\n"
     ]
    }
   ],
   "source": [
    "df_combined = df.merge(metadata, left_index=True, right_index=True, how='inner')\n",
    "print(f\"df: {df.shape}\")\n",
    "print(f\"metadata: {metadata.shape}\")\n",
    "print(f\"df_combined: {df_combined.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a2cc3d8d-eeaf-4827-9b14-acb8027ab8dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>taxa_functions</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>10.085904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1447437_20171212_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>59.260325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1128691_20171206_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1128691_20171218_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1128691_20180116_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>14.478968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552445</th>\n",
       "      <td>895293_20180502_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552446</th>\n",
       "      <td>896213_20180427_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552447</th>\n",
       "      <td>913873_20180417_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552448</th>\n",
       "      <td>980574_20180403_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552449</th>\n",
       "      <td>983493_20180123_S</td>\n",
       "      <td>Methylomirabilis</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>552450 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_id                                 taxa_functions  \\\n",
       "0       1068841_20180306_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "1       1447437_20171212_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "2       1128691_20171206_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "3       1128691_20171218_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "4       1128691_20180116_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "...                    ...                                            ...   \n",
       "552445   895293_20180502_S                               Methylomirabilis   \n",
       "552446   896213_20180427_S                               Methylomirabilis   \n",
       "552447   913873_20180417_S                               Methylomirabilis   \n",
       "552448   980574_20180403_S                               Methylomirabilis   \n",
       "552449   983493_20180123_S                               Methylomirabilis   \n",
       "\n",
       "        abundance  \n",
       "0       10.085904  \n",
       "1       59.260325  \n",
       "2        0.000000  \n",
       "3        0.000000  \n",
       "4       14.478968  \n",
       "...           ...  \n",
       "552445   0.000000  \n",
       "552446   0.000000  \n",
       "552447   0.000000  \n",
       "552448   0.000000  \n",
       "552449   0.000000  \n",
       "\n",
       "[552450 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_long = df.reset_index().melt(id_vars='index', var_name='taxa_functions', value_name='abundance')\n",
    "df_long.rename(columns={'index': 'sample_id'}, inplace=True)\n",
    "df_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "37c1b0ef-1843-49be-af92-c96ec69f0699",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>taxa_functions</th>\n",
       "      <th>abundance</th>\n",
       "      <th>minion</th>\n",
       "      <th>MGI</th>\n",
       "      <th>pwCF_ID</th>\n",
       "      <th>Sample date</th>\n",
       "      <th>IP vs OP</th>\n",
       "      <th>Hospital</th>\n",
       "      <th>Room</th>\n",
       "      <th>...</th>\n",
       "      <th>Sum of meds</th>\n",
       "      <th>Sum of antifungals</th>\n",
       "      <th>Sum of steroid + mabs</th>\n",
       "      <th>DNA_extraction_ conc</th>\n",
       "      <th>SAGC ULN</th>\n",
       "      <th>DNA Conc. (ng/ul)</th>\n",
       "      <th>Index I7</th>\n",
       "      <th>Index I5</th>\n",
       "      <th>Mean_Size_BP</th>\n",
       "      <th>Total Clusters Passing Filter (Million)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</td>\n",
       "      <td>10.085904</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2-O-alpha-mannosyl-D-glycerate utilization</td>\n",
       "      <td>2784.895948</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2-aminophenol Metabolism</td>\n",
       "      <td>516.160945</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2-ketoacid oxidoreductases disambiguation</td>\n",
       "      <td>231.778018</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>2-oxoglutarate dehydrogenase</td>\n",
       "      <td>122.573392</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841_20180306_S</td>\n",
       "      <td>1068841</td>\n",
       "      <td>3/06/2018</td>\n",
       "      <td>OP</td>\n",
       "      <td>RAH</td>\n",
       "      <td>Chest Clinic 7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.07</td>\n",
       "      <td>SAGCFN_22_01754</td>\n",
       "      <td>42.8</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>ATAGAGGC</td>\n",
       "      <td>417</td>\n",
       "      <td>2.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 169 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_id                                 taxa_functions  \\\n",
       "0  1068841_20180306_S  2,3-diacetamido-2,3-dideoxy-d-mannuronic acid   \n",
       "1  1068841_20180306_S     2-O-alpha-mannosyl-D-glycerate utilization   \n",
       "2  1068841_20180306_S                       2-aminophenol Metabolism   \n",
       "3  1068841_20180306_S      2-ketoacid oxidoreductases disambiguation   \n",
       "4  1068841_20180306_S                  2-oxoglutarate dehydrogenase    \n",
       "\n",
       "     abundance              minion                 MGI  pwCF_ID Sample date  \\\n",
       "0    10.085904  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "1  2784.895948  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "2   516.160945  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "3   231.778018  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "4   122.573392  1068841_20180306_S  1068841_20180306_S  1068841   3/06/2018   \n",
       "\n",
       "  IP vs OP Hospital            Room  ...  Sum of meds  Sum of antifungals  \\\n",
       "0       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "1       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "2       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "3       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "4       OP      RAH  Chest Clinic 7  ...            0                   0   \n",
       "\n",
       "  Sum of steroid + mabs DNA_extraction_ conc         SAGC ULN  \\\n",
       "0                     0                 1.07  SAGCFN_22_01754   \n",
       "1                     0                 1.07  SAGCFN_22_01754   \n",
       "2                     0                 1.07  SAGCFN_22_01754   \n",
       "3                     0                 1.07  SAGCFN_22_01754   \n",
       "4                     0                 1.07  SAGCFN_22_01754   \n",
       "\n",
       "   DNA Conc. (ng/ul)  Index I7  Index I5  Mean_Size_BP  \\\n",
       "0               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "1               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "2               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "3               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "4               42.8  CGCTCATT  ATAGAGGC           417   \n",
       "\n",
       "   Total Clusters Passing Filter (Million)  \n",
       "0                                      2.8  \n",
       "1                                      2.8  \n",
       "2                                      2.8  \n",
       "3                                      2.8  \n",
       "4                                      2.8  \n",
       "\n",
       "[5 rows x 169 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data =  pd.merge(df_long, metadata, left_on='sample_id', right_on=sequence_type)\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0c47cf3-104c-4eb6-a86d-b3b84dcae618",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns_no_spaces = {}\n",
    "for c in merged_data.columns:\n",
    "    if ' ' in c:\n",
    "        columns_no_spaces[c] = c.replace(' ', '_')\n",
    "        \n",
    "merged_data.rename(columns = columns_no_spaces, inplace=True)\n",
    "\n",
    "\n",
    "df_combined.rename(columns = columns_no_spaces, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f048ac6-d1fe-4053-bd6c-1c8db769b0d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['2,3-diacetamido-2,3-dideoxy-d-mannuronic acid',\n",
       " '2-O-alpha-mannosyl-D-glycerate utilization',\n",
       " '2-aminophenol Metabolism',\n",
       " '2-ketoacid oxidoreductases disambiguation',\n",
       " '2-oxoglutarate dehydrogenase ',\n",
       " '2-phosphoglycolate salvage',\n",
       " '3-amino-5-hydroxybenzoic Acid Synthesis',\n",
       " '4-hydroxybenzoyl-CoA reductase',\n",
       " '5-methylaminomethyl-2-thiouridine',\n",
       " 'A Hypothetical Protein Related to Proline Metabolism',\n",
       " 'A new toxin - antitoxin system',\n",
       " 'ABC transporter YeiABEF',\n",
       " 'ABC transporter YxeMNO',\n",
       " 'ABC transporter of unknown substrate X',\n",
       " 'ABC transporter tungstate (TC 3.A.1.6.2)',\n",
       " 'ABC-type iron transport system',\n",
       " 'AMP to 3-phosphoglycerate',\n",
       " 'ATP-dependent Nuclease',\n",
       " 'ATP-dependent RNA helicases, bacterial',\n",
       " 'AaeAB efflux system for hydroxylated, aromatic carboxylic acids',\n",
       " 'Accessory colonization factor',\n",
       " 'Acetoin, butanediol metabolism',\n",
       " 'Acetolactate synthase subunits',\n",
       " 'Acetophenone carboxylase 1',\n",
       " 'Acetyl-CoA Pathway Wood-Ljungdahl',\n",
       " 'Actinobacterial signal transduction system MtrAB-LpqB',\n",
       " 'Activation of fatty acids as acyl-adenylates cluster subgroup FadD10',\n",
       " 'Acyclic terpene utilization',\n",
       " 'Acyl carrier protein',\n",
       " 'Acyl-CoA thioesterase II',\n",
       " 'Adaptation to d-cysteine',\n",
       " 'Adenosyl nucleosidases',\n",
       " 'Adenylylsulfate reductase',\n",
       " 'Adhesins in Staphylococcus',\n",
       " 'Aerotolerance operon',\n",
       " 'Agrobacterium opine transport',\n",
       " 'Agrocinopine utilization',\n",
       " 'Alginate biosynthesis',\n",
       " 'Alkane biosynthesis in bacteria',\n",
       " 'Alkane synthesis in bacteria 2',\n",
       " 'Alpha-acetolactate operon',\n",
       " 'Amidase clustered with urea and nitrile hydratase functions',\n",
       " 'Amino acid racemase',\n",
       " 'Aminoglycoside modifying enzymes: N-acetyltransferases',\n",
       " 'Aminoglycoside modifying enzymes: O-nucleotidyltransferases',\n",
       " 'Aminoglycoside modifying enzymes: O-phosphotransferases',\n",
       " 'An Arabinose Sensor',\n",
       " 'Anaerobic Oxidative Degradation of L-Ornithine',\n",
       " 'Anaerobic dimethyl sulfoxide reductase (EC 1.8.5.3)',\n",
       " 'Anaerobic glycerol-3-phosphate dehydrogenase subunits',\n",
       " 'Anaerobic module of TCA',\n",
       " 'Anaerobic toluene and ethylbenzene degradation',\n",
       " 'Antibiotic anticapsin - bacilysin',\n",
       " 'Antibiotic targets in DNA processing',\n",
       " 'Antibiotic targets in cell wall biosynthesis',\n",
       " 'Antibiotic targets in metabolic pathways',\n",
       " 'Antibiotic targets in protein synthesis',\n",
       " 'Antibiotic targets in transcription',\n",
       " 'Archaeal membrane bound hydrogenases',\n",
       " 'Arginine biosynthesis',\n",
       " 'Arginine biosynthesis via N-acetyl-L-citrulline',\n",
       " 'Arginine decarboxylase and Agmatinase cluster',\n",
       " 'Arginine succinyltransferase pathway',\n",
       " 'Aromatic amino acid interconversions with aryl acids',\n",
       " 'Arsenic resistance',\n",
       " 'Aspartate to Homoserine module',\n",
       " 'Aspartate to Threonine Module',\n",
       " 'AttEFGH ABC Transport System',\n",
       " 'Autoinducer 2 (AI-2) transport and processing (lsrACDBFGE operon)',\n",
       " 'BOX elements in Streptococci',\n",
       " 'Bacillibactin Siderophore',\n",
       " 'Bacillithiol synthesis',\n",
       " 'Bacillus cannibalism system',\n",
       " 'Bacitracin resistance',\n",
       " 'BactLuciferaseComplex',\n",
       " 'Bacterial Sphingolipids',\n",
       " 'Bacterial cell division related cluster 2',\n",
       " 'Bacterial checkpoint-control-related cluster',\n",
       " 'Bacterial pathways for dimethylsulfoniopropionate and acrylate catabolism',\n",
       " 'Bacteriocin resistance related cluster',\n",
       " 'Bacteriocins in Lactobacilli',\n",
       " 'Bacteroides capsular polysaccharide transcription antitermination proteins',\n",
       " 'BarA-UvrY(SirA) two-component regulatory system',\n",
       " 'Barnase-barstar complex',\n",
       " 'Benzoyl-CoA pathway of anaerobic aromatic metabolism',\n",
       " 'Beta-lactamases Ambler class A',\n",
       " 'Beta-lactamases Ambler class B',\n",
       " 'Beta-lactamases Ambler class C',\n",
       " 'Beta-lactamases Ambler class D',\n",
       " 'Betaine lipids in bactera',\n",
       " 'Bicyclomycin resistance cluster',\n",
       " 'Bile hydrolysis',\n",
       " 'Biofilm Adhesin Biosynthesis',\n",
       " 'Biofilm formation in Staphylococcus',\n",
       " 'Biogenesis of c-type cytochromes',\n",
       " 'Biogenesis of cbb3-type cytochrome c oxidases',\n",
       " 'Biogenesis of cytochrome c oxidases',\n",
       " \"Biosynthesis of 4,4'-diapolycopene dialdehyde\",\n",
       " 'Biosynthesis of Arabinogalactan in Mycobacteria',\n",
       " 'Biosynthesis of thiamin antivitamin Bacimethrin',\n",
       " 'Biotin biosynthesis',\n",
       " 'Biotin synthesis & utilization',\n",
       " 'Biotin synthesis cluster',\n",
       " 'Box C-D RNA-guided RNA methyltransferase',\n",
       " 'Branched-Chain Amino Acid Biosynthesis',\n",
       " 'Branched-chain amino acids and alpha-keto acids utilization as energy sources',\n",
       " 'BtuS-PduT gene cluster for metalloporphyrine salvage',\n",
       " 'Butyrate kinase pathway',\n",
       " 'CMP-legionaminic acid synthesis',\n",
       " 'CMP-pseudaminic acid synthesis',\n",
       " 'Cadmium resistance',\n",
       " 'Calvin-Benson cycle',\n",
       " 'Carbon monoxide induced hydrogenase',\n",
       " 'Carbon storage regulator',\n",
       " 'Carbonic anhydrase',\n",
       " 'Carboxysome',\n",
       " 'Cardiolipin biosynthesis',\n",
       " 'Carnitine metabolism and transport',\n",
       " 'Carnitine operon',\n",
       " 'Carotenogenesis related cluster',\n",
       " 'Cell division cluster',\n",
       " 'Cell division initiation related cluster',\n",
       " 'Cell division related cluster',\n",
       " 'Cell division related cluster including coaD',\n",
       " 'Cell envelope-associated LytR-CpsA-Psr transcriptional attenuators',\n",
       " 'Cell invasion-related virulence operon MCE',\n",
       " 'Cell wall-associated cluster in Mycobacterium',\n",
       " 'Chaperones GroEL GroES and Thermosome',\n",
       " 'Chemotaxis in Escherichia coli and Bacillus subtilis',\n",
       " 'Chitobiose',\n",
       " 'Chloramphenicol resistance',\n",
       " 'Chlorophyll Biosynthesis in plants and prokaryotes',\n",
       " 'Chlorophyll Degradation',\n",
       " 'Chlorophyll a synthase and phytol kinase cluster',\n",
       " 'Chlorosome',\n",
       " 'Cholera toxin',\n",
       " 'Cholesterol catabolism',\n",
       " 'Choline uptake and conversion to betaine clusters',\n",
       " 'Chorismate Synthesis',\n",
       " 'Citrate lyase',\n",
       " 'Citrate uptake cluster',\n",
       " 'Cluster containing Glutathione synthetase',\n",
       " 'Cluster related to 2-thiouridine modification of tRNAs',\n",
       " 'Cluster related to PAS - para-aminosalicylic acid resistance',\n",
       " 'CoA disulfide thiol-disulfide redox system',\n",
       " 'Coat proteins CotJABC',\n",
       " 'Cob(I)alamin adenosyltransferase',\n",
       " 'Cobalamin synthesis',\n",
       " 'Cobalamin synthesis from Cob(II)yrinate a,c diamide',\n",
       " 'Cobalamin synthesis, Precorin-2 to Cob(II)yrinate a,c diamide, aerobic',\n",
       " 'Cobalamin synthesis, Precorin-2 to Cob(II)yrinate a,c diamide, anaerobic',\n",
       " 'Coenzyme A',\n",
       " 'Coenzyme A -- gjo',\n",
       " 'Coenzyme A Biosynthesis cluster',\n",
       " 'Coenzyme B synthesis',\n",
       " 'Coenzyme F420 synthesis',\n",
       " 'Coenzyme F420-H2 dehydrogenase (methanophenazine)',\n",
       " 'Coenzyme PQQ synthesis',\n",
       " 'Colanic acid synthesis',\n",
       " 'Cold shock proteins of CSP family',\n",
       " 'Colicin V and Bacteriocin Production Cluster',\n",
       " 'Conjugative transfer',\n",
       " 'Conjugative transposon, Bacteroidales',\n",
       " 'Control of cell elongation - division cycle in Bacilli',\n",
       " 'Copper Transport System',\n",
       " 'Copper homeostasis: copper tolerance',\n",
       " 'Copper transport and blue copper proteins',\n",
       " 'Copper uptake system CopCD',\n",
       " 'CtsR and MscAB regulation of protein degradation',\n",
       " 'Cyanate hydrolysis',\n",
       " 'Cyanobacterial bypass in the TCA',\n",
       " 'Cyanophycin Metabolism',\n",
       " 'Cyclic Di-GMP Turnover Proteins',\n",
       " 'Cysteine dioxygenase containing cluster 1',\n",
       " 'Cysteine synthesis',\n",
       " 'CytR regulation',\n",
       " 'Cytochrome b6-f complex in plants (plastidial) and cyanobacteria',\n",
       " 'Cytochrome bo ubiquinol oxidase ',\n",
       " 'Cytochrome c oxidase EC 1.9.3.1',\n",
       " 'Cytochrome d ubiquinol oxidase operon',\n",
       " 'Cytolysin and Lipase operon in Vibrio',\n",
       " 'D-3-phosphoglycerate to phosphoserine module',\n",
       " 'D-Galacturonate and D-Glucuronate Utilization',\n",
       " 'D-alanylation of teichoic acid',\n",
       " 'D-allose utilzation',\n",
       " 'D-altritol utilization',\n",
       " 'D-arabinitol utilization',\n",
       " 'D-arabinose utilization',\n",
       " 'D-galacturonate utilization, oxidative pathway',\n",
       " 'D-gluconate and ketogluconates metabolism',\n",
       " 'D-glucosaminate utilization',\n",
       " 'D-threonate, L-threonate and D-erythronate utilization',\n",
       " 'D-tyrosyl-tRNA(Tyr) deacylase',\n",
       " 'DAP (1,3-diaminopropane) production',\n",
       " 'DNA Helicase of Unknown Function',\n",
       " 'DNA Repair Base Excision',\n",
       " 'DNA internalization-related cluster',\n",
       " 'DNA phosphorothioation-dependent restriction system DptFGH',\n",
       " 'DNA processing cluster',\n",
       " 'DNA recombination, archaeal',\n",
       " 'DNA repair system including RecA, MutS and a hypothetical protein',\n",
       " 'DNA repair, UvrABC system',\n",
       " 'DNA repair, bacterial',\n",
       " 'DNA repair, bacterial DinG and relatives',\n",
       " 'DNA repair, bacterial MutHLS system',\n",
       " 'DNA repair, bacterial RecBCD pathway',\n",
       " 'DNA repair, bacterial RecFOR pathway',\n",
       " 'DNA repair, bacterial SbcCD exonuclease',\n",
       " 'DNA repair, bacterial UmuCD system',\n",
       " 'DNA repair, bacterial UvrD and related helicases',\n",
       " 'DNA repair, bacterial photolyase',\n",
       " 'DNA sulfur modification cluster',\n",
       " 'DNA topoisomerases, Type II, ATP-dependent',\n",
       " 'Damaged nicotinamide nucleotide NAD(P)HX repair',\n",
       " 'De Novo Pyrimidine Synthesis',\n",
       " 'DeNovo Purine Biosynthesis',\n",
       " 'Degradation of branched-chain amino acids and alpha-keto acids [Leu, Val]',\n",
       " 'Dehydrogenase complexes',\n",
       " 'Delta(6)-desaturase containing cluster',\n",
       " 'Denitrifying reductase gene clusters',\n",
       " 'Desulfovibrio high-molecular-weight cytochrome cluster',\n",
       " 'DevBCA efflux pump transporter (of ABC type) cyanobacterial',\n",
       " 'Diaminopimelate Synthesis ',\n",
       " 'Dihydropyrimidine dehydrogenase',\n",
       " 'Dihydroxyacetone kinases',\n",
       " 'Dioxygenolytic carbon monoxide-forming enzymes',\n",
       " 'Diphthamide biosynthesis',\n",
       " 'Diphtheria toxin',\n",
       " 'Dipicolinate Synthesis',\n",
       " 'Dissimilatory nitrite reductase',\n",
       " 'Dpp dipeptide ABC transport system',\n",
       " 'ESAT-6 protein secretion system in Mycobacteria (locus ESX-5)',\n",
       " 'ESAT-6 proteins secretion system in Firmicutes and Actinobacteria',\n",
       " 'ESAT-6 proteins secretion system in Firmicutes extended',\n",
       " 'ESAT-6 proteins secretion system in Mycobacteria (locus ESX-1)',\n",
       " 'ESAT-6 proteins secretion system in Mycobacteria (locus ESX-3)',\n",
       " 'ESAT-6 proteins secretion system in Mycobacteria (locus ESX-4)',\n",
       " 'EcsAB transporter affecting expression and secretion of secretory preproteins',\n",
       " 'Ectoine synthesis',\n",
       " 'Ectoine, hydroxyectoine uptake and catabolism',\n",
       " 'Efflux ABC transporter MdlAB',\n",
       " 'Efflux ABC transporter YadGH',\n",
       " 'Efflux ABC transporters BmrCD and LmrCD involved in multidrug resistance',\n",
       " 'Electron-bifurcating caffeyl-CoA reductase-Etf complex',\n",
       " 'Encapsulation of DyP-type peroxidase or ferritin-like protein oligomers',\n",
       " 'Energy conserving hydrogenase b, Methanococcales-Methanobacteriales-Methanopyrales',\n",
       " 'Energy-conserving hydrogenase (ferredoxin)',\n",
       " 'Enoyl-[ACP] reductases disambiguation',\n",
       " 'Entericidin',\n",
       " 'Enterobacterial common antigen (LPS O-antigen)',\n",
       " 'Entner-Doudoroff Pathway',\n",
       " 'EnvZ and OmpR regulon',\n",
       " 'Ergothioneine biosynthesis',\n",
       " 'Ethanolamine utilization',\n",
       " 'Ethylmalonyl-CoA pathway of C2 assimilation',\n",
       " 'Exosome',\n",
       " 'Exosporium',\n",
       " 'Extracellular matrix proteins (PEL) involved in glucose-rich biofilm formation in Pseudomonas',\n",
       " 'Extracellular matrix proteins (PSL) involved in mannose-rich biofilm formation in Pseudomonas',\n",
       " 'F0F1-type ATP synthase',\n",
       " 'Fatty Acid Biosynthesis FASII',\n",
       " 'Fatty Acid Biosynthesis cluster',\n",
       " 'Fatty acid catabolic operon fadN-fadA-fadE (yusJKL)',\n",
       " 'Fatty acid degradation',\n",
       " 'Fatty acid synthesis',\n",
       " 'Fe(III) respiration - Shewanella type',\n",
       " 'Fe-S cluster assembly',\n",
       " 'Fermentations: Lactate',\n",
       " 'Fermentations: Mixed acid',\n",
       " 'Ferrous iron transporter EfeUOB, low-pH-induced',\n",
       " 'Fet iron export system',\n",
       " 'Flagellum',\n",
       " 'Folate Biosynthesis',\n",
       " 'Folate biosynthesis cluster',\n",
       " 'Forespore to mother cell channel',\n",
       " 'Formaldehyde assimilation: Ribulose monophosphate pathway',\n",
       " 'Formate dehydrogenase',\n",
       " 'Fosfomycin resistance',\n",
       " 'Fructoselysine and glucoselysine',\n",
       " 'FtsEX cell-division-associated signaling system',\n",
       " 'Fusaric acid resistance cluster',\n",
       " 'Fusidic acid resistance',\n",
       " 'G-protein-coupled receptor (GPCR) system in Actinobacteria',\n",
       " 'GDP-fucose to mannose-1P module',\n",
       " 'GMP synthase',\n",
       " 'Galactosamine Substituent of Arabinogalactan in Mycobacteria',\n",
       " 'Galactose utilization',\n",
       " 'Gallic acid utilization',\n",
       " 'Gene Transfer Agent',\n",
       " 'General Secretion Pathway',\n",
       " 'Gentisate pathway',\n",
       " 'GlrKR two-component regulatory system',\n",
       " 'Glutaconate CoA-transferase or 3-oxoadipate CoA-transferase Subunits',\n",
       " 'Glutamate dehydrogenases',\n",
       " 'Glutamate fermentation',\n",
       " 'Glutamine synthetases',\n",
       " 'Glutathione analogs: mycothiol',\n",
       " 'Glutathione-dependent pathway of formaldehyde detoxification',\n",
       " 'Glutathione: Biosynthesis and gamma-glutamyl cycle',\n",
       " 'Glutathione: Non-redox reactions',\n",
       " 'Glutathione: Redox cycle',\n",
       " 'Glutathionylspermidine and Trypanothione',\n",
       " 'Glycerol ABC transporter',\n",
       " 'Glycerol fermentation to 1,3-propanediol',\n",
       " 'Glycerol utilization cluster',\n",
       " 'Glycerolipid and Glycerophospholipid Metabolism in Bacteria',\n",
       " 'Glycine betaine synthesis from choline',\n",
       " 'Glycine cleavage system',\n",
       " 'Glycine reductase, sarcosine reductase and betaine reductase',\n",
       " 'Glycogen metabolism',\n",
       " 'Glycolate, glyoxylate interconversions',\n",
       " 'Glycolysis and Gluconeogenesis',\n",
       " 'Glycolysis and Gluconeogenesis, including Archaeal enzymes',\n",
       " 'Glycyl-tRNA synthetase',\n",
       " 'Glyoxylate bypass',\n",
       " 'Gram-negative cluster',\n",
       " 'HMG CoA Synthesis',\n",
       " 'HPr catabolite repression system',\n",
       " 'HU-alpha and dsDNA mimic protein interaction',\n",
       " 'Head-to-head olefinic hydrocarbon biosynthesis',\n",
       " 'Heat shock dnaK gene cluster extended',\n",
       " 'Heme Biosynthesis: protoporphyrin-, coproporphyrin- and siroheme-dependent pathways',\n",
       " 'Heme O and Heme A biosynthesis',\n",
       " 'Heme O and Heme A biosynthesis (with selected terminal oxidases)',\n",
       " 'Heme and heme d1 biosynthesis from siroheme',\n",
       " 'Heme, hemin uptake and utilization systems in GramPositives',\n",
       " 'Heterocyst formation in cyanobacteria',\n",
       " 'Hexose phosphate transport system',\n",
       " 'Hfl operon',\n",
       " 'High affinity phosphate transporter and control of PHO regulon',\n",
       " 'HipAB system implicated in growth arrest, persistence and drug tolerance',\n",
       " 'Histidine Biosynthesis',\n",
       " 'Histidine Biosynthesis -- gjo',\n",
       " 'Histidine Degradation',\n",
       " 'Homogentisate pathway',\n",
       " 'Homoprotocatechuate degradative cluster',\n",
       " 'Hopanoid biosynthesis',\n",
       " 'Hyaluronate utilization',\n",
       " 'Hyaluronic Acid-containing Cell Walls',\n",
       " 'Hydrogen-sensing hydrogenase system',\n",
       " 'Hydrolysis of sphingomyelin',\n",
       " 'Hydroxy-fatty acid production as stress response',\n",
       " 'Hydroxyaromatic decarboxylase family',\n",
       " 'Hyperosmotic potassium uptake',\n",
       " 'IMP and Xanthine conversion to XMP and GMP module',\n",
       " 'IbrA and IbrB: co-activators of prophage gene expression',\n",
       " 'Indole-pyruvate oxidoreductase complex',\n",
       " 'Indoleacetamide metabolism',\n",
       " 'Inner membrane proteins of MarC family, not involved in antibiotic resistance',\n",
       " 'Inositol catabolism',\n",
       " 'Inositol-containing phospholipids',\n",
       " 'Integrons',\n",
       " 'Iron Scavenging cluster in Thermus',\n",
       " 'Iron acquisition in Streptococcus',\n",
       " 'Iron siderophore sensor & receptor system',\n",
       " 'Iron transport system including ABC transporter',\n",
       " 'Iron(III) dicitrate transport system Fec',\n",
       " 'Isoniazid inductible proteins cluster',\n",
       " 'Isoprenoid Biosynthesis: Interconversions',\n",
       " 'KDO2-Lipid A biosynthesis',\n",
       " 'L-2-amino-4-methoxy-trans-3-butenoic acid synthesis',\n",
       " 'L-2-amino-adipate to lysine module',\n",
       " 'L-Amino acid oxidases: lysine-epsilon oxidase LodA',\n",
       " 'L-ascorbate utilization (and related gene clusters)',\n",
       " 'L-fucose dissimilation cluster',\n",
       " 'L-fucose utilization',\n",
       " 'L-galactonate utilization',\n",
       " 'L-rhamnonate utilization cluster',\n",
       " 'L-serine dehydratase subunits',\n",
       " 'Lactose utilization',\n",
       " 'Lambdoid phage',\n",
       " 'Lasso peptide maturation cluster',\n",
       " 'Late competence',\n",
       " 'Lipid orphans',\n",
       " 'Lipid-linked oligosaccharide synthesis related cluster',\n",
       " 'Lipoic acid metabolism',\n",
       " 'Lipopolysaccharide in K12',\n",
       " 'Lipopolysaccharide transporter LptABC',\n",
       " 'Lipopolysaccharide-related cluster in Alphaproteobacteria',\n",
       " 'Lipoprotein transporter to outer membrane Lol',\n",
       " 'Lipoylated proteins',\n",
       " 'Lipoylation-related cluster',\n",
       " 'Listeria Pathogenicity Island LIPI-1 extended',\n",
       " 'Listeria surface proteins: Internalin-like proteins',\n",
       " 'Lpt lipopolysaccharide transport system',\n",
       " 'Lysine Biosynthesis via alpha-aminoadipate AAA',\n",
       " 'Lysine DAP biosynthetic pathway',\n",
       " 'Lysine fermentation',\n",
       " 'Lysine fermentation to crotonoyl-CoA',\n",
       " 'Lysine leader peptide',\n",
       " 'Macrolides, lincosamides, streptogramins, ketolides, oxazolidinones (MLSKO) resistance: efflux systems',\n",
       " 'Macrolides, lincosamides, streptogramins, ketolides, oxazolidinones (MLSKO) resistance: enzymatic degradation',\n",
       " 'Macrolides, lincosamides, streptogramins, ketolides, oxazolidinones (MLSKO) resistance: rRNA methylases',\n",
       " 'Macrolides, lincosamides, streptogramins, ketolides, oxazolidinones (MLSKO) resistance: ribosomal protection',\n",
       " 'Macromolecular synthesis operon',\n",
       " 'Magnesium transport',\n",
       " 'Maintenance of outer membrane lipid asymmetry',\n",
       " 'Malonate decarboxylase',\n",
       " 'Mannosylglucosylglycerate synthesis',\n",
       " 'MazEF toxin-antitoxing (programmed cell death) system',\n",
       " 'Mebrane bound hydrogenases',\n",
       " 'Mechanisms of PAS resistance',\n",
       " 'Membrane bound hydrogenase Mbh',\n",
       " 'Membrane bound hydrogenase Mbh -- gjo',\n",
       " 'Menaquinone biosynthesis from chorismate via 1,4-dihydroxy-2-naphthoate',\n",
       " 'Menaquinone biosynthesis from chorismate via 1,4-dihydroxy-6-naphthoate',\n",
       " 'Metabolite repair',\n",
       " 'Metal chelatases',\n",
       " 'Methanofuran biosynthesis',\n",
       " 'Methanogenesis',\n",
       " 'Methanogenesis from methylated compounds',\n",
       " 'Methionine Salvage',\n",
       " 'Methylenetetrahydrofolate reductase',\n",
       " 'Methylglyoxal Metabolism',\n",
       " 'Methylglyoxal resistance in Bacillus subtilis',\n",
       " 'Methylhydantoinase',\n",
       " 'Methylthiotransferases',\n",
       " 'Mevalonate metabolic pathway',\n",
       " 'MexXY System of Pseudomonas aeruginosa',\n",
       " 'Modification of eukaryotic initiation factor 5A',\n",
       " 'Molybdenum cofactor biosynthesis',\n",
       " 'Molybdopterin cytosine dinucleotide',\n",
       " 'Mu-like phage',\n",
       " 'Muconate lactonizing enzyme family',\n",
       " 'MukBEF Chromosome Condensation',\n",
       " 'Multi-subunit cation antiporter',\n",
       " 'Multidrug Resistance, Tripartite Systems Found in Gram Negative Bacteria TEMP',\n",
       " 'Multidrug resistance RND efflux system MexEFOprN',\n",
       " 'Multiple Antibiotic Resistance MAR locus',\n",
       " 'Mupirocin resistance',\n",
       " 'Murein hydrolase regulation and cell death',\n",
       " 'Mycobacterial FadD proteins (fatty acid CoA- and AMP- ligases)',\n",
       " 'Mycobacterial FadE proteins Acyl-CoA dehydrogenase',\n",
       " 'Mycobacterial HadABC proteins ((3R)-hydroxyacyl-ACP dehydrates of FASII)',\n",
       " 'Mycobacterial cell wall virulence lipid phthiocerol dimycocerosate (PDIM)',\n",
       " 'Mycobacterial gene cluster associated with resistance against FAS-II antibiotics',\n",
       " 'Mycobacterial heme acquisition system',\n",
       " 'Mycobacterial lipid and multidrug efflux system proteins MmpL, MmpS',\n",
       " 'Mycobacterial lipoproteins (LPR group)',\n",
       " 'Mycobacterial multidrug efflux systems implicated in AMR ',\n",
       " 'Mycobacterial signal transduction system MprAB required for persistent infections',\n",
       " 'Mycofactocin system cluster',\n",
       " 'Mycolic acid synthesis 2',\n",
       " 'Mycosporine synthesis cluster',\n",
       " \"N,N'-diacetylbacillosamine\",\n",
       " 'N-acetylneuraminate utilization',\n",
       " 'NAD and NADP cofactor biosynthesis bacterial',\n",
       " 'NAD and NADP cofactor biosynthesis global',\n",
       " 'NAD-reducing [NiFe] hydrogenase HoxFUYH(E)',\n",
       " 'NADH ubiquinone oxidoreductase',\n",
       " 'NADH ubiquinone oxidoreductase vs. multi-subunit cation antiporter',\n",
       " 'NADP-reducing [FeFe]-hydrogenase',\n",
       " 'Na(+) H(+) antiporter',\n",
       " 'Na-driven 2-hydroxyglutarate pathway',\n",
       " 'Na-translocating NADH-quinone oxidoreductase',\n",
       " 'NhaA, NhaD and Sodium-dependent phosphate transporters',\n",
       " 'NiFe hydrogenase maturation',\n",
       " 'Nickel-requiring',\n",
       " 'Nicotinate catabolism, anaerobic',\n",
       " 'Nicotinic acid utilization',\n",
       " 'Nitrilase',\n",
       " 'Nitrogen fixation',\n",
       " 'Nonmevalonate Branch of Isoprenoid Biosynthesis',\n",
       " 'Nucleoside ABC transporter cluster',\n",
       " 'Nucleoside triphosphate pyrophosphohydrolase MazG',\n",
       " 'Nucleoside uptake and degradation cluster',\n",
       " 'Nudix proteins (nucleoside triphosphate hydrolases)',\n",
       " 'O-methyl phosphoramidate modification in capsular polysaccharide',\n",
       " 'Oligopeptide degradation cluster',\n",
       " 'One-carbon metabolism by tetrahydropterines',\n",
       " 'Osmoregulation',\n",
       " 'Osmotic stress cluster',\n",
       " 'Outer membrane',\n",
       " 'Outer membrane porins in Mycobacteria',\n",
       " 'Outer membrane porins in Pseudomonas and Acinetobacter',\n",
       " 'Outer membrane proteases (Omptins)',\n",
       " 'Oxaloacetate decarboxylase Na-pump',\n",
       " 'Oxygen-sensing diguanylate cyclase and phosphodiesterase couple for c-di-GMP control',\n",
       " 'Oxygen-sensing two-component regulatory system NreABC',\n",
       " 'P2-like phage',\n",
       " 'PE multigene family in Mycobacterium tuberculosis complex (MTBC)',\n",
       " 'PEB1, the major cell-binding factor of Campylobacter jejuni',\n",
       " 'PPE multigene family in Mycobacterium tuberculosis complex (MTBC)',\n",
       " 'Particulate methane monooxygenase (pMMO)',\n",
       " 'Pentalenolactone Biosynthesis and Resistance',\n",
       " 'Pentose phosphate pathway',\n",
       " 'Peptidase clustering with DAP',\n",
       " 'Peptide ABC transport system Sap',\n",
       " 'Peptide methionine sulfoxide reductase',\n",
       " 'Peptidyl-prolyl cis-trans isomerase',\n",
       " 'Periplasmic disulfide interchange',\n",
       " 'Periplasmic nitrate reductase EC 1.7.99.4',\n",
       " 'Phage shock protein operon',\n",
       " 'Phd-Doc, YdcE-YdcD toxin-antitoxin (programmed cell death) systems',\n",
       " 'Phenazine biosynthesis',\n",
       " 'Phenol hydroxylase',\n",
       " 'Phenylacetyl-CoA catabolic pathway (core)',\n",
       " 'Phenylalanine and Tyrosine synthesis 1',\n",
       " 'Phosphatidylinositol mannosides biosynthesis related cluster',\n",
       " 'Phosphoenolpyruvate phosphomutase',\n",
       " 'Phospholipid and Fatty acid biosynthesis related cluster',\n",
       " 'Phosphonate (phosphite) dehydrogenase',\n",
       " 'Phosphonate metabolism',\n",
       " 'Phosphonoalanine utilization',\n",
       " 'Photorespiration (oxidative C2 cycle)',\n",
       " 'Photosystem I',\n",
       " 'Photosystem II',\n",
       " 'Phr peptides - Rap phosphatases signaling',\n",
       " 'Plasmid replication',\n",
       " 'Poly-gamma-glutamate biosynthesis',\n",
       " 'Polyadenylation, bacterial',\n",
       " 'Polyamine Metabolism',\n",
       " 'Polymyxin resistance',\n",
       " 'Polymyxin resistance, lipid A modifications with 4-amino-4-deoxy-L-arabinose',\n",
       " 'Polymyxin resistance, lipid A modifications with phosphoethanolamine',\n",
       " 'Polysulfide reductase EC 1.12.98.4 and Thiosulfate reductase',\n",
       " 'Polyunsaturated Fatty Acids synthesis',\n",
       " 'Pore-forming cytolytic toxins',\n",
       " 'Possible DNA repair related cluster',\n",
       " 'Possible RNA modification and stress response cluster',\n",
       " 'Possible heterocyst differentiation related cluster',\n",
       " 'Possible new toxin-antitoxin system including DivIC',\n",
       " 'Possible stress related actinobacterial cluster',\n",
       " 'Post-translational modification of proteins',\n",
       " 'Potassium-transporting ATPase',\n",
       " 'Potential DNA repair cluster',\n",
       " 'Predicted secretion system W clustering with cell division proteins',\n",
       " 'Predicted secretion system X',\n",
       " 'Prenylated indole derivatives in actinomycetes',\n",
       " 'Programmed cell death toxin-antitoxin PezAT',\n",
       " 'Programmed frameshift',\n",
       " 'Proline Synthesis',\n",
       " 'Proline biosynthesis (for review)',\n",
       " 'Proline, 4-hydroxyproline uptake and utilization',\n",
       " 'Propanediol utilization',\n",
       " 'Proteasome bacterial',\n",
       " 'Protection from Reactive Oxygen Species',\n",
       " 'Protein chaperones',\n",
       " 'Protein degradation',\n",
       " \"Protein-O-mannosyltransferase and 16S rRNA (cytidine(1402)-2'-O)-methyltransferase cluster\",\n",
       " 'Proteorhodopsin',\n",
       " 'Pterin synthesis related cluster',\n",
       " 'Purine catabolism in Bacillus subtilis',\n",
       " 'Purine nucleotide synthesis regulator',\n",
       " 'Putative TldE-TldD proteolytic complex',\n",
       " 'Putative oxidase COG2907',\n",
       " 'Putrescine utilization',\n",
       " 'Pyridoxin (Vitamin B6) Biosynthesis',\n",
       " 'Pyridoxin(Vitamin B6) Degradation Pathway',\n",
       " 'Pyrroloquinoline Quinone biosynthesis',\n",
       " 'Pyruvate Alanine Serine Interconversions',\n",
       " 'Pyruvate formate-lyase cluster',\n",
       " 'Pyruvate formate-lyase cluster with possible role in choline utilization',\n",
       " 'Pyruvate metabolism I: anaplerotic reactions, PEP',\n",
       " 'Pyruvate metabolism II: acetyl-CoA, acetogenesis from pyruvate',\n",
       " 'Pyruvate:ferredoxin oxidoreductase',\n",
       " 'Queuosine and archaeosine',\n",
       " 'Quinolone (Fluoroquinolone) resistance via Pentapeptide repeat proteins',\n",
       " 'Quinone oxidoreductase family',\n",
       " 'Quorum sensing in Yersinia',\n",
       " 'Quorum sensing regulation in Pseudomonas',\n",
       " 'Quorum-sensing related cluster',\n",
       " 'RNA polymerase, archaeal',\n",
       " 'RNA polymerase, archaeal initiation factors',\n",
       " 'RNA polymerase, bacterial',\n",
       " 'RNA processing and degradation, bacterial',\n",
       " 'RNA pseudouridine synthases',\n",
       " 'RSF1010 like IncQ broad-host-range plasmids',\n",
       " 'Rcn nickel and cobalt homeostasis system',\n",
       " 'Rcs two-component regulator of capsule synthesis',\n",
       " 'RecA and RecX',\n",
       " 'Repair of Iron Centers',\n",
       " 'Replication-associated recombination protein RarA',\n",
       " 'Resistance to Daptomycin',\n",
       " 'Resistance to Ethambutol',\n",
       " 'Resistance to Ethionamide and Isoniazid',\n",
       " 'Resistance to Triclosan',\n",
       " 'Resistance to Vancomycin and Teicoplanin',\n",
       " 'Resistance to chromium compounds',\n",
       " 'Resistance to the fluoroquinolones norfloxacin and ciprofloxacin',\n",
       " 'Respiratory Complex I',\n",
       " 'Respiratory Systems - Shewanella (updated)',\n",
       " 'Respiratory complex I in plants (plastidial) and cyanobacteria',\n",
       " 'Rhamnolipids in Pseudomonas',\n",
       " 'Ribitol utilization',\n",
       " 'Riboflavin, FMN and FAD biosynthesis in Archaea',\n",
       " 'Riboflavin, FMN and FAD metabolism with fusion events',\n",
       " 'Ribonuclease H',\n",
       " 'Ribonuclease J family',\n",
       " 'Ribonucleotide reduction',\n",
       " 'Ribosomal hibernation related cluster',\n",
       " 'Ribosomal protein S12p Asp methylthiotransferase',\n",
       " 'Ribosomal protein S5p acylation',\n",
       " 'Ribosomal proteins, single-copy',\n",
       " 'Ribosome LSU, bacterial',\n",
       " 'Ribosome LSU, eukaryotic and archaeal',\n",
       " 'Ribosome SSU, eukaryotic and archaeal',\n",
       " 'Ribosome activity modulation',\n",
       " 'Ribosome biogenesis, archaeal',\n",
       " 'RuBisCO operon',\n",
       " 'Rut pathway of pyrimidine degradation',\n",
       " 'RuvABC plus a hypothetical',\n",
       " 'S-Adenosyl-L-homocysteine recycling',\n",
       " 'S-methylmethionine',\n",
       " 'Salmochelin-mediated Iron Acquisition',\n",
       " 'Sarcosine oxidases, monomeric and heterotetrameric',\n",
       " 'SecY2-SecA2 Specialized Transport System',\n",
       " 'Selenocysteine metabolism',\n",
       " 'Selenoprotein O',\n",
       " 'Sex pheromones in Enterococcus faecalis and other Firmicutes',\n",
       " 'Shikimate kinase SK3 cluster',\n",
       " 'Siderophore Achromobactin',\n",
       " 'Siderophore Aerobactin',\n",
       " 'Siderophore Anthrachelin',\n",
       " 'Siderophore Enterobactin',\n",
       " 'Siderophore Mycobactin',\n",
       " 'Siderophore Pyoverdine',\n",
       " 'Siderophore Staphylobactin',\n",
       " 'Siderophore Yersiniabactin Biosynthesis',\n",
       " 'Siderophore assembly kit',\n",
       " 'Siderophore pyochelin',\n",
       " 'Sigma54-dependent transcription related gene cluster',\n",
       " 'Sigma54-dependent transcription related gene cluster 2',\n",
       " 'Signal peptidase',\n",
       " 'Small efflux transporters of SMR family',\n",
       " 'Sodium-transporting ATPase',\n",
       " 'Sox sulfur oxidation genes',\n",
       " 'Sox thiosulfate oxidation system',\n",
       " 'SpoVS protein family',\n",
       " 'Spore germinant receptors',\n",
       " 'Spore germination',\n",
       " 'Sporulation Cluster',\n",
       " 'Sporulation Cluster III A',\n",
       " 'Sporulation gene orphans',\n",
       " 'Sporulation proteins SigEG cluster',\n",
       " 'Sporulation proteins SpoIIIAA-SpoIIIAH',\n",
       " 'Sporulation proteins SpoVA cluster',\n",
       " 'Sporulation-associated proteins with broader functions',\n",
       " 'Streptococcal Hyaluronic Acid Capsule',\n",
       " 'Streptolysin S',\n",
       " 'Stress proteins YciF, YciE',\n",
       " 'Stress response and cell wall lysis cluster',\n",
       " 'Stringent Response, (p)ppGpp metabolism',\n",
       " 'Succinate dehydrogenase and Fumarate reductase cpmlexes',\n",
       " 'Sucrose to levan conversions',\n",
       " 'Sugar-phosphate stress regulation',\n",
       " 'Sulfate assimilation related cluster',\n",
       " 'Sulfate reduction-associated complexes',\n",
       " 'Sulfhydrogenase',\n",
       " 'Sulfoquinovose biosynthesis cluster',\n",
       " 'Sulfur metabolism related cluster',\n",
       " 'Sulfur transfer pathway CsdAEL',\n",
       " 'Sulfur transport related cluster',\n",
       " 'Sulfurtransferase and Ssec cluster',\n",
       " 'Sulphoglycolysis',\n",
       " 'TAM transport system',\n",
       " 'TCA Cycle',\n",
       " 'Tagatose utilization',\n",
       " 'Taurine and alkanesulfonate utilization as sulfur sources',\n",
       " 'Teichuronic acid biosynthesis',\n",
       " 'Teicoplanin resistance in Staphylococci',\n",
       " 'Tetracycline resistance, all mechanisms',\n",
       " 'Tetracycline resistance, ribosomal protection type',\n",
       " 'Tetrathionate respiration',\n",
       " 'The Chv two-component regulatory system',\n",
       " 'Thiamin, hydroxymethylpyrimidine selected transporters',\n",
       " 'Thiamin, thiazole, hydroxymethylpyrimidine salvage and uptake',\n",
       " 'Thiazole.oxazole-modified microcins',\n",
       " 'Thiocyanate hydrolase',\n",
       " 'Threonine anaerobic catabolism gene cluster',\n",
       " 'Threonine synthase cluster',\n",
       " 'Threonylcarbamoyladenosine',\n",
       " 'Tocopherols and tocotrienols (vitamin E) in plants and cyanobacteria',\n",
       " 'Tol-Pal Cell Envelope Complex',\n",
       " 'Tolerance to colicin E2 and CreBC signal transduction system',\n",
       " 'Toluene 4-monooxygenase (T4MO)',\n",
       " 'TonB-ExbBD energy transducing system',\n",
       " 'Toxin-Antitoxin MT1',\n",
       " 'Toxin-Antitoxin system HigAB',\n",
       " 'Toxin-antitoxin replicon stabilization systems',\n",
       " 'Toxins',\n",
       " 'Trans-envelope signaling system VreARI in Pseudomonas',\n",
       " 'Trans-translation by stalled ribosomes',\n",
       " 'Transcription factors, cyanobacterial RpoD-like sigma factors',\n",
       " 'Transcriptional regulator AfsR',\n",
       " 'Translation elongation factor G family',\n",
       " 'Translation elongation factors, bacterial',\n",
       " 'Translation elongation factors, eukaryotic and archaeal',\n",
       " 'Translation initiation factors, bacterial',\n",
       " 'Translation initiation factors, eukaryotic and archaeal',\n",
       " 'Translation termination factors, bacterial',\n",
       " 'Tricarballylate Utilization',\n",
       " 'Tricarboxylate transport system',\n",
       " 'Trimethylamine-N-oxide operon',\n",
       " 'Tripartite efflux system (of RND type) associated with Geranylgeranyl-PP synthase',\n",
       " 'Tripartite multidrug efflux systems (of MFS type) EmrKY-TolC and EmrAB-TolC',\n",
       " 'Tripartite multidrug efflux systems (of MFS type) in Gram-negative bacteria',\n",
       " 'Tripartite multidrug efflux systems (of RND type) AcrAB-TolC and related transporters',\n",
       " 'Tripartite multidrug efflux systems (of RND type) CmeABC and CmeDEF',\n",
       " 'Tripartite multidrug efflux systems (of RND type) MdtABC-TolC',\n",
       " 'Tripartite multidrug efflux systems (of RND type) in Pseudomonas',\n",
       " 'Trk and Ktr potassium uptake systems',\n",
       " 'Tryptophan synthesis',\n",
       " 'Twin-arginine translocation system',\n",
       " 'Twin-arginine translocation system related cluster',\n",
       " 'Two cell division clusters relating to chromosome partitioning',\n",
       " 'Two partner secretion pathway (TPS)',\n",
       " 'Type 4 conjugative transfer system, IncI1 type',\n",
       " 'Type 4 secretion and conjugative transfer',\n",
       " 'Type I restriction-modification systems',\n",
       " 'Type I secretion systems disambiguation',\n",
       " 'Type III secretion system',\n",
       " 'Type III secretion system orphans',\n",
       " 'Type III secretion systems',\n",
       " 'Type IV pilus',\n",
       " 'Type VI secretion system',\n",
       " 'Ubiquinone biosynthesis',\n",
       " 'Undecaprenyl-diphosphatases',\n",
       " 'Universal GTPases',\n",
       " 'Universal stress protein family',\n",
       " 'Urea carboxylase and Allophanate hydrolase cluster',\n",
       " 'Urea cycle',\n",
       " 'Urease subunits',\n",
       " 'Uridine to beta-Alanine and Thymine to 3-aminoisobutanoate module',\n",
       " 'Utilization systems for glycans and polysaccharides (PULs, SUS)',\n",
       " 'V-Type ATP synthase',\n",
       " 'Vancomycin resistance, D-Ala-D-Ser, D-Ala-Lac ligases',\n",
       " 'Vanillate and syringate utilization',\n",
       " 'Vibrioferrin synthesis',\n",
       " 'Vir-like type 4 secretion system',\n",
       " 'VraTSR and LiaFSR three-component regulatory systems',\n",
       " 'Wall polysaccharide pyruvylation',\n",
       " 'Widespread colonization island',\n",
       " 'Xanthine dehydrogenase subunits',\n",
       " 'YoeB-YefM toxin-antitoxin system',\n",
       " 'YrdC-YciO-Sua5 and associated protein families',\n",
       " 'ZraPRS-based zinc resistance',\n",
       " 'alpha-glucoside ABC transporter',\n",
       " 'beta-Barrel Assembly Machinery',\n",
       " 'dTDP-3-acetamido-3,6-dideoxy-alpha-D-galactose synthesis',\n",
       " 'dTDP-rhamnose synthesis',\n",
       " 'dTDP-rhamnose synthesis -- gjo',\n",
       " 'mRNA 3-end processing factor containing cluster',\n",
       " 'p-Aminobenzoyl-Glutamate Utilization',\n",
       " 'rRNA modification related cluster including tlyA',\n",
       " 'tRNA aminoacylation, Ala',\n",
       " 'tRNA aminoacylation, Arg',\n",
       " 'tRNA aminoacylation, Asn',\n",
       " 'tRNA aminoacylation, Asp',\n",
       " 'tRNA aminoacylation, Cys',\n",
       " 'tRNA aminoacylation, Gln',\n",
       " 'tRNA aminoacylation, Glu',\n",
       " 'tRNA aminoacylation, Gly',\n",
       " 'tRNA aminoacylation, His',\n",
       " 'tRNA aminoacylation, Ile',\n",
       " 'tRNA aminoacylation, Leu',\n",
       " 'tRNA aminoacylation, Lys',\n",
       " 'tRNA aminoacylation, Met',\n",
       " 'tRNA aminoacylation, Phe',\n",
       " 'tRNA aminoacylation, Pro',\n",
       " 'tRNA aminoacylation, Ser',\n",
       " 'tRNA aminoacylation, Thr',\n",
       " 'tRNA aminoacylation, Trp',\n",
       " 'tRNA aminoacylation, Tyr',\n",
       " 'tRNA aminoacylation, Val',\n",
       " 'tRNA modification, 5-oxyacetyl uridine (cmo5U34)',\n",
       " 'tRNA nucleotidyltransferase',\n",
       " 'tRNA splicing',\n",
       " 'tRNA thiolation',\n",
       " 'Aegiribacteria',\n",
       " 'Kapaibacterium',\n",
       " 'Stahlbacteria',\n",
       " 'Vermiphilus',\n",
       " 'Babela',\n",
       " 'Chromulinivorax',\n",
       " 'Ozemobacter',\n",
       " 'Desulfacyla',\n",
       " 'Dissulfurimicrobium',\n",
       " 'Dissulfurirhabdus',\n",
       " 'Bradymonas',\n",
       " 'Persicimonas',\n",
       " 'Lujinxingia',\n",
       " 'Microvenator',\n",
       " 'Acididesulfobacter',\n",
       " 'Acidulodesulfobacterium',\n",
       " 'Abditibacterium',\n",
       " 'Chloracidobacterium',\n",
       " 'Polarisedimenticola',\n",
       " 'Acanthopleuribacter',\n",
       " 'Sulfidibacter',\n",
       " 'Geothrix',\n",
       " 'Holophaga',\n",
       " 'Thermotomaculum',\n",
       " 'Angelobacter',\n",
       " 'Bryobacter',\n",
       " 'Paludibaculum',\n",
       " 'Solibacter',\n",
       " 'Sulfopaludibacter',\n",
       " 'Acidoferrum',\n",
       " 'Acidicapsa',\n",
       " 'Acidipila',\n",
       " 'Acidisarcina',\n",
       " 'Acidobacterium',\n",
       " 'Bryocella',\n",
       " 'Sulfotelmatomonas',\n",
       " 'Edaphobacter',\n",
       " 'Granulicella',\n",
       " 'Occallatibacter',\n",
       " 'Paracidobacterium',\n",
       " 'Pseudacidobacterium',\n",
       " 'Silvibacterium',\n",
       " 'Terracidiphilus',\n",
       " 'Terriglobus',\n",
       " 'Korobacter',\n",
       " 'Sulfotelmatobacter',\n",
       " 'Sulfomarinibacter',\n",
       " 'Thermoanaerobaculum',\n",
       " 'Luteitalea',\n",
       " 'Hakubella',\n",
       " 'Acidiferrimicrobium',\n",
       " 'Acidimicrobium',\n",
       " 'Aciditerrimonas',\n",
       " 'Acidithrix',\n",
       " 'Ferrimicrobium',\n",
       " 'Ferrithrix',\n",
       " 'Actinomarinicola',\n",
       " 'Aquihabitans',\n",
       " 'Dermatobacter',\n",
       " 'Iamia',\n",
       " 'Rhabdothermincola',\n",
       " 'Desertimonas',\n",
       " 'Ilumatobacter',\n",
       " 'Microthrix',\n",
       " 'Halopolyspora',\n",
       " 'Acidothermus',\n",
       " 'Actinobaculum',\n",
       " 'Actinomyces',\n",
       " 'Actinotignum',\n",
       " 'Ancrocorticia',\n",
       " 'Arcanobacterium',\n",
       " 'Boudabousia',\n",
       " 'Bowdeniella',\n",
       " 'Buchananella',\n",
       " 'Changpingibacter',\n",
       " 'Flaviflexus',\n",
       " 'Gleimia',\n",
       " 'Mobiluncus',\n",
       " 'Nanchangia',\n",
       " 'Neoactinobaculum',\n",
       " 'Pauljensenia',\n",
       " 'Peptidiphaga',\n",
       " 'Schaalia',\n",
       " 'Scrofimicrobium',\n",
       " 'Trueperella',\n",
       " 'Varibaculum',\n",
       " 'Winkia',\n",
       " 'Actinopolyspora',\n",
       " 'Aeriscardovia',\n",
       " 'Alloscardovia',\n",
       " 'Bifidobacterium',\n",
       " 'Bombiscardovia',\n",
       " 'Galliscardovia',\n",
       " 'Gardnerella',\n",
       " 'Parascardovia',\n",
       " 'Pseudoscardovia',\n",
       " 'Scardovia',\n",
       " 'Planktophila',\n",
       " 'Actinocrinis',\n",
       " 'Actinospica',\n",
       " 'Catenulispora',\n",
       " 'Cryptosporangium',\n",
       " 'Frankia',\n",
       " 'Parafrankia',\n",
       " 'Protofrankia',\n",
       " 'Pseudofrankia',\n",
       " 'Antricoccus',\n",
       " 'Blastococcus',\n",
       " 'Cumulibacter',\n",
       " 'Geodermatophilus',\n",
       " 'Goekera',\n",
       " 'Klenkia',\n",
       " 'Modestobacter',\n",
       " 'Trujillonella',\n",
       " 'Glycomyces',\n",
       " 'Haloglycomyces',\n",
       " 'Natronoglycomyces',\n",
       " 'Stackebrandtia',\n",
       " 'Jatrophihabitans',\n",
       " 'Haloactinopolyspora',\n",
       " 'Jiangella',\n",
       " 'Phytoactinopolyspora',\n",
       " 'Angustibacter',\n",
       " 'Kineococcus',\n",
       " 'Kineosporia',\n",
       " 'Pseudokineococcus',\n",
       " 'Quadrisphaera',\n",
       " 'Carbonactinospora',\n",
       " 'Actinacidiphila',\n",
       " 'Allostreptomyces',\n",
       " 'Embleya',\n",
       " 'Kitasatospora',\n",
       " 'Mangrovactinospora',\n",
       " 'Peterkaempfera',\n",
       " 'Phaeacidiphilus',\n",
       " 'Streptacidiphilus',\n",
       " 'Streptantibioticus',\n",
       " 'Streptomyces',\n",
       " 'Wenjunlia',\n",
       " 'Yinghuangia',\n",
       " 'Luteimicrobium',\n",
       " 'Pengzhenrongella',\n",
       " 'Timonella',\n",
       " 'Beutenbergia',\n",
       " 'Miniimonas',\n",
       " 'Salana',\n",
       " 'Serinibacter',\n",
       " 'Bogoriella',\n",
       " 'Georgenia',\n",
       " 'Brevibacterium',\n",
       " 'Sediminivirga',\n",
       " 'Spelaeicoccus',\n",
       " 'Actinotalea',\n",
       " 'Cellulomonas',\n",
       " 'Oerskovia',\n",
       " 'Paraoerskovia',\n",
       " 'Pseudactinotalea',\n",
       " 'Sediminihabitans',\n",
       " 'Demequina',\n",
       " 'Brachybacterium',\n",
       " 'Dermabacter',\n",
       " 'Devriesea',\n",
       " 'Helcobacillus',\n",
       " 'Allobranchiibius',\n",
       " 'Barrientosiimonas',\n",
       " 'Branchiibius',\n",
       " 'Calidifontibacter',\n",
       " 'Demetria',\n",
       " 'Dermacoccus',\n",
       " 'Flexivirga',\n",
       " 'Leekyejoonella',\n",
       " 'Luteipulveratus',\n",
       " 'Metallococcus',\n",
       " 'Rudaeicoccus',\n",
       " 'Yimella',\n",
       " 'Agilicoccus',\n",
       " 'Austwickia',\n",
       " 'Dermatophilus',\n",
       " 'Gephyromycinifex',\n",
       " 'Kineosphaera',\n",
       " 'Mobilicoccus',\n",
       " 'Piscicoccus',\n",
       " 'Arsenicicoccus',\n",
       " 'Humibacillus',\n",
       " 'Intrasporangium',\n",
       " 'Janibacter',\n",
       " 'Knoellia',\n",
       " 'Kribbia',\n",
       " 'Lapillicoccus',\n",
       " 'Marihabitans',\n",
       " 'Ornithinibacter',\n",
       " 'Ornithinicoccus',\n",
       " 'Oryzihumus',\n",
       " 'Pedococcus',\n",
       " 'Phycicoccus',\n",
       " 'Segeticoccus',\n",
       " 'Terrabacter',\n",
       " 'Terracoccus',\n",
       " 'Tetrasphaera',\n",
       " 'Flavimobilis',\n",
       " 'Jonesia',\n",
       " 'Kytococcus',\n",
       " 'Agreia',\n",
       " 'Agrococcus',\n",
       " 'Agromyces',\n",
       " 'Alpinimonas',\n",
       " 'Amnibacterium',\n",
       " 'Aquiluna',\n",
       " 'Aurantimicrobium',\n",
       " 'Lumbricidophila',\n",
       " 'Canibacter',\n",
       " 'Clavibacter',\n",
       " 'Cnuibacter',\n",
       " 'Compostimonas',\n",
       " 'Conyzicola',\n",
       " 'Cryobacterium',\n",
       " 'Curtobacterium',\n",
       " 'Diaminobutyricibacter',\n",
       " 'Diaminobutyricimonas',\n",
       " 'Frigoribacterium',\n",
       " 'Frondihabitans',\n",
       " 'Galbitalea',\n",
       " 'Glaciibacter',\n",
       " 'Glaciihabitans',\n",
       " 'Gryllotalpicola',\n",
       " 'Gulosibacter',\n",
       " 'Herbiconiux',\n",
       " 'Homoserinibacter',\n",
       " 'Homoserinimonas',\n",
       " 'Humibacter',\n",
       " 'Klugiella',\n",
       " ...]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df_combined.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a29357e3-b502-41c1-94c7-2bb9ca8e7b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = smf.mixedlm(\n",
    "    'N12M_Pseudomonas_aeruginosa ~ abundance + Pseudomonas_Culture + CS_Pseudomonas_aeruginosa', \n",
    "    merged_data,\n",
    "    groups=merged_data[\"pwCF_ID\"] \n",
    ")\n",
    "result = model.fit()\n",
    "\n",
    "# Print the model summary\n",
    "print(result.summary())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ca2cc9f-0782-415b-a47d-52cbe48fedaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "if False:\n",
    "    model = smf.mixedlm(\n",
    "        'abundance ~ CS_Pseudomonas_aeruginosa + taxa_functions', \n",
    "        merged_data,\n",
    "        groups=merged_data[\"pwCF_ID\"] \n",
    "    )\n",
    "    result = model.fit()\n",
    "\n",
    "    # Print the model summary\n",
    "    print(result.summary())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce94b2ee-0be5-4869-9769-08caf6a6789d",
   "metadata": {},
   "source": [
    "## Write the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f413e87e-e2d0-4b90-8bbf-cdb0509b531b",
   "metadata": {},
   "outputs": [],
   "source": [
    "coefficients = pd.DataFrame({\n",
    "    \"Variable\": result.params.index,\n",
    "    \"Estimate\": result.params.values,\n",
    "    \"Std Error\": result.bse.values,\n",
    "    \"P-value\": result.pvalues.values\n",
    "})\n",
    "coefficients.to_csv(os.path.join('lmm', f'{taxa}_model_coefficients.tsv'), sep=\"\\t\", index=False)\n",
    "\n",
    "with open(os.path.join('lmm', f'{taxa}_model_results.tsv'), 'w') as f:\n",
    "    f.write(result.summary().as_text())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1a55639-6ff9-4e37-8dd8-4bdd90e575bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "# Extract fixed effect coefficients and confidence intervals\n",
    "params = result.fe_params  # Fixed effect coefficients\n",
    "conf_int = result.conf_int()  # Confidence intervals\n",
    "conf_int.columns = ['lower', 'upper']\n",
    "\n",
    "# Combine coefficients and confidence intervals\n",
    "coefficients = pd.DataFrame({\n",
    "    'coef': params,\n",
    "    'lower': conf_int['lower'],\n",
    "    'upper': conf_int['upper']\n",
    "})\n",
    "\n",
    "# Plot the coefficients with error bars\n",
    "plt.figure(figsize=(8, 6))\n",
    "plt.errorbar(coefficients.index, coefficients['coef'], \n",
    "             yerr=(coefficients['coef'] - coefficients['lower'], coefficients['upper'] - coefficients['coef']), \n",
    "             fmt='o', capsize=5)\n",
    "plt.axhline(0, color='gray', linestyle='--', linewidth=1)\n",
    "plt.xticks(rotation=45)\n",
    "plt.title(\"Fixed Effect Coefficients with Confidence Intervals\")\n",
    "plt.ylabel(\"Coefficient Value\")\n",
    "plt.xlabel(\"Predictor\")\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4e2dacf-2866-4264-aad6-71b688c12fdc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bioinformatics]",
   "language": "python",
   "name": "conda-env-bioinformatics-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
