{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8617db5-c2c7-40fb-9cbc-a01cc326e9cb",
   "metadata": {},
   "source": [
    "# PCA Figure\n",
    "\n",
    "A recapitulation of Jess' primer figure but written in python by Rob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a9de0900-a0c5-4e98-986d-b676d9aeef89",
   "metadata": {},
   "source": [
    "df = pd.read_csv(\"../MGI/Taxonomy/MGI_reads_species.rawcounts.tsv.gz\", sep='\\t', compression='gzip')\n",
    "df = df[df['taxonomy'].str.contains('k__Bacteria')]\n",
    "df = df[~df['taxonomy'].str.endswith('s__')]\n",
    "df = df[~df['taxonomy'].str.contains('g__;')]\n",
    "df = df.set_index('taxonomy')\n",
    "df.index = df.index.str.split(';').str[-1].str.replace('s__', '')\n",
    "df = df.T\n",
    "df.head()"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e669e52-75c8-49da-a6fa-ef0d95582294",
   "metadata": {},
   "source": [
    "corrections = {\n",
    "    '1085876_20180227_S': '1723809_20180227_S',\n",
    "    '1447437_20171212_S': '1112926_20171212_S',\n",
    "    '1588281_20180508_S': '1598281_20180508_S',\n",
    "    '1590009_20171212_S': '1255498_20171212_S',\n",
    "    '1651490_20171215_S': '1316979_20171215_S',\n",
    "    '715927_20180226_S': '652927_20180226_S',\n",
    "    '778851_20170918_S': '777851_20170918_S'\n",
    "}"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4914f557-67ab-4253-9d1c-e53d11f23245",
   "metadata": {},
   "source": [
    "# Load metadata table\n",
    "# This is generic so we can copy/paste. Yes, it should be a function, but its not.\n",
    "\n",
    "# sequence_type = 'minion_ID'\n",
    "sequence_type = 'MGI_ID'\n",
    "\n",
    "# metadata = pd.read_csv(\"../Metadata/Metadata20241026.txt\", encoding='utf-8', sep=\"\\t\", index_col=0)\n",
    "metadata = pd.read_csv(\"../Metadata/Metadata20241026.txt\", encoding='windows-1252', sep=\"\\t\", index_col=0)\n",
    "metadata = metadata[~metadata[sequence_type].isna()]\n",
    "todrop = []\n",
    "for s in metadata[sequence_type]:\n",
    "    if s not in df.index:\n",
    "        print(f\"ERROR: {s} not found in data frame, dropped from metadata\", file=sys.stderr)\n",
    "        todrop.append(s)\n",
    "\n",
    "metadata.drop(todrop, inplace=True)\n",
    "\n",
    "todrop = []\n",
    "for s in df.index:\n",
    "    if not metadata[sequence_type].str.contains(s).any():\n",
    "        print(f\"ERROR: {s} not found in metadata, dropped from dataframe\", file=sys.stderr)\n",
    "        todrop.append(s)\n",
    "df.drop(todrop, inplace=True)\n",
    "        \n",
    "if metadata.shape[0] != df.shape[0]:\n",
    "    print(f\"ERROR: we have {metadata.shape[0]} rows in metadata and {df.shape[0]} data rows\", file=sys.stderr)\n",
    "metadata.head()"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4d1a5ced-5923-4ec6-b489-ff86eb95e102",
   "metadata": {},
   "source": [
    "sns.kdeplot(data=df['Pseudomonas aeruginosa'])"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "caeaf1b2-a912-4d90-9b78-2f857b552c5c",
   "metadata": {},
   "source": [
    "genus = 'Pseudomonas aeruginosa'\n",
    "list(df[df[genus] < 100][genus])\n",
    "#df[df[genus] > 0].shape"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5f827423-6a93-4049-8107-6397ac9596d3",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Create some bins for the sizes\n",
    "# Define the bins and labels\n",
    "# Non zero mean is the mean of all the non zero values\n",
    "genus = 'Pseudomonas aeruginosa'\n",
    "mn = np.mean(df[df[genus] > 0][genus])\n",
    "md = np.median(df[df[genus] > 0][genus])\n",
    "print(f\"The mean for {genus} is {mn} and the median is {md}\", file=sys.stderr)\n",
    "\n",
    "nonzeromean = np.median(df[df[genus] > 0][genus])\n",
    "bins = [-1, 0, nonzeromean/2, nonzeromean*2, float('inf')]  # Include -1 to capture 0 in 'small'\n",
    "\n",
    "#bins = [-1, 200, 5000, 10000, float('inf')]\n",
    "labels = ['small', 'medium', 'large']\n",
    "sizelist = [50, 200, 500, 1000]\n",
    "marker_shapes = ['s', 'V', '^', 'o'] # see https://seaborn.pydata.org/tutorial/properties.html#marker\n",
    "\n",
    "# Create a new column 'size' based on the conditions\n",
    "sizes = list(pd.cut(df['Pseudomonas aeruginosa'], bins=bins, labels=sizelist))\n",
    "markers = list(pd.cut(df['Pseudomonas aeruginosa'], bins=bins, labels=marker_shapes))"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "66f24f97-fc05-419b-98a1-0f47bd0cf95b",
   "metadata": {},
   "source": [
    "# Normalization/Transformation\n",
    "# Square root transform\n",
    "sqrt_df = np.sqrt(df)\n",
    "\n",
    "# perform the PCA\n",
    "pca = PCA(n_components=2)\n",
    "\n",
    "pca_result = pca.fit_transform(sqrt_df)\n",
    "# Create a DataFrame with PCA results\n",
    "pca_df = pd.DataFrame(data=pca_result, index=df.index, columns=['PC1', 'PC2'])\n",
    "\n",
    "# Get loadings\n",
    "loadings = pca.components_.T * np.sqrt(pca.explained_variance_)\n",
    "# Create a DataFrame for loadings with feature names\n",
    "# Select the top 5 most important loadings\n",
    "loadings_df = pd.DataFrame(loadings, index=df.columns, columns=['PC1', 'PC2'])\n",
    "\n",
    "# Create a DataFrame for top loadings\n",
    "top_loadings_df = loadings_df.loc[loadings_df['PC1'].abs().sort_values(ascending=False).index]\n",
    "top_loadings_df.head()\n",
    "\n",
    "\n",
    "explained_variance = pca.explained_variance_ratio_ * 100\n",
    "pc1_variance = explained_variance[0]\n",
    "pc2_variance = explained_variance[1]"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ddffccb8-afeb-4030-89aa-afab02646617",
   "metadata": {},
   "source": [
    "# Map colors based on the metadata column\n",
    "colours = np.where(metadata['CS_Pseudomonas aeruginosa'] == 1, 'blue', 'red')\n",
    "\n",
    "# Plot the PCA results\n",
    "plt.figure(figsize=(16, 12))\n",
    "plt.scatter(pca_df['PC1'], pca_df['PC2'], alpha=0.2, c=colours, s=sizes)\n",
    "plt.title('Pseudomonas detected by MGI sequencing coloured whether positive - present' + f\"\\nMedian number of '{genus}' sequences: {md:.0f}\")\n",
    "plt.xlabel(f'Principal Component 1 ({pc1_variance:.3f}%)')\n",
    "plt.ylabel(f'Principal Component 2 ({pc2_variance:.3f}%)')\n",
    "\n",
    "# add the loadings ... we only plot maxloadings here\n",
    "maxloadings = 15\n",
    "if len(loadings) < maxloadings:\n",
    "    maxloadings = len(loadings)\n",
    "\n",
    "plotscaler = 2\n",
    "texts = []\n",
    "for i in range(maxloadings):\n",
    "    xpos = top_loadings_df.iloc[i, 0]*plotscaler\n",
    "    ypos = top_loadings_df.iloc[i, 1]*plotscaler\n",
    "    plt.arrow(0, 0, xpos, ypos, \n",
    "              color='black', alpha=0.5, width=0.05)\n",
    "    texts.append(plt.text(xpos, ypos, top_loadings_df.index[i], color='black'))\n",
    "\n",
    "# specifically add Pseudomonas\n",
    "pidx = top_loadings_df.index.get_loc('Pseudomonas aeruginosa')\n",
    "xpos = top_loadings_df.iloc[pidx, 0]*plotscaler\n",
    "ypos = top_loadings_df.iloc[pidx, 1]*plotscaler\n",
    "plt.arrow(0, 0, xpos, ypos, color='black', alpha=0.5, width=0.05)\n",
    "texts.append(plt.text(xpos, ypos, top_loadings_df.index[pidx], color='black'))\n",
    "\n",
    "    \n",
    "adjust_text(texts)\n",
    "\n",
    "# Add a legend\n",
    "blue_patch = plt.Line2D([0], [0], marker='o', color='w', label='Pseudomonas culture positive', \n",
    "                         markerfacecolor='blue', alpha=0.2, markersize=10)\n",
    "red_patch = plt.Line2D([0], [0], marker='o', color='w', label='Pseudomonas culture negative', \n",
    "                        markerfacecolor='red', alpha=0.2, markersize=10)\n",
    "small_patch = plt.scatter([], [], color='k', label=f'{bins[1]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[0])\n",
    "med1_patch = plt.scatter([], [], color='k', label=f'{bins[1]:.0f} - {bins[2]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[1])\n",
    "med2_patch = plt.scatter([], [], color='k', label=f'{bins[2]:.0f} - {bins[3]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[2])\n",
    "big_patch = plt.scatter([], [], color='k', label=f'>{bins[3]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[3])\n",
    "\n",
    "plt.legend(handles=[blue_patch, red_patch, small_patch, med1_patch, med2_patch, big_patch])\n",
    "\n",
    "\n",
    "plt.savefig('img/MGI_species_pseudomonas_present.png')\n",
    "# Show the plot\n",
    "plt.show()"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ddd2dddc-e95d-49b8-8ca6-e261525ff5b8",
   "metadata": {},
   "source": [
    "columns_to_check = ['CS_Pseudomonas aeruginosa', 'P12M_2_Pseudomonas aeruginosa']\n",
    "colours = metadata[columns_to_check].isin([1]).any(axis=1).map({True: 'blue', False: 'red'})\n",
    "\n",
    "# Plot the PCA results\n",
    "plt.figure(figsize=(16, 12))\n",
    "plt.scatter(pca_df['PC1'], pca_df['PC2'], alpha=0.2, c=colours, s=sizes)\n",
    "plt.title('Pseudomonas detected by MGI sequencing coloured whether positive - past or present' + f\"\\nMedian number of '{genus}' sequences: {md:.0f}\")\n",
    "plt.xlabel(f'Principal Component 1 ({pc1_variance:.3f}%)')\n",
    "plt.ylabel(f'Principal Component 2 ({pc2_variance:.3f}%)')\n",
    "\n",
    "# add the loadings ... we only plot maxloadings here\n",
    "maxloadings = 15\n",
    "if len(loadings) < maxloadings:\n",
    "    maxloadings = len(loadings)\n",
    "\n",
    "plotscaler = 2\n",
    "texts = []\n",
    "for i in range(maxloadings):\n",
    "    xpos = top_loadings_df.iloc[i, 0]*plotscaler\n",
    "    ypos = top_loadings_df.iloc[i, 1]*plotscaler\n",
    "    plt.arrow(0, 0, xpos, ypos, \n",
    "              color='black', alpha=0.5, width=0.05)\n",
    "    texts.append(plt.text(xpos, ypos, top_loadings_df.index[i], color='black'))\n",
    "\n",
    "# specifically add Pseudomonas\n",
    "pidx = top_loadings_df.index.get_loc('Pseudomonas aeruginosa')\n",
    "xpos = top_loadings_df.iloc[pidx, 0]*plotscaler\n",
    "ypos = top_loadings_df.iloc[pidx, 1]*plotscaler\n",
    "plt.arrow(0, 0, xpos, ypos, color='black', alpha=0.5, width=0.05)\n",
    "texts.append(plt.text(xpos, ypos, top_loadings_df.index[pidx], color='black'))\n",
    "\n",
    "    \n",
    "adjust_text(texts)\n",
    "\n",
    "# Add a legend\n",
    "blue_patch = plt.Line2D([0], [0], marker='o', color='w', label='Pseudomonas culture positive', \n",
    "                         markerfacecolor='blue', alpha=0.2, markersize=10)\n",
    "red_patch = plt.Line2D([0], [0], marker='o', color='w', label='Pseudomonas culture negative', \n",
    "                        markerfacecolor='red', alpha=0.2, markersize=10)\n",
    "small_patch = plt.scatter([], [], color='k', label=f'{bins[1]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[0])\n",
    "med1_patch = plt.scatter([], [], color='k', label=f'{bins[1]:.0f} - {bins[2]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[1])\n",
    "med2_patch = plt.scatter([], [], color='k', label=f'{bins[2]:.0f} - {bins[3]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[2])\n",
    "big_patch = plt.scatter([], [], color='k', label=f'>{bins[3]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[3])\n",
    "\n",
    "plt.legend(handles=[blue_patch, red_patch, small_patch, med1_patch, med2_patch, big_patch])\n",
    "\n",
    "\n",
    "plt.savefig('img/MGI_species_pseudomonas_past_present.png')\n",
    "# Show the plot\n",
    "plt.show()"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "06db2650-ff8f-48ce-bbac-2a3b447c2a2c",
   "metadata": {},
   "source": [
    "columns_to_check = ['CS_Pseudomonas aeruginosa', 'N12M_Pseudomonas aeruginosa', 'P12M_2_Pseudomonas aeruginosa']\n",
    "colours = metadata[columns_to_check].isin([1]).any(axis=1).map({True: 'blue', False: 'red'})\n",
    "\n",
    "# Plot the PCA results\n",
    "plt.figure(figsize=(16, 12))\n",
    "plt.scatter(pca_df['PC1'], pca_df['PC2'], alpha=0.2, c=colours, s=sizes)\n",
    "plt.title('Pseudomonas detected by MGI sequencing coloured whether positive - past, present, or future' + f\"\\nMedian number of '{genus}' sequences: {md:.0f}\")\n",
    "plt.xlabel(f'Principal Component 1 ({pc1_variance:.3f}%)')\n",
    "plt.ylabel(f'Principal Component 2 ({pc2_variance:.3f}%)')\n",
    "\n",
    "# add the loadings ... we only plot maxloadings here\n",
    "maxloadings = 15\n",
    "if len(loadings) < maxloadings:\n",
    "    maxloadings = len(loadings)\n",
    "\n",
    "plotscaler = 2\n",
    "texts = []\n",
    "for i in range(maxloadings):\n",
    "    xpos = top_loadings_df.iloc[i, 0]*plotscaler\n",
    "    ypos = top_loadings_df.iloc[i, 1]*plotscaler\n",
    "    plt.arrow(0, 0, xpos, ypos, \n",
    "              color='black', alpha=0.5, width=0.05)\n",
    "    texts.append(plt.text(xpos, ypos, top_loadings_df.index[i], color='black'))\n",
    "\n",
    "# specifically add Pseudomonas\n",
    "pidx = top_loadings_df.index.get_loc('Pseudomonas aeruginosa')\n",
    "xpos = top_loadings_df.iloc[pidx, 0]*plotscaler\n",
    "ypos = top_loadings_df.iloc[pidx, 1]*plotscaler\n",
    "plt.arrow(0, 0, xpos, ypos, color='black', alpha=0.5, width=0.05)\n",
    "texts.append(plt.text(xpos, ypos, top_loadings_df.index[pidx], color='black'))\n",
    "\n",
    "    \n",
    "adjust_text(texts)\n",
    "\n",
    "# Add a legend\n",
    "blue_patch = plt.Line2D([0], [0], marker='o', color='w', label='Pseudomonas culture positive', \n",
    "                         markerfacecolor='blue', alpha=0.2, markersize=10)\n",
    "red_patch = plt.Line2D([0], [0], marker='o', color='w', label='Pseudomonas culture negative', \n",
    "                        markerfacecolor='red', alpha=0.2, markersize=10)\n",
    "small_patch = plt.scatter([], [], color='k', label=f'{bins[1]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[0])\n",
    "med1_patch = plt.scatter([], [], color='k', label=f'{bins[1]:.0f} - {bins[2]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[1])\n",
    "med2_patch = plt.scatter([], [], color='k', label=f'{bins[2]:.0f} - {bins[3]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[2])\n",
    "big_patch = plt.scatter([], [], color='k', label=f'>{bins[3]:.0f} sequences', \n",
    "                        alpha=0.2, s=sizelist[3])\n",
    "\n",
    "plt.legend(handles=[blue_patch, red_patch, small_patch, med1_patch, med2_patch, big_patch])\n",
    "\n",
    "plt.savefig('img/MGI_species_pseudomonas_past_present_future.png')\n",
    "# Show the plot\n",
    "plt.show()"
   ],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "997ebebc-fc03-48f9-8510-62218ea0b53b",
   "metadata": {},
   "source": [],
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22567523-b3f0-41a2-bd74-2ff0d4c5f92f",
   "metadata": {},
   "source": [],
   "outputs": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BioinfieConda",
   "language": "python",
   "name": "bioinfieconda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
